{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PB: Leave-one-out-Labels\n",
    "\n",
    "* The following notebook is trained on data generated from revised R script [Oct 12 2017]\n",
    "    * Exact Match [1] and Homozygous [0] Reference data points\n",
    "    * Removed all data points with Gtcons and GTconswithoutXX -1\n",
    "* 5k randomly selected deletions test data was also processed through same R script\n",
    "* Balanced Training Set for GTconswithoutXX labels:\n",
    "    * 200 Hom Var\n",
    "    * 200 Hom Ref\n",
    "    * 200 Het Var\n",
    "* **Train/Prediction Label:** GTconswithoutXX\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/cross_validation.py:44: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "\n",
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/grid_search.py:43: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Imports\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from fancyimpute import KNN\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy import stats\n",
    "from matplotlib import pyplot\n",
    "from sklearn import preprocessing\n",
    "from scipy.linalg import svd\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import seaborn as sns\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "import plotly.plotly as py\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import metrics\n",
    "from sklearn.grid_search import GridSearchCV\n",
    "from sklearn.metrics import f1_score, precision_score\n",
    "from sklearn import preprocessing\n",
    "from ggplot import *\n",
    "from bokeh.charts import TimeSeries\n",
    "from bokeh.models import HoverTool\n",
    "from bokeh.plotting import show\n",
    "from bokeh.charts import Scatter, Histogram, output_file, show\n",
    "from bokeh.plotting import figure, show, output_file, ColumnDataSource\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.charts import Bar, output_file, show\n",
    "import bokeh.palettes as palettes\n",
    "from bokeh.models import HoverTool, BoxSelectTool, Legend\n",
    "from sklearn import (manifold, datasets, decomposition, ensemble,\n",
    "                     discriminant_analysis, random_projection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>sample</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HG002</td>\n",
       "      <td>13</td>\n",
       "      <td>37165972</td>\n",
       "      <td>37166066</td>\n",
       "      <td>-93</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>171.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17450.66667</td>\n",
       "      <td>7372.823301</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11405.2</td>\n",
       "      <td>4795.917183</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom sample  id     start       end  Size       type    SVtype  refN_cnt  \\\n",
       "0     1  HG002  13  37165972  37166066   -93  Insertion  Deletion         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            171.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       198.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            4.0               17450.66667   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7372.823301              15.0                     11405.2   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4795.917183                              15.0  \n",
       "\n",
       "[1 rows x 41 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Training Data\n",
    "# SVanalyzer generated training data\n",
    "df_train = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/training_data/tech_sep/HG002/DEL/PB_EM_HR_HG002_DEL_balanced_label_LOO_min1.csv')\n",
    "df_train_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/training_data/tech_sep/HG002/DEL/PB_EM_HR_HG002_DEL_balanced_label_LOO_min1.csv')\n",
    "df_train.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set = pd.DataFrame()\n",
    "train_set = df_train_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set['GTcons'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "train_set['GTcons'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "train_set['GTcons'].replace(2, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='imbalance'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    200\n",
       "1    200\n",
       "2    150\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(train_set['GTconswithoutpacbio.GT'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train the model only on the rows that have an Exact Match or Homozygous Reference Label\n",
    "# This step removes any row that has in 'Inaccurate Call' label\n",
    "df_train = df_train[(df_train['Label'] == 1) | (df_train['Label'] == 0)]\n",
    "df_train_2 = df_train_2[(df_train_2['Label'] == 1) | (df_train_2['Label'] == 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    131\n",
       "1    419\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# There are only Exact Match [1] and Homozygous Reference Labels [0]\n",
    "pd.value_counts(df_train['Label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='hom_ref'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>sample</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HG002</td>\n",
       "      <td>859</td>\n",
       "      <td>37568322</td>\n",
       "      <td>37568587</td>\n",
       "      <td>-115</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.37823</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom sample   id     start       end  Size       type    SVtype  refN_cnt  \\\n",
       "0     1  HG002  859  37568322  37568587  -115  Insertion  Deletion         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0               7611.37823              31.0                 8407.903226   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "\n",
       "[1 rows x 40 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Test Data\n",
    "# SVanalyzer generated training data\n",
    "df_test = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/test/final_df/tech_sep/DEL/HG002/PB_HG002_DEL_min1.csv')\n",
    "df_test_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/test/final_df/tech_sep/DEL/HG002/PB_HG002_DEL_min1.csv')\n",
    "df_test.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_test.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Label'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(c) - set(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_train.drop(['Label'], axis=1, inplace = True)\n",
    "# df_train.drop(['GTconswithoutIll300x.GT'], axis=1, inplace = True)\n",
    "df_train.drop(['GTcons'], axis=1, inplace = True)\n",
    "df_train.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_train.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_train.drop(['sample'], axis=1, inplace = True)\n",
    "df_train.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_train.drop(['type'], axis=1, inplace = True)\n",
    "df_train.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>37165972</td>\n",
       "      <td>37166066</td>\n",
       "      <td>-93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.829787</td>\n",
       "      <td>...</td>\n",
       "      <td>171.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17450.66667</td>\n",
       "      <td>7372.823301</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11405.2</td>\n",
       "      <td>4795.917183</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end  Size  refN_cnt  refN_pct  segdup_cnt  segdup_pct  \\\n",
       "0     1  37165972  37166066   -93         0         0           0         0.0   \n",
       "\n",
       "   tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0              2       0.829787                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            171.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       198.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            4.0               17450.66667   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7372.823301              15.0                     11405.2   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4795.917183                              15.0  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_train['chrom'].replace('X', 23, inplace=True)\n",
    "df_train['chrom'].replace('Y', 24, inplace=True)\n",
    "df_test['chrom'].replace('X', 23, inplace=True)\n",
    "df_test['chrom'].replace('Y', 24, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GTconflict', 'GTcons', 'GTsupp', 'SVtype', 'id', 'sample', 'type'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(d) - set(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_test.drop(['GTconswithoutpacbio.GT'], axis=1, inplace = True)\n",
    "df_test.drop(['GTcons'], axis=1, inplace = True)\n",
    "df_test.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_test.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_test.drop(['sample'], axis=1, inplace = True)\n",
    "df_test.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_test.drop(['type'], axis=1, inplace = True)\n",
    "df_test.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Impute missing values using KNN\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>37165972</td>\n",
       "      <td>37166066</td>\n",
       "      <td>-93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.829787</td>\n",
       "      <td>...</td>\n",
       "      <td>171.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17450.66667</td>\n",
       "      <td>7372.823301</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11405.20000</td>\n",
       "      <td>4795.917183</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>65326531</td>\n",
       "      <td>65326651</td>\n",
       "      <td>-120</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>83753489</td>\n",
       "      <td>83753698</td>\n",
       "      <td>-209</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>82.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18304.63415</td>\n",
       "      <td>4939.332831</td>\n",
       "      <td>41.0</td>\n",
       "      <td>11648.90244</td>\n",
       "      <td>2866.042175</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end  Size  refN_cnt  refN_pct  segdup_cnt  segdup_pct  \\\n",
       "0     1  37165972  37166066   -93         0         0           0         0.0   \n",
       "1     1  65326531  65326651  -120         0         0           0         0.0   \n",
       "2     1  83753489  83753698  -209         0         0           1         1.0   \n",
       "\n",
       "   tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0              2       0.829787                ...                  \n",
       "1              1       1.000000                ...                  \n",
       "2              1       1.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            171.0   \n",
       "1                                            187.0   \n",
       "2                                             82.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       198.0                             0.0   \n",
       "1                       236.0                             1.0   \n",
       "2                       160.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            4.0               17450.66667   \n",
       "1                           11.0               18112.46296   \n",
       "2                            7.0               18304.63415   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7372.823301              15.0                 11405.20000   \n",
       "1              7134.848382              54.0                 11277.83333   \n",
       "2              4939.332831              41.0                 11648.90244   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4795.917183                              15.0  \n",
       "1                4197.626206                              54.0  \n",
       "2                2866.042175                              41.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store training data in a new variable which will be converted to a matrix\n",
    "X = df_train\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/550 with 0 missing, elapsed time: 0.103\n",
      "Imputing row 101/550 with 0 missing, elapsed time: 0.104\n",
      "Imputing row 201/550 with 22 missing, elapsed time: 0.105\n",
      "Imputing row 301/550 with 0 missing, elapsed time: 0.188\n",
      "Imputing row 401/550 with 0 missing, elapsed time: 0.189\n",
      "Imputing row 501/550 with 0 missing, elapsed time: 0.190\n"
     ]
    }
   ],
   "source": [
    "# Convert dataframe to matrix\n",
    "X=X.as_matrix()\n",
    "\n",
    "#Imput missing values from three closest observations\n",
    "X_imputed=KNN(k=3).complete(X)\n",
    "X=pd.DataFrame(X_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37165972.0</td>\n",
       "      <td>37166066.0</td>\n",
       "      <td>-93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.829787</td>\n",
       "      <td>...</td>\n",
       "      <td>171.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17450.66667</td>\n",
       "      <td>7372.823301</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11405.20000</td>\n",
       "      <td>4795.917183</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>65326531.0</td>\n",
       "      <td>65326651.0</td>\n",
       "      <td>-120.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>83753489.0</td>\n",
       "      <td>83753698.0</td>\n",
       "      <td>-209.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>82.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18304.63415</td>\n",
       "      <td>4939.332831</td>\n",
       "      <td>41.0</td>\n",
       "      <td>11648.90244</td>\n",
       "      <td>2866.042175</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom       start         end   Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0  37165972.0  37166066.0  -93.0       0.0       0.0         0.0   \n",
       "1    1.0  65326531.0  65326651.0 -120.0       0.0       0.0         0.0   \n",
       "2    1.0  83753489.0  83753698.0 -209.0       0.0       0.0         1.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0            2.0       0.829787                ...                  \n",
       "1         0.0            1.0       1.000000                ...                  \n",
       "2         1.0            1.0       1.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            171.0   \n",
       "1                                            187.0   \n",
       "2                                             82.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       198.0                             0.0   \n",
       "1                       236.0                             1.0   \n",
       "2                       160.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            4.0               17450.66667   \n",
       "1                           11.0               18112.46296   \n",
       "2                            7.0               18304.63415   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7372.823301              15.0                 11405.20000   \n",
       "1              7134.848382              54.0                 11277.83333   \n",
       "2              4939.332831              41.0                 11648.90244   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4795.917183                              15.0  \n",
       "1                4197.626206                              54.0  \n",
       "2                2866.042175                              41.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header values in a list, will be used later to re-label the matrix post KNN imputation\n",
    "dftrain_header = list(df_train.columns.values)\n",
    "X.columns = dftrain_header\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Store Labels in a new 'Y' DataFrame\n",
    "Y = pd.DataFrame()\n",
    "Y = X['GTconswithoutpacbio.GT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    200\n",
       "1.0    200\n",
       "2.0    150\n",
       "dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Count the number of labels\n",
    "pd.value_counts(Y.values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Note: originally selected 1000 of each label --> find out why some are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Remove labels from feature set\n",
    "X.drop(['GTconswithoutpacbio.GT'],axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X4 = X.reindex_axis(sorted(X.columns), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Machine Learning\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='machine_learning'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the following section a random forest model will be trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](http://scikit-learn.org/0.16/modules/generated/sklearn.cross_validation.train_test_split.html) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data. The trained model will be used in a [later section](#predict) to predict the consensus GT for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "       * In the following section, svanalyzer data was used to train a random forest (RF) model. The features for the svanalyzer dataset include: svviz features, GA4GH features [RefN, Segmental Duplications, Tandem Repeat], preliminary R script analysis [consensus GT, GTsup].\n",
    "       * The RF classifier will predict the consensus GT labels:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "       \n",
    "       * In the [following section](#prediction_step), the trained RF model will be used to predict labels for genotype labels for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Train Random Forest Classifier **"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='train_test'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train Test Split\n",
    "# Train on 70% of the data and test on 30%\n",
    "X_train, X_test, y_train, y_test = train_test_split(X4, Y, test_size=0.3, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n",
       "            max_depth=None, max_features='auto', max_leaf_nodes=None,\n",
       "            min_impurity_split=1e-07, min_samples_leaf=1,\n",
       "            min_samples_split=2, min_weight_fraction_leaf=0.0,\n",
       "            n_estimators=10, n_jobs=1, oob_score=False, random_state=None,\n",
       "            verbose=0, warm_start=False)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%timeit\n",
    "# Train Random Forest Classifier\n",
    "model = RandomForestClassifier() \n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature ranking:\n",
      "1. feature 8 (0.110852)\n",
      "2. feature 5 (0.106484)\n",
      "3. feature 21 (0.087524)\n",
      "4. feature 6 (0.075837)\n",
      "5. feature 20 (0.067547)\n",
      "6. feature 9 (0.066713)\n",
      "7. feature 23 (0.062730)\n",
      "8. feature 4 (0.058097)\n",
      "9. feature 24 (0.050953)\n",
      "10. feature 3 (0.033797)\n",
      "11. feature 22 (0.027929)\n",
      "12. feature 19 (0.025084)\n",
      "13. feature 0 (0.022663)\n",
      "14. feature 1 (0.021099)\n",
      "15. feature 14 (0.020681)\n",
      "16. feature 13 (0.018100)\n",
      "17. feature 7 (0.014982)\n",
      "18. feature 27 (0.012432)\n",
      "19. feature 18 (0.011582)\n",
      "20. feature 31 (0.011521)\n",
      "21. feature 29 (0.011455)\n",
      "22. feature 15 (0.011403)\n",
      "23. feature 10 (0.010423)\n",
      "24. feature 11 (0.010344)\n",
      "25. feature 30 (0.010166)\n",
      "26. feature 12 (0.009770)\n",
      "27. feature 2 (0.009590)\n",
      "28. feature 28 (0.008085)\n",
      "29. feature 16 (0.006770)\n",
      "30. feature 17 (0.005389)\n",
      "31. feature 25 (0.000000)\n",
      "32. feature 26 (0.000000)\n"
     ]
    }
   ],
   "source": [
    "# Feature Importance\n",
    "importances = model.feature_importances_\n",
    "indices = np.argsort(importances)[::-1]\n",
    "print(\"Feature ranking:\")\n",
    "\n",
    "for f in range(X_test.shape[1]):\n",
    "    print(\"%d. feature %d (%f)\" % (f + 1, indices[f], importances[indices[f]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Another Resource](https://stackoverflow.com/questions/37877542/how-to-label-the-feature-importance-with-forests-of-trees)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='feature_importance'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/ipykernel/__main__.py:3: FutureWarning:\n",
      "\n",
      "sort is deprecated, use sort_values(inplace=True) for INPLACE sorting\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x116a367b8>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAHeCAYAAABg00vYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXu8rdW8/997t7uodnTZosPRkdZHueSSS0q5nHCQ23GE\n04mShHKJg1JIwnEkHKWLNhUVcaIcSuRWSTioqK+z8SNddLHtbrqv3x/fZ7aeNddzGeOZc6619lzf\n9+u1X2vPOcd8xhjzmfM7xvheF01OThIEQRCMF4vnegBBEATB8AnhHgRBMIaEcA+CIBhDQrgHQRCM\nISHcgyAIxpAQ7kEQBGPIkrkeQI/rrrtphk/mhhuuy8qVtya9P7XtsNvFNRfmNcdtPnHN1ff7sWzZ\n0kVV7ef1zn3JkjWG3nbY7eKaC/Oa4zafuOZ4fT9gngv3IAiCoBsh3IMgCMaQEO5BEARjSAj3IAiC\nMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGEPmTYRqEARBMJM9P3zujOeWv+sZre+LnXsQBMEY\nEsI9CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIKYgiAIZpmq\nwCRIC05KJXbuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMIa1+\n7pIWA0cB2wC3A3uZ2Yq+NusC5wCvMbPLU94TBEEQjI6UnfuLgHXMbDvgXcDh5RclbQv8ANgi9T1B\nEATBaEkR7jsAZwGY2YXAtn2vrw28GLg84z1BEATBCEkR7hsAq0qP75Z0rzrHzM43syty3hMEQRCM\nlhSBeyOwtPR4sZndNez3bLjhuixZssaM55ctW1rRuprUtsNuF9dcmNcct/nENefu+zGKa6cI9/OB\nXYAvSXoycMko3rNy5a0znlu2bCnXXXdTQnfpbYfdLq65MK85bvOJa87d96NM6nvK7eoEfYpwPx3Y\nWdIFwCJgD0mvBNY3s2NT35M04iAIgmAotAp3M7sH2Kfv6csr2j2t5T1BEATBLBFBTEEQBGNICPcg\nCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiC\nMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEk\nhHsQBMEYklIgOwiCIEhgzw+fO+O55e96xhyMJHbuQRAEY0kI9yAIgjEkhHsQBMEYEsI9CIJgDAnh\nHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxpDX9gKTFwFHANsDtwF5mtqL0\n+i7Ae4C7gOVmdpykNYETgM2Bu4HXmtnlwx9+EARBUEXKzv1FwDpmth3wLuDw3guFED8CeBawE7C3\npE2B5wJLzOwpwPuBw4Y98CAIgqCeFOG+A3AWgJldCGxbem0rYIWZrTSzO4DzgB2B3wBLil3/BsCd\nQx11EARB0EhKVsgNgFWlx3dLWmJmd1W8dhNwX+BmXCVzObAJ8Py2TjbccF2WLFljxvPLli1NGGJe\n22G3i2suzGuO23zimsO/5lyOIUW43wiUr7S4EOxVry0F/gq8FTjbzA6Q9GDgXEmPMrPb6jpZufLW\nGc8tW7aU6667KWGI6W2H3S6uuTCvOW7ziWsO/5o9ctqnti23qxP0KWqZ83EdOpKeDFxSeu0yYEtJ\nG0laC1fJ/AhYydSO/i/AmsDMbXkQBEEwElJ27qcDO0u6AFgE7CHplcD6ZnaspP2Bs/GFYrmZXSnp\nCGC5pB8CawEHmtktI5pDEARB0EercDeze4B9+p6+vPT6mcCZfe+5GXjZMAYYBEEQ5BNBTEEQBGNI\nCPcgCIIxJIR7EATBGJJiUA2CIFiw7PnhcyufX/6uZ8zySPKInXsQBMEYEsI9CIJgDAnhHgRBMIaE\ncA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAP\ngiAYQ0K4B0EQjCEh3IMgCMaQSPkbBMGCpCqV73xP45tD7NyDIAjGkNi5B0EwNqyuhTVGQezcgyAI\nxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhCtkEATznnEPOBoFsXMPgiAYQ1p3\n7pIWA0cB2wC3A3uZ2YrS67sA7wHuApab2XHF8wcALwDWAo4ys+OHP/wgCIKgihS1zIuAdcxsO0lP\nBg4HXgggaU3gCOAJwC3A+ZLOALYCngJsD6wLvH0EYw+CIAhqSFHL7ACcBWBmFwLbll7bClhhZivN\n7A7gPGBH4NnAJcDpwJnA14c56CAIgqCZlJ37BsCq0uO7JS0xs7sqXrsJuC+wCfAQ4PnAPwBnSHq4\nmU3WdbLhhuuyZMkaM55ftmxpwhDz2g67XVxzYV5z3OazOl1zPoxhvl8zRbjfCJSvtLgQ7FWvLQX+\nCtwAXF7s5k3SbcAy4Nq6TlauvHXGc8uWLeW6625KGGJ622G3i2suzGuO23xWp2v2yGmf2nZ1vGad\noE9Ry5wPPBeg0LlfUnrtMmBLSRtJWgtXyfwIV888R9IiSZsB6+ECPwiCIJgFUnbupwM7S7oAWATs\nIemVwPpmdqyk/YGz8YViuZldCVwpaUfgouL5N5rZ3aOZQhAEQdBPq3A3s3uAffqevrz0+pm40bT/\nfe8YeHRBEIw1EZw0OiKIKQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3\nIAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhKVkhgyAIkolkYPOD2LkHQRCM\nISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh\n3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENai3VIWgwcBWwD\n3A7sZWYrSq/vArwHuAtYbmbHlV67P/AzYGczu3zIYw+CIAhqSNm5vwhYx8y2A94FHN57QdKawBHA\ns4CdgL0lbVp67Rjgb8MedBAEQdBMinDfATgLwMwuBLYtvbYVsMLMVprZHcB5wI7Fax8FjgauGt5w\ngyAIghRShPsGwKrS47slLal57SbgvpJeDVxnZmcPZZRBEARBFikFsm8ElpYeLzazu2peWwr8FXgT\nMCnpH4HHACdKeoGZXVPXyYYbrsuSJWvMeH7ZsqUVratJbTvsdnHNhXnNcZvPqK45H8awEK+ZItzP\nB3YBviTpycAlpdcuA7aUtBFwM66S+aiZfbnXQNL3gH2aBDvAypW3znhu2bKlXHfdTQlDTG877HZx\nzYV5zXGbz6iu2SOnfWrbuKZTJ+hThPvpwM6SLgAWAXtIeiWwvpkdK2l/4GxcxbPczK5MGl0QBEEw\nMlqFu5ndA+zT9/TlpdfPBM5seP/Tug4uCIIg6EYEMQVBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEk\nhHsQBMEYEsI9CIJgDAnhHgRBMIakBDEFQRCw54fPnfHc8nc9Yw5GEqQQO/cgCIIxJIR7EATBGBLC\nPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgS6QeC\nYAETKQXGl9i5B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgS3jJBMIaEF0wQO/cgCIIx\nJIR7EATBGBLCPQiCYAwJnXsQrCaEHj3IIXbuQRAEY0jrzl3SYuAoYBvgdmAvM1tRen0X4D3AXcBy\nMztO0prAcmBzYG3gA2Z2xvCHHwRBEFSRopZ5EbCOmW0n6cnA4cALAQohfgTwBOAW4HxJZwDPBW4w\ns3+TtBHwCyCEexBUEOqWYBSkCPcdgLMAzOxCSduWXtsKWGFmKwEknQfsCJwGfLloswjf1QfBgqFK\nYEMI7WD2SBHuGwCrSo/vlrTEzO6qeO0m4L5mdjOApKW4kD+orZMNN1yXJUvWmPH8smVLE4aY13bY\n7eKaC/OaOX3PhzHENRfWNVOE+41A+UqLC8Fe9dpS4K8Akh4MnA4cZWYnt3WycuWtM55btmwp1113\nU8IQ09sOu11cc2FeM6fvMqnvGXa7uOb4XrNO0KcI9/OBXYAvFTr3S0qvXQZsWejVb8ZVMh+VtCnw\nLWBfM/tO0miDIAiCoZEi3E8HdpZ0Aa4/30PSK4H1zexYSfsDZ+NulcvN7EpJnwA2BA6WdHBxnX8y\ns7+NYA5BEARBH63C3czuAfbpe/ry0utnAmf2vefNwJuHMcAgmE+EZ0uwuhBBTEEQBGNICPcgCIIx\nJIR7EATBGBLCPQiCYAwJ4R4EQTCGRMrfYMETqQKCcSR27kEQBGNICPcgCIIxJIR7EATBGBI692Bs\niWjSYCETO/cgCIIxJHbuQTY5O+LUtoO0a+o/CBYqsXMPgiAYQ2LnPuaMYpcdBMH8J4T7PCIEcRAE\nwyKE+ywQgjgIgtkmdO5BEARjSOzcOxJeG0EQzGdi5x4EQTCGhHAPgiAYQ0It00cYP4MgGAdi5x4E\nQTCGLIidexg/gyBYaMTOPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSKsr\npKTFwFHANsDtwF5mtqL0+i7Ae4C7gOVmdlzbe4IgCILRkrJzfxGwjpltB7wLOLz3gqQ1gSOAZwE7\nAXtL2rTpPUEQBMHoSRHuOwBnAZjZhcC2pde2AlaY2UozuwM4D9ix5T1BEATBiEkR7hsAq0qP75a0\npOa1m4D7trwnCIIgGDGLJicnGxtI+hhwoZl9qXj8JzN7UPH/RwMfNrPnFo+PAM4HnlL3niAIgmD0\npOzczwd6wvvJwCWl1y4DtpS0kaS1cJXMj1reEwRBEIyYlJ17z/Pl0cAiYA/gccD6ZnZsyVtmMe4t\nc2TVe8zs8tFNIwiCICjTKtyDIAiC1Y8IYgqCIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGPIgggskvT3\nfU/dCVxvZncO0jYIgvFH0oPM7E+lxzIzG7TtqJl33jKSPmVm+5Yen2hmu/e1WQNYAzgV2BV3t1wM\nfMPMZhRGlXQx8CDAgAngFnxhe4eZfb5L27ZxSroamATWBtYFriiue62Zbd5lPjnzlvSe/s+hh5m9\nv+IzOhB4B3Brcd1JM9us6v2S/g74D+D+wGnAxWb244p2B5nZB0qPP2RmB1S0a73nfe03AO4BXgx8\n3cxWll7L+m4U79kZ2B+/VwA0tH0w8ApgnVLb95deT77vpfE+ou96F9X0/Ujg08CGwOeBS83s6xXt\nFgFP6LvmD2qu2Th3SV80s12r3ltHMadXAw8Bzi3GeX2XcUrasa6fqjml9C3pu/g9qrpmee6PBHrf\n9XcUT68BfMjMHtN3zaS2qX0XbbPm3s+82blLeiNwELCRpJcUTy8GflXRfE/gQOABuBBehP/Yf1hz\n+d8DzzCz6yVtCHwGeC3wTfxHktw2dZxm9sBiXp8HDjCzKyRthida6zqfnHn/ufj7omJO5+M/pP6T\nSY9dgc3M7Naa18sciyeDOxj4AXAC8OTei5JeA+wFbCXpucXTawBrAgeU2uXc8957TgW+jkdBLwZe\nggv5HrnfDfB78hZcELdxGvDturaZ9x3gG7hg7S1Qk/icqvgEHmdyHHA8/p2cIdyBr+ALb2+Mk/h9\nqqJt7stqnm/iGOAqYGfgJ8CJFEGNHcb5+uLvFsBaxfUeC9wMPK1j3/sUf98LfBX/bTwReH5fuw2B\nlwOb4gs6+HfpqIp+U9um9g35c5/O5OTkvPo3MTFxYEbbPRPbXdj3+Lzi74+6tk0d58TExAV9j2f0\n2WE+Se2Ktt/qe3xOTbuvTkxMLEq85rl9f7/X9/raExMTm09MTBw7MTHxkOLfgycmJtYewj3/QfH3\nu8Xfbw/hM/pGRtvKz6/rfZ+YmPh+Rt/f6fvcv5vS9yBzn5iY+MPExMQHq/41vOe7feM8fwjj/J+J\niYklxf/XmJiYOGuQvsufZ+nxuTXtHpcxzqS2qX3nzL3/37zZuZe4VNIhZvZeSWcBHzOzb9W0PUfS\nO6g5Ipf4maRT8NQI2wG/kLQrU7vbLm1Tx/lrSScBF+G7zZ/Vzjx9PqntwHfFW5jZbyUJT+xWxVrA\nJZJ6qSImzeyVNW1vk/RsYI0ivcRt5RfN7Hbg/0na3Mz+UHONMjn3fK1il/9rSZsAS2va5XxG10o6\nGvg5xZHZzI5tGOvL+9r+pqLdZaX7vh319/0HxWd5WWmcf6xp+xdJrwPWK8bw15p2l0vazMyuqnm9\nTNvcb8VPQDksKe4NkpbiO9hBx/nA8vXxHf8gfVO0eQ1Tv807apptLOkbTP8uVartctom9g3pc5/G\nfBTu7wOeXvx/V/zoWfdDbzwi9zCzN0p6AZ6i+CQz+0Yh6M4coG3qOPfG1QYTwKlm9rWGoSbNJ6Md\nwJuB04s8+38CXlfT7j8SrtVjb+CjwCbA25k6PvazUtILceFwD9QKwveRfs8/UrR5G/Am4NCadjmf\n0e+Lvw8o/jYZoh5T/OsxCVT9eF/L1H3/YsN93xT4OFOCehL/sVfxGlzldD2eRnvPmnY7AH+UdF3v\nmnX2E9rnfo2ZnVDz3jrejasbHghciH8HBx3n8cCvJF2K2yg+PGDfAP9atH8Z8GugbjOTo7ZLbZva\nN6TPfRrzUbjfaWarAMxslaS7G9reZGYHtV2wWMHXAa4GNpG0u5mdOGDb1HGuh+ubrwTu29R36nwy\n2gFsaGaP7j2Q9DLgpxXtLgGejevFFwGbAd+vueZbzezlCX3fH/+i96gThMn33Mz+u/iSPxrX/V9Z\n0zT5MzKzQyQ9kOlzr2v79PLjImFe+XHZoNv7jNaQdG7NDu7hZrZVyjhxQXVwz3NL0ofxYjj9Y5xI\nvF7K3JtOmnU82MwkaRnuaVZnQEweJ/A1fMHeAvg/YONB+i7Yy8z26z2Q9CFKNqESfzSzbyeOM7Vt\nat+QPvdpzEfhfpGkk3G1yBPx42IdqUfkr+FGlrLhpo7UtqnjzOk7dT6t7SQ9H9geeIWk3k5wMfBC\n4EsV1zwdVw08ClezNBlWt5Z0PzOrUwtQjKlREJZIvueS9sV3xBsBnwO2BPataJr6WSLpeFx1sh5w\nH+B3lAzEfW1fh3uX9IThnfjuvEfZoNtLljdJvUH34kK1VR5n3RH9TcBTJL3UzG7BP6uqMT4ZN7ze\nK7DN7Nk1bRvnbmZv793rQh22fjHOUxvcg/cGvmBm19W8njzOGi+UbfHd6zSPldS+Uw3+JXLUdo1t\nc/ruMPdpzDvhbmb7SXoR/oP5kpnNUJ2USD0iLzaz3RKHkNQ2Y5w5fafOJ6XdL/EV/m9M6UzvwXeU\nVSwys30kLce/fE3eJVsDN0i6vrhm5XE6QRAC2ff85Xhq6e+Y2Sck/aSmXepnCV7r9xG4p8WBwJcb\n+n8j7qlwEL6bKp9MMLPjgOMk7Wlmyxuu02NH4Hl943xoTdtLgf8Czi4Ebd1G4dO4+uql+ImsblGF\nlrkX9+UgXKi8B1eZPRZfvP6z5pprS/o5/r2bpN5+kzLOHI+V1L4/D3ynmO9hpWteW3PNftVVE21t\nc/rOnfs05p1wl3R/XP8q4P6Sziv7MffxT8BWZvbz4kv4PzXtLpb0JOAXtO+OktpmjDO57/6dbgOt\n8zazK4ATCqMe+K59O1y/V8VdktbBd3CTNHw3zOwhieNsFIQ9Mu/54mJ8PcF2e0271O8GwA1mNilp\nvcIFtmlOV5nZ1ZKWmtn3JL23pt1PJW2H/xg/CHzQzL7T36inMis+gxvMrEkNSWEDuhkXsnUR5teb\n2SmSnmVm75NUp16D9rnvh6vrAFaa2QGS7gucQ71wf2fTHHLGaWY/BH4o6XFm9r/gacjNrM5Q2tp3\nyeD/VlyA3onv+E8EqhwAPps4n9a2OX13mPs05mP6gS/i6oF34kfEkxrafp6p3dkE7m9dxU74jvVy\nfEVvyi2f2jZ1nDsBpwDXAb9p6lvS7yX9rvSvTj2ROm+Aj+HGvffjBpzjatodCbwVN2RewdQOpGqc\nj5L0E0lXS/q5pMfWNL3KzK4GlprZ96j31Mm55yfjvtAPK7wSvlrTLucz+pmktwNXyf3o79PQdlWx\nWEwWJ5NNatodjS88B+Gfe+UiIOlpkn4HnA38Vh5UVMf34d4Alr3wE1EV90h6BLCuXFpv1HDNtrkv\nNrMb+vpfRbPa7ue43/Y78DiLumI9OePcStLLJb0KuLoY8yB9g59SHo8vUnfiNpwqvojLhC/hvuYn\nN1wztW1q35A+92nMu507gJkdXfz3l3IDYB1/Z2afLd7zEXn0V9X1tsnoO6dtyjjfhAvOq/Dd6/9r\nuOTDi7+L8Bv/LzXtkuZd8AQze4uk75rZ0yXN2D0W1/kKgKSNgNPM7MaGa34SNwj9UtJj8PltX9Eu\nVRDm3PNz8GPtI/1tdnFNu+TPyMwOlLQ+bmv4J9w9rY69cMPWAbjHzn417W7Dg7HWMrMLVW8k/gCw\ng5ldJY/8/e9ijveiqZD2kyX11Fq3UB/stD+uavkkLlxq1UMJc79Pqe37Ss+vUXfNor/vA1/ANzef\nA14wyDhxY/I/4YLz7/FNyEcH6Bs8gvgM4M1mtrukf6xqZGbb9f4v6X40COKMtkl9F6TOfRrzUbhf\nLulfge/iAu6G3he6wiA2KWnCzH4jaQv6vnAqwtol/Yg+/aSZPaVr28xxHorrVb+M69jOx12bZlAc\n2XqcL7egV9E47z7WkPR4/Ci4FjV+4fJQ56OKa50m6Q9mVjlOXD//y2LMv5B0V027vYCHMSUIqwyf\nkHfPjzezHSj5hdeQ/BlpZjqFzamOgQDfsW6L/8jOxPXglf3jR+1vFItVnfHxbiv8vM3sSkm3VbTZ\nv/h3THHdjYC78SL0M+wIZvYrSXfgxuYX4S6wlSTM/UeS9jWzT5Xesw9u/K5jYzP7r+L/v5D00qpG\nOeNkKpbiJjO7XVKd7Erqu2AtXHD+TNLWuEqyjVXU20Ry2ub0nTr3acxH4f7w4t9eped6X+r+L/Jb\ngC/KfbivYqYPd88HOsVtr67t2v0NM8d5j5n9RRJmdpukm+oGUAjz3sKyGfUBGG3zLnMiLrT3xI1X\nx9S0+wC+CH0F1xHXLkLA3XJvnB8W76nTe3/IpnLGvE3SiVS7V+bc81vkhdjLvvNVu6Ocz6gxnUIf\nqaH1u+LeLN/E7Q4vB5D0EJse2HWjpP2KfncE/lJxrc8XKron4SqHo3G/+EOqBqh0jyJon/u7geWS\n9sRVZv9Q/K3N/QPcR9IDzOya4vOvXFgzx/lb3G/9rXI7R92JLanvgrfhi8phwG7U+MSXNnyL8HQM\n51S1y2yb1HdB6tynMe+Ee51RUdL7KtpehFvu+9u+18wOMbPeDmRtXLc1gR+V31ZxrV7bXc3sI8V1\nHon/eB83wDhXFEJ7Y0nvotpg06Osj/8lcFZVo7Z597U9iinr+lsa2vYWocm2RQhfKD6Ku2T9Gtfp\nl8dRlTNmETXG3Jx7DlxQ/N20YXxZnxFwHzM7V57ozGp2zz22MLO9JO1gZmcW97Sq/+vxvDHgJ5Ie\nn2X6grUb/lkdhn8+VYFJ/wm8yszukPQB4DnACnzhOKOifapHEbTM3TzX0MsLQbk58Cczq4st6HEw\ncIGkVcAGuMGwiuRxmtkektY3s5sl/dTMrgGQ9EKbHiCW2jdmdgFT36fyyeR0MyvnKypv+G4ryYq6\nObW2zeg7Z+7TmI8G1TpqM6RVsFPf4xPx3c6T8R/X5xre+0hJ+xRGi5Op16nW0T/OfXCBfh6uJ33t\njHdM8QXcj/iJuCvV3zL77p93TtucRehm4FgzeyS+21tVftHMjjRPoHUYvsN9Oi6IZiyqLcy454Vg\n/in+2fyiQlC3UfUZNaZT6CMrvL2CRX2PP21m/25mzzezd1i1l9AaZnaxPAHZemb2v4VNpM4VMtWj\nCBLnbmZ/NrMf9wR7cXKrxMzOwU8rOwOvsAovoQ7jxMxuLv5eU3r6zX1tUvtu4n59j+/GNzPfAI6W\ntHnDe3PapvQNpM29n9VJuPf/KHLa3mJm3zSzVWb2PzT/IF+NC4Dn4MbI8/OGOb1vM7vLzI42szea\n2X9ZvQsm+JH/ofhRbnM8I2XnvjPblhehm5muIunnVKbUVX9hZmbNHs/Ej6aH4UagusyIqWPsqa72\nwHNxvEpSq2Gp7Zr47m4P2tMpgO+yz8f17hfiXkg59AvktSU9WtI6ktZSdaBXT1//HDylApLWxDcC\nVaR6FEHe3MtsWfeCPIjnZcXp5V8lfWII46xj2v3M6LuJ/nt0HO7BtT2utqpTV+a2Tem7icbf+7xT\nyzSQM+n+tldIOgjP7/x44HZJzwKwIkFVnyF1TTy447uFrrwu18eg4+xnSzPr7Va/KumCxtaD9d3f\ndj1cl9zT+b6Y6khW8N3j1wHM7GRJdQtBL9XuQWZ2qqSmU0vKGAF2NLPtAYof7oWDXtPM/iRpb0rJ\nnuows+9711pmLRGYiUzgUczl8fUb4b4t6XzgwcALCgPxp3C3u6oxfkruFfVI4HIzq3UHzJk7gDys\n/wYza1qoH2dm+xTXf7OkynTDOeNsoP9+JvWdyTpm1lN/fVXuoz6MtoPS+HtfnYT7IEzi7msPxVe7\nP+MRX5NMJahKMbqOmnUkrWtmt0q6D83GoGHzLVznW05gVSfc75D7Y1+Iq5DqTkJr4kbcH0h6Os2R\nkqmsqalAjkUMtpgCXhwE32mtKl1zhp2laPs63Di7joqAHzPbOqO7/pPdo9reYGb/IekMYJW5y+QW\nuFrs9JoxPhH/Pq8DPL3YoLyhpm3S3Iv7dzxwI3A/Sa8tVCCVSNrYzG6QuwNWypmcceaQ0ncmSyQ9\nyswukdR2v3LajpTVSbh3VjkUBonH4hGQv6raIfQ8GFRRaYe8o3fOOPv5BO7nfSke4l8X/TiMvvvb\nrjKzPRJk80qhAAAgAElEQVTfuxeuV/wkviDUeaLsges+j8dz2rwqY3xVYwTfrZ4v6ULce6Ry95p5\nTZnZFonvfzPuHVMXQTv9wtJGZlb2gDm37/W9cEP3ur3nzGyG+5yZlVMC/xb3oKjjBNy9MWWMqXM/\nlBZ//BLvxyN0V+KBa3UCO2ecdfTfz9S+m+gfz5twj6EH4qfbWiNtZtuUvptYPdQymlm79F7M81vv\n3qVt0f5Q3EPhIuBNhUW6LnQ6KVVssbN+HcWCARxjnkipyUWsETP7gqQf4j7H11pNXu+cvlVfdqy/\n7dly/+V7PVqsppSXma0AXlR4E91eCJuqdv+HZ7GD+lNA1nzM7HBJZ+Ouk8vrjvKZ9+ciKbnW5cXA\nFdaSJkDSTnhw1xqSTgP+YGbHm1l/iuLX47llrum/xgD8n5l9LrFt6txT/PEpXv+6pG8Cy/oMgIOM\nE0lb4rr+i4ErzTM+fqxj31U+/hcXRuN/7mv6a2Bvm0plUVspLLVtRt/Jc+9n3gh3fAfW8w+l+P/D\n8NV3bfNcKV3agkd3PdHM7imE3Y+oz4uRmir2FNzX+iz8WPtZYLeKvpOR+7CubR41eFrh9lSVZz2n\n70q/7Iq2T8WNpD1vkhklzwpVzPG4imtPPMT7OkmfMbNc42+n+RR6+wkz+3dJ35J0kplVpSvI+YxW\nAT+R52xprB+LL5C/k/TbUtuqhGS94LW2uIHrLa2gSQ5fkacSKC/UdafP1Lm3+uNL+gdc4PwLnsfo\ntOK6u5lZlW0keZya7hN/Av5739eKJHMd+ob0+IYv4LmJfo7bSF5Gff711LbJsRVtc69j3gh3mx62\nuxZ+vFofF8yd2xb8CY/MXIXrgZv8VFNTxW5sZr0kRV8rdtyD8gIze3zR578URrQq4Z7Td5JfNrC+\nmTWFQINnBXyimd1ZXGdn/ITzPfI9e8rkzOf1TKW6fR7+w6gS7jnXfAawkZnVRdqWeR3+g21Md0xL\n3ICkDxb/Xas4ifwvU9+3AxPG0cQb8UWlbYyQPvcUf/xP4SekuyR9DPi3ou0XqK75mTPOsk/8xzXT\nJz63b0iPb8hJ95HaNie2om3ulcwb4d5D0ja4H/q3cVfEWt/XjLabAb+R9Etcl32HCk8Um+kJk5oq\n9leStjez8wvDyR/k7mmLrNndsYl7JK1lHqyyJvWuqjl9p/plpyxqd5pH/j20+P+K4rqNKooEcuZz\nd08QFYtMnUE155q/wYOi2oJzwDcKP7H2zHxtcQM9NciNuM/+X/Ed/uEJY2jjhpoTXxWNc9dUXptN\ncTe/ntF1GTP1w+ub2RmSNsaLZpxTXKPue5wzzjaf+Ny+IT2+ISfdR2rbnNiKrHiAHvNGuBc34UD8\nCLO3mZ03jLYFdQm4ZmDpBSaeCjxb0p1MZef7Dc35uNs4Gheyl+A65bovfk7fqWXHtin+9aha1Cbl\neS2eh2cxRJ50al0GI2c+vV34Rbhba1WEZu41t8dz71xfPG5Sy6zNlNG7twhWHbv3wQ3PveC1ae6i\nVpSuK3ZhLzevcftDfLPSqEtN4HpJxzD9NFCXwKpt7v15bWBKwPd/P3oC6pkUhmNJi6jPBpozzlPw\nU9pDVO0Tn9s3pJeLzEllkdo2tW9on3sliyYnB/YkGwqSfowb/f4TmHaE7b/hOW2L9g/DBXy54kvl\nh66KAhOWVw5sIOR+xA8Ffmtu+KwNMy52KX+xhlJikv7V3FCbUnas6v33hutL2h3XEa6J/7DXxQOY\n/svqk4zl9NU6n6LdY3BD6WVWnxVyYKo+d7mhdBpm9n315YyRVzja34rygZI+Z2avrujjgvLpUdL3\nzOxpA457hpeVZUby1sx9Ezy4boVVVOGSdDi+idgWj8T+Da4yvdnMZvh7545T0lYU2UDxcnZ/Lb2W\n1XfvPWaWGzU9bfypn2t/29y+m+Zex7zZuTOVh2M96iPvurQFj4Q7HS/Ie1XLe1ILTHyXmdkj6yr9\nJGMeGNMfHPNmSoEuysvgmFTyrIF7hZmZnSjpq7iHzO1yd689zKypFGIrOfOR9Gg8De0fgCMkVRbB\nGNL9mfa5F9eoK3zRnzPmWcD/yEviXYNvRqr4Q6F/75UYTFENNVIncFSRt6SB/u/cHni+/cuAhxfC\nqt8D6u14FO3HzOynhTrs17iL70DjlBuwX2JmpxWPz2X6553Vd0FSucgGBkn3kdx3wtwrmU/pB55S\n3OxF5km/7v03YFvw1ftDeNKjV9OcdCq1wMQ++FHqDbgXRJdCwqn0+7P2Mjheg+tpm3x515YX1DhV\n0inyWqWd+pZ7Gt0GnFLor28ALiu+bIOQM59eEYx301AEg+Hcn0HiBlbgAvJMScJzjlSxB15i7bn4\nol5lqBwWlXlLauifz+uBxxRCd1uqk+9Nmtk3cSGLmV1iHsl6WH/bDuO8FThSHkw1Y3wd++6Vi/yz\nvPDMVZnjHOT7kdN349zrmE87903k/sBPLX4M91Kh08xpC64rfgCwVFLbbj+pwITZNL/gy+WFb0dF\nv5oiJ4NjasmzlL7LxZ8N/5L1UgwMQs58kopgDOn+DJLOATP7WaHKqq3uZGa3AR/vMLYuDDKfG5jK\ncdMzAE9D+YWnc8Z5HfDPeEj/JsA0754ufVt6uciccSa1zey7ce51zCfh/kzg0bgPZ13O8S5twfNe\nvxh3mWsr49ZfYKIyK6Q8H0ePB5KmHhoWORkcf47rybfG9ZD9QTTJWH7x51Ry5pNUBGOO7w8UkbNm\ndlmxWThylvsfCpJOYco75mfyyODHUZ2xNLfwdA6LzGylpOfhhW8ePWjfhepmOfAg/NS456AqxlQy\n+26beyXzRrgXuqcfyPNN7IgHylyIC6TObYv2P2AqIKfSu0JFIrESy3CPkDpvmQeW/n8b7vs8KvqP\nYW/Ad9EpaYRzyo6l9A1wjqR3UErRYPVBMinkzCe1CMYw7k/2sVvStmb2UzzQqfyd+mSH/ueS3tyP\nrnjtlKo32FTx531w1U3v+/EP9AXEdeCAoo+bJb0AD/MftO/UcpF1lFWWE1YdDzOjbYe+G+dex7wR\n7iUOwVezrXDd6gF4rpfObSUdBryGko93hatb733lyNfe415ysXLqg88mzSaTYkWfwHPg9Ip39LvG\nfd3M+hejOpLKjik9TQEkpmjIIHk+1lIEo+v9UWZ4t+pzxjwTzzff/z2c9j2aIyrzljTNvWdAlrQL\nsK2ZvVfSWTSnb/4yHlbf+37MiHZOHaekvcyjn19YCLY2cvpOKhepogRn6fGJZrY7038bJwFPajBa\n9/+OWvvuMPdpzEfhvoOZ7Sgv6HyCpCb/z9S2zwMeYg0BUbh75LMT3Jt6iao2xqNeL8GL/P6ZmkyC\nqcjTEveKFL9N0pfM7OM2M8x4ZXGzf8NUqbm6XUNq2bHUNAWQnqIhlZz51NFbkLPvjzLCu9WSM8aK\noBwrkrBJegRwh3menVlBGXlLMuZ+CF50Bfz09E2KWIcKHmAJabITx9n7/l3e//5B+i5oLBepmRXF\nFhX/fgXQ99v4naRrgfsWxtF7U6OY2WYVv6OUUpW5c5/GfPKW6bFE0jq4QXMN6r0Mctr+nPZ81T0j\n7eslnVz+V25kZtuZpz/4FZ7j5Fn4TrupuG8qzwOeau6XuyP1aYjvD7wV+DQulKuOzj16Zcd+jpf1\nOrim3RZm9h7gb8UPuyn441JJL5czoaKY9QDkzKeOXhBMl/vzcnxR+6uZfRzPNllHL2dMpWePpJ0l\n/VHSmoVB/gzgpBEb3Ps5FlfHrYnvWpvcAVPnfqcVPvvF36bf5eXyqlEDj9PMzpa0jXnQ1ym47WRN\n6u1mqX2DqwJfhQf5/Rt96kCbqih2SCGgH2hmDzCzZ1aM8xVmdn/gM6W2D6zQECT1XVwzd+7TmI87\n94/jbmvLgB/TfPxLbXspcLWka5hKjtQfpZhrpH2Qmd0EYGa3yH2+B+XPeGDQzbiuv9I33Wpqjta0\nPUfS03B3qoea1xatIqd8XGqKhtQxJs8ng5z7kxPe3ebZ05R/Z+BAr0Ry8pakzv2iYqPT88dvMjzu\nAPxRHvU6SX3Eb+s4Je0P7Cppezxo8SEUMQ5UR1un9g1T5SLPKU4w08pFlozy1/cZ6JsiaQ+QZ6H9\nO+Dr+GlkRW7fRf+5c5/GfBTu++KGhS2B3xc61kHb7oobVmoDBspGWvPUsNOo0KV9S9L3cf3qE+lW\nIqx37V4VqPsD/6epHDg31LT/AG5DuNe9qu4LLC87tsLMPirpYHnEatUXIzVNAWb2dEn3xaMVf2tF\nfceu5MyngX6DVc79yQnvbvPsGVX+nRxy8pYkzd3M9pN7/Qg4zaaqDVW1TT3JpYzzX4Cn4N+NV+LV\nyv6qmiplGX2Du6j2Tgu9cpHl2rB1G4ImF8jjmTL2X1M8rgp2ausbMufez3wU7pO4MczwRFpNWfJS\n2/4Br6PamnCnSrAX3K+v3bslPR4/8p9gg4XB51aBej6wecp8SC879mDfPLWnKZD0z7gucgnwpWIX\n+4HMOZTJmU9vDI1FMHLuj5n9l6RvUwrvbui6MWcMo8u/k0Ny3pLUuUvaCPfVvxrYUNIB5oGBVW0f\ngavWNsSF1qVWlGXsMM6bzOxuSY8DfmdTEZ2VnkwZfUNLuUibni7ggZTSl9RcD9yBYbmk3czsAtUn\nLkspVZk1937mo3DP8Z9Obftg4LeSflc8nswwuvSYJuwklfveWZ6g6grgSKuuYF+LTVWBKpcd61EV\nrdmzISQJQ6WVHctJU7A/nnv6LDy69KfF364kz6fNoFlql3x/lBfefQyeM+boou3ncC+jHifgIfpr\n4t47j6TIv9M2tyHyVjNL2jBkzP10fF6PwnfYtzZc9pN45O1xTO1kqwRsyjgnC5vOHhRuzHLvnrpA\nntS+IbFcpDxP0HZ4upP74LEylbnXi/YPL/4+qGGcKX3nzn0a8064F8aDYbfdteNwmrgPXursh/iN\nfgIeMHECeX7kZVLLjqXYEHqklh1buzC6GlO6yrqCBHeb55aZNLNJSbe0jLeNnPmkFsHIuT+98O6D\nzey7NO+MGnPG2Ijy72SSkzMlde6LzGyfYtHci5aoZDNbUXw/rquwS+SM8yDcgHgNrs/eCV8sazO9\nJvYN6eUit8E9ro7Bg6S+3HDNN+HahK2KdnW/t5S+s+deZt4J92GiKT/RfZipJxu0IMIyM+v5Mp8t\n6VtmdnCD2iOF1LJjrTaEHpZediwnTcF58sjFBxU6/aTiAQ0kz4f0VAU59ycnvHsFrkI4U9Ju9HmN\naCr/zqmSdsXtJislnWtDSCyXSC9vyfX4jrDJqJg697vknmnr4b+lJtnxF7mn0HryGgF197V1nGb2\nE9x//EHmQTwX4llT6xb/1L6Ty0XieecnJa1nZtdresaTfp5jpWJCg/TdYe7TGGvhTr2faJc8x/27\n6Q0kPdzMLi+OYUvlKWsHCXNPLTvWakNQftmx5DQF5mUAn4Pn4b6sQaeZSrJNhPRUBTn3Jyu825pz\nxlTl37kb19HPCpaXtyR17kfii9oluFtp0879NfhncD0eLVrpBpoyzkL4bQZ8RB4VDe7h82Gme2wl\n9638cpE/k/R24Kri91mZJ6jguZKOsJoauzl9d5j7NMZauJtZL8jiCdYXYYYH6cxA0oPx6MJpofU2\nMwBkX+ALxbH7CjxV8K7kZ8Ark1p2LMWGkFt2LDlNQbFwTOCCa2tJW5vZR1rGPOh8epQNmjcz06DZ\nI+f+5IR3N+aMsan8O28ws6MarjMylJe3JHXuk/hnvRK3jTRFAN+Cp9nu/YYehgfmdRnnhvjvcVPc\nHtVLVlf32ab0nVUustjMLMXz6fSCDOvYBF8Efs+UerP8Xc7pO3fu0xhr4a6ZEWbgH9Cv69+VFlpv\nZhdJegbT3QF/OuCQU8uOpdgQcsuOJaUpKPga8N+02wZSybGJrIdH0vY8ZV4M9OcVT7o/ygjvVn7O\nmFeS+CMcAa15S3LmXnAwLpSulUc6n0l9hOo38DiNlUxVbXpJRbvWcZrZD4EfSnpPzSm2S9/Z7qpW\nxEzg6rjnU2+k3aVlfMl9d5j7NMZauJvZkbix6EAz+2DrG5yk0HoN3x0Q0suO3UVf2DYz1RO5ZcdS\n0xQAXGFm72ufTjIp8+nxLXxx7p1uJqkQ7on3Jye8OzdnzC2SjqBw04XGwJdhk5IzJTe0/QYzu7a4\n5p8l3djQdh0zSylkkZTbpeCZuHNAGyl9D+quumXdCzY9eR3FQlB+rkvfqXOfxlgLd3WLMEspFA3D\ndwcEN9aB62uh3jZwLF5I+WA8AOUEZrpmXSqPKNwWeG2hnng/ff7gJXppClYBG+CukXWcKenDTLcN\nVKq5EkmZT49VVuRtaaH1/lgpvFteK/e1uMphhout5eeM6QWaNBWGGRWteUty5l5wk6SzcdXd44F1\n5RWkqmJLfiAPTrqs1F+V/3xKfpUea2vKm6u3WFZ5c6X0Xeeu2pi5Ux4DcoN5EZBU+heCLn2nzn0a\nYy3cqY8wayI1tH7Y7oCY2SFKC5ZICS+vKztW+SWy9DQF4Pq/y3B3L+hmoC6TEy5/tjyta3lhqfKA\nab0/ygjvbjCEHWcVJQGLe/mPuGdDbTrqEbEn7mbXW4Bn5C3JmXtBOXK1rRTgpnhqkPLpqsqG0jrO\nEqneXK19W6a7qrwC0vHAjcD9JL22p+aso24hyO27oFPBnbEW7jZV2Pnv29qW3vP0Qk+9BR4VVpfS\n4LxiZzwsd8CcYInWsG3zCNNvFte9P5674it4zosZuyilpykA/2I2ZevMJSdc/qnA2kyFdNeldE25\nPznh3Vk5Y4pdbWrq6mHTmreE/LD+5PgT4OFmtlV7s6Rx9vhfXMhtRpGzZZC+zexGSUslvZDC+Crp\nUTUn0EPxDLRXyTNZ/jdQKdxTFoLMviF97tMYa+Fe4ov4l3gx7k/9f3iCoRlI+hf8+H4Z8EhJ7zOz\nz1c0PQp4UdFuD9xXeFBSgyWSw8slHYXX5+ylIa3bRaWmKQAv6nwA020Dg+QqT54Pbij+x4Rrptyf\nnPDuXCNcTurqYZOSt2Sg0PYWLi4W6bJq846O4+yxHN+s7ERzzpbUvsEdA65ieu73Ku42s6uKa13Z\ncrJMXQhS+4b0uU9jQQh3KwUVyEPwmwxb+wOPN3cNW4rrqKuE+xeA9+EudgfiPuWDZjdMCpYwsz+R\nno/mibiapSnLI5CcpgBcbTRR/IMBC1FkzifVJpJyf3LCu3MNYTmpq4dNSt6SgULbW9gR/5zu7Yvq\nwJuUcfZIzdmS2jfAYjPbraHPHjdK2g8/Ie7IlKdWFakLQWrfkD73aSwI4d7HKpojvO6xIsuhmd3U\ncHPuwW/2u83sVElN+sJUGoMlJF2Nf1nXxoXKFbia5Toz27zmmivwo19TLhBIT1Nwr1FxUDrOZ5vi\nX486m0jK/ckJ787NGXME6amrh01K3pKBQtubMLOkGp+J47wXJeRsyegbfJf/JOAXNO/yd8M/r8Nw\n28CeDddMXQhS+waS89VMY0EId02l1F2E/9i+3dD8d5IOZ+rm1IUkrwl8BLfOP536WqvJWEuwhHnh\nACR9HjjAzK6QFyZoEhx/j6tRep44lQFClp6mgEIl8058wejlgslN0dtpPtaX+13u6VFF6/2xjPDu\nXEOYmZ0mz7b4MFztUZm+eUS05i3JmXsuhcB+K9MDAasW4NTcLpCYsyWjb3DVRtkvfdouv/hs/oQb\naY9jSq25jPoYj9SFoLHvPlLz1UxjQQh3ph/5bzOzPze03QP/ku2M35x3NbTreVC8EK+qMjCWFizx\nUCvKdhW6vSaDcaMRT/lpCsA/z83MrO00kEryfOR5Q/ZnyqPoTqbUQ2Va748ywruVmTNG7imzBI8X\nOFWemOvk/najwBLyluTMvQNHAG+hPRAwNbcLZnYp/v1E0oOtuvxjct/FNbdpabJ/8e8YpnTiPQE/\n7Z7nLgQJfZfbps59GgtFuPfCd3uWacysNjsibuz4Cu6atRkVATXm/s09H+cZQTRDoi5Y4teSTsJ3\n9k/Bj//TkOdfqTPSlP2Sc9MUAPweP10Mi9b5lHhjMa6D8ICnt1Q1Srw/OeHduTljDsO9UHpRl1/C\nw+JHhvJypgwU2t7CH82s9nScOc7ee/4dd2+8H7CHpLPMbP/cvotr9U7yMyifanvXN/eg2wSPdl5h\n1VkskxaC1L77xps692ksFOGemkoX/NjzaeCluJA7Fnj26IY2E7UHS+yNh91PAKea2dcq2qRGHuam\nKQBXcVwi6ZLicVN64BRS5tPjKjO7WtJSM/uepPd27dQywrstP2fMrXjZxLsKL5tBYwFSSM5bkjP3\nDlwrdz8tG73LTgxZuV0K/hlXk55lZlvL88536Rsyi+NI2gNXQ14GPFzSe81s2oYhYyHILcwD6XOf\nxkIR7qmpdMENe2cCbzGz3Yvj9ayg9GCJ9fDj/pV4tfXdrc9H1gq/5MLD4wnUB0blpikAXyiHSet8\nSqySJ+yaLFQ0mwyh/5zw7tScMTfi0bHHynMcXdtxbDl0KfHXKbS9hd8Xf+sirbuM8+7iej2Vap2X\nUlvf5eI4D8PVkeXfRpXe//XAY8xTTK+H/04qT4NtC0GHviF97tNYKMI9NZUu+K70zbjnyta44Jkt\nRuEjezr+Bfo7XIBehdfN7JGbpgBm+tjeKc+m+UWrL1PYRM589sKNlAcAbwP269BfP2srPbw7NWfM\ny4AtzOzXhU75uCGMs40ueUty5p7Kd/se31nSSXcd5/eKf7sVn///dOy7zMn472MH/PtXl677Bty2\nA66ObMramroQpPYN6XOfxkIR7qmpdMGDaF6I60x3I6HK+BAZhY/sJma2naTP4IKwf7HISlNQsA3+\nJe9VOXowXlvz2bjOPpec+dyDF5i4G3dr/X1z8yRywrtTc8bcH3i+pmfXHPYOuZ8ueUs6hba38AF8\np/kz4LHAHcA68lQN/9llnGb2bryIO5J+0rCJaOu7zM1m9iFJW5rZnpKm5aiXF6TpGUV/VngUPY5m\ne1PqQtDYd5mMuU9joQj31FS6mNn58rziG+DqmWwXvwEYhY9sz6NlPTP7W7/u1zLTFBTcz6by2x8j\nr3L0b5K6FqTImc+X8QLI/8zwbCLJ4d2WnjMmKXX0MMl11yzoFNrewq3Ao4vd69r49+kl+Pf6P7uM\nU30ujoVTRJWLY2PffW0nJT0AL+SyHjN3z0dXXP+Uiue6LARtfZevnTr3aSwU4Z6aSrc/v8u6uJ97\nbTHcITMKH9n/lnQw8MviC3dzVSOlpykAtwdsYh5FuzGuJ1+T9JSp/eTMZ108ovLNQ7SJJId3Kz1n\nTFLq6GFj+XlLOoW2t7DMzG4rxnN78V25o2yk7zDOVBfH1r5LHIKnqDgJz+N0UvlFM/t+Ma5dgG3N\n7L2SzqI6DiN5IUjpu49k984yC0W496fSbSKnGO5Q0Gh9ZI+UtMg8rcH/UHwWkl7Y55WSnKYAeC/w\nY3lO7/Vxdc/bqC5UnTLG5PkwGptITnh3as6Y1DQJoyDHhtEptL2FrxanuItwY/4Zxed06QDjbHVx\nzOy7l020l0PpjN7zhRH0kFLTQ5hKXbErvhhOK1SSuRDk9A3pc5/GghDuFR8WAJJON7MX9z2dUwx3\nWCQHSwDIK+e8kSlL+8bWEHZdqF4ws0tKT78Z/3H1SE1T0Itm/Qa++FxbXP+stvfVkTmft+E7nqHa\nRJQe3p2aMyY1dfQoyLFhdAptb8LMDpX0Nfx0s9zMLpW79/bvbnPGmeLimNN3EzMcBsxsVXH9VWr2\n6mldCDL7hsS597MghHsD96t4LqcY7lDI8JHt8QHcbWof3Dtg5w7d9mf/a01TIOlTZrav+gIxCh1g\nnQonheT5mNkFTBk1P1UaQ9VCnUpOeHdSzhjrS5Mwy+TYMDqFtlehonSfpgfQbVN8P/oLeuSOs9HF\nsUPfTfT/Ni6Se5T9CD/hNtkwchaClL4hwb2zioUu3Kt8YA8s3LJuo70Y7lBRQrBEwdVm9iNJ+5jZ\n5yS9ukN3/XNPyTV+aPG3SyBGE8OYT9VCnYRlhHdbS84YSV82s5dqKila+b2zZZxPtmHkzD2B3NJ9\nreMsqSyb9Ndd+m6i/77tJ4+tEHCamZ1R/TYgbyFo7Dtj7pUsdOE+A7l/ebme5+ZMBQ+MmlQf2dsl\n7QisKS9ysXHXDpWepgDgzQ1qqtzdUZlhzKdzBKgywrvVkjPGzF5a/K2sAlZh6xg6OTaMnLknMCkv\nHn51SuPEcaaqLLP6zkHSRvgJ/mpgQ0kHmNmHqtpmLgRtZKlr+wnhPpOcep7DJtVH9iL83n0A3013\nuY+941/OTsc69JPCMOYzCDnh3YPmjOm3dQyNfnVZmQa1WafQ9hrqTn/T8v3njLOssqxqXzJAJvVd\nvGeixbjdrxo5HT9NPwo/0dfapXIWgra+M+ZeyUIX7lW5ZnLqeQ4FJfrISnoNHqG5FVPRtmv0t6u4\n/qPwvC2/MrOeMP8YZKUpyGqbQtf5jICc8O5Bc8YMo9JRHV3UZZ1C26uwmjz/cj/2MsNU6+2U2Te4\n2+GTGuw0u/c9XmRm+0hajn9fawOOSFwIevar0uMTzWz3ir6baHRZXRDCXR4a/wqm53h+v00F4pTJ\nqec5LFJ9ZD8PfAdXgRxWPHcPDXlLJB3ElO3gbZK+ZGYfN7Mz+5q2pSno2raJ7Pk0kJIUro7vkR7e\nPWjOmJElELNueUu+R4fQ9iYkvR9XMa6FLxa/wd2LBxlnHdMWy7a+C34n6Vo8PuOq0jUmzWyzCrvD\nXYWH1Hr4/WuSm40LQfGdOQjYSNJLir4XAb8CyLR5NG4UFoRwJy9aMKee51BI9ZE1s9uB/1eMMZXn\nAdub2T3Fjvs8vDp8P21pCrq2raXLfDIX6tRx5IR3z0XOmFyS85Zkzj2VF+CBXkfgJ8S6RGs5+VXq\n6F8sW/s2s1cASDrSzN6Y0MeRuCy4BPgTzTv3xoXAzI4EjpR0oJl9MKHvJho3CgtFuOdEC77VzIbt\nDaLDaHkAAB7BSURBVJLKoD6yVfwZ38HcjO9mrqtp15imYIC2w2boYf3KC+8eNGfMKNUyPZLzlmTO\nPZWrzaNDl5rZCtVXy0oe5wj6Bi8veCh+Av06cLEVGSr7mMR34SvxqOTPNlyzcSGQ1NvIXF/6P5Dm\nu57DQhHuOdGCW0u6X4OP+SgZ1Ef2XkpGq/sD/yfpl8DWuNG2iqQ0BR3aDptRhPXnhHcnLy5Nto4R\nk5y3hI6h7S38SdKeeAbND1Hvpto6zg7Gz9S+wSOqv4kXf2lKvXAwnn/+Wkmb4jmn6jZdbQtBpRcV\n3dR1oZYhL1pwa3xVvb5o16k+aEcG9ZEtk3X6sPQ0BVltR8AowvpzwruTFpcMW8coyMlb0im0vYXX\n4ZlCTwNejXsXIekhPX17xjhzjZ+pfUN66oUbzOxaADP7szztRh2NC0HZu6Uw9rY6JXQ1vi6anJzN\nE/XcIU9wtQUeeHJ9h/fPhuCi5CN72YA+sr3rPZFSiUGgqcRg1fsra4QO2rYrkvrzdU8O2qekz+GG\n89bw7sLo+GNaFpfi5DTN1mFms+VSW0mV61zO3IfQf9L3ozzOwpPsmXjxmBvoM34O0rfc7fMNuF5+\nd+CkKrdDSafjqs3vA4/Hd9/fA2ZEv0r6tpn9Y93j0vPlBIX3weXSk/va3Gt8Lc19EX4SfGbbnBfE\nzl3Sv+A+1JcBj5T0PjP7fOZlRuaf3EOD+8hWkVNisIocHfHI9cnmKRoGWqgryAnvTj0Fpto6ZpMq\nlUOn0PaOpH4/7h1nB+NnTt+pqRe+Wvr/lS393CTpbKYWgnXlmUT7F4LWBIWDGl8XhHDHo7web2Y3\nS1qKR37mCvfZMIQlB0tkkFNisIqcH/vIj4FDWqh718oO764LKCldM9fWMZvc+x3uMvchkPr9qPqt\npRo/c/p+jplt1/bGXnxHIqkLQWuCwkGNrwtFuN9jZjcDmNlN6haYNBv6q5xgiVRySgyuDgxjoS5f\nKym8W+k5Y+bK0yqF8rgHCm0fMVW/tVTjZw7PlXSEmXV2XOgnYyFISVA4kPF1oQj330k6nKkKR7+d\n4/HUkRMskUpOicEq5pVahuEs1BTvTw7vtsScMTYVoDPD1sEAGReHTc7ch8gg349B885X9b0JLlx/\nz5TzxCDZTZMxT1C4FI/GrkxQ2MX4WmahCPc9cCv6zvgO9l0drjEbgisnWCKV5BKDOa57krYEtsTL\nsl1pntN9Ntz8ZnOhztkZ9ttkBrV1ZNPBbbCJQXfFVaTmrakcpwbLO1/V9y4Vz80aZnZT8d8zJT0f\nVzfNoMr4SkK+q4Ui3NfGf3hfAV6Lr3z9blH3Mpf+yaQHS6SSVGIwx3VP0r7Ai3Er/gl4+tt9Z8nN\nbxgLdSqDnFoGtXV0IddtsInOmxl5YNT++O8OADN7hpkd2tcux8UvyfiZ2nfx/B/63vt8GuTCiNmy\n4bVO1eEWinD/MvBp4KW0FFWeY//knGCJVPpLDNbp61LTFICrG3YEvmNmH5f0kwHHmEPWQj0ggxiT\n58LWkZszpYlBbEyNgVHqll8lyfjZ1ncLTQJ2JMirRN1gZpVFXwo6VYdbKMJ9XVxQvsXaiyrnCLlh\nkxMskYSZHZKor8tx3VtMoaMsHt8+6DgzSF6o55hBbR3ZDNFtcFAaA6M6uvilGj+zg7ISBexQkfR0\n3Ch8I15w/rVmVpejqVN1uIUi3HOKKs+lf3Kqj2wybfq6jq57J+M674fIa6l+taHtsMlZqAdlELVM\nsq1jBLS6DQ5ZP99PY83Pji5+qcbP5HqjmQJ22ByKF1u/Sl4g6L+pScCXYnytYqEI97cDL6ShqPI8\n8U/OCZZIpU1f18V17xw8Ve8jATOziwcaYR45C3USmbrfVJtMkq1jRKS4DQ5TP99Pf2BUP11c/FKN\nn219l0kWsCPgbjO7CsDMrmzz+ko1vpZZEMLdzM6X9DtgA3zXV6WamHP/5MxgiVQa9XUdXfeON7Md\n8ECi2aZ1oU6li+43wyaTausYBSlug8PUz0+jUAU+D99UmM3MTZTt4pdq/Gzru48sATtkbpS0H1Ne\nX3/JeG+SbWBBCPc+1cS6uPvcNFei1cU/uQOp+roc171b5DlWDC+uMWu70sSFOvVaXXS/STaZDFvH\nSGhzGxylfl6ejXFL/LN5laSnmtnbK9p1cvErqBRwqX0XDCJgB2U3fGNxGG472rPtDbm2gQUh3Mlz\nJZp1/+RRkqGvy3Hdu6D4u+mAw8smZaHOuFYX3W+STWZAwTUoqTlTYPCw/ip2NLPtASR9Ariwpl22\ni1+CgEvtGzoI2EHRVNqHTfFCL72o4GXUyJyutoGFItxzXInmwj95pCTq65Jd9zKPvsOmk89vDcm6\n3w42mWGOM5dUt0EYTVj/mpIWm9k9TAmvKpJ/lxkCrrXvLgJ2iHRJ+9DJNrBQhHuOK9G45WLpp05f\nl+y6l3n0HTadfH6ryNT95tpkhjbODuTkTBk0rL+KLwLnywu5PAk4taZdzu8yVcCl9D1neXWslPZB\n0ibA5sAKay4O1Mk2sCCEe6GaWB/PtNjmSjTr/smzQcJxNsd1L+foO2w6+fw2kaJC6WCTGfo4M8jK\nmdKmn8/FzA4vXHofDnzGzH5V0y7HxS9JwKX03VHADhVJewDvxJ0SHi7P5fOlmuadbAMLQrgXK/1/\n4Efq0/Cb+eea5nPpnzx0Mo6zOa57qcfuoZO5UKcydJtMV9/kIZGTMyVHP5+EpEfjC+WfgE9K+qCZ\nfaeqbYaLX5KAy+k7U8AOm9cDjzGz2+QlBs8F6vruZBtYEMIdj2I8HA/v/wH+A60zbs2lf/IoSD3O\n5rju9R99vziswbaRuVCnMhKbTBff5GGQ6jZYkKOfT+VoYF+8jN67gY/gcRFtNLn4pQq4nL5zBOyw\nuQG4s/j/36jQFAxqG1gowv0+ZnaupIPMzFp0VnPpnzwKUo+zya57fUff5WZ2yQjGXUfOQp3KbNhk\nZj1vSWLfQ89pjp+qfgWsZWYXqqXQe5PKsIOAy+m7VcAOG3nZwN74f1ZskB5X9N/PQLaBhSLcb5P0\nbGANSU/GvwCVzLV/8ghIPc4mu+5Jei0wYWb/Lulbkk4ys6YizMMkZ6FOIlOFkmWTyfVNHiaJfY8i\np/kkcCLwDUkvY0qA9o8vRWWYK+Ba+84UsMPm6IrnKqthDWobWCjCfW/go/gX+e34caySOfZPHgWp\nx9kcvfPrgScW/38evnDMlnBPXqhzyFChJNlkuvomD4PMvkeR03xX/PvRc7Gs8zRqVRl2EHApfScL\n2GFjZt8HkLQLsK2ZvVfSWXg2y0q62gaG4fa0OvBWM3u5mT3CzF5qZr9vaNsTcmfjfsyzGZI8NArP\nB5g6zr4N+Ay+W6niBvOCG+tZe9Hpu83sLgAzu5PZVV3tjed0b12oB6BJjXG9pGMkvU7S3uoLfirR\nE1yPAbbH677OFsl9m9kfyv/w+r2Dcg9wX+DfgL/H8+9XMU1lSMNvrRBw5+F69B8Vu/JOfZvZ9wsh\nuwHwjOL/BzDdA2rUHMJUPqJdgfc2tO3ZBl4MbIv/lltZKDv3rSXdL/E4M5f+ycMk9zibo3f+mqQf\n4uqLxzG9AtGoeauZjSQPUKIaI9UmM5d5Swbpexi2gV6h90fjqo66Qu85Ln6pxs/UvsEFbK/M4K74\nbn/Q+gmp3GlmqwDMbNUobAMLRrjjO67rmdIr1unS59I/eWjkHmdz9M5m9gFJXwcEnGCzmxUyZ6FO\nIkeNkWGTmcu8Jdl9D9k2kFroPcfFL1XA5RSZzxGww+YiSScDP8LVSD/vbzCobWBBCHcze0jV8yqK\nGve1nUv/5KGTo69r0zvLi3e/DvgkcD3+w7xN0tvN7JpRzaGPnIU6leTw7gybzKznLenS94hsA42F\n3nM8YDoIuJwi860CdlSY2X6SXoRvkE4zszMqmg1kG1gQwr2B/qLGwNz5J4+Irr68VcfzT+IJsxbj\nxbx/grudfRqvqTpychbqDHLUGI2G50F9kwehY9+jyGl+JPBW4Ft4ubvz+l7PURnmCri2vu8lUcCO\nBEkb4ZuDq4ENJR1gZh/qG1+28bXMQhfuKdVm5tI/eRhk6etajuePMLPti53RU4GXmtmdkpIMPCOm\ncqFOJEeN0WaTmbO8JR37HrptwMy+AvcKsNPM7Ma+15NVhrkCrq3vMikCdoT0bAOPwg3JQ7cNLHTh\nXuvlMZf+ycMg9zibeDzvnWi2By4qPGVgftglBikLl6NCabTJDOqbPAgd+x66bUDSjsBRwBrAaZL+\nYGbHV7TLcfFLEnCpfRfkCNhhM3LbwEJxhUxG0tPlxSDOwavV7DzXY+rI0fgO7m34rvYU4N+B99S0\nT3Gfu7lw/XsncKqkxZL+DfjjsAffgWx3zA7uor1atsfgrnNfAF5Qc+1U172hk9n3brjL4GHAgxmO\nbeAD+EJxDfBB6vPV5Lj4TRNwQJ2AS+0bCgGLF53ZGdiooe2wybYNSNpP0kkk2gYWunCv2u3NpX/y\n0Ojgy5vic7wPsAW+azoB30m9lNH4ms8G+xd/j8EXw0+X/l+Lmd1kZneZl9Z7Qk2zTr7JQ6K17y4L\nWwb3mNlfcGP3bUyd+PrJURmmCrjUviFPwA6bI/E4jUvwJGe1sTdmth9uJ1sXVzW9KaWDBaOWUVpR\nY5hb/+RRkKqvaz2eF8FN75T0fPOAp+8A3yl2hrOVcKmObLXMkFQodTaZWc9bktn3KG0DK+Q5/zeW\n9C76kpZ1cfHLMH429t1Hv4BtUo0Mm0lcHbMSuB3PzFlJV9vAghDuSi9qDHPrnzwKUvV1rXrnwnPo\nKcArJfXyjyzGC1bPmnDPWKhTr5cd3l1nkxnUN3kQcvoesW3gDfj35zzgFuC1fa9nu/hlCLi2vssk\nC9gRcDDwRDO7VtKmeD3gOiNpJ9vAghDuJBY1LphL/+RR0OjLm+k+90tgY1xYWPHcPdRX2hk6mQt1\nKsnuogmG5znLW9Kl7y4LWwJfN7Nn1b3Y0cUvVcA19t1HjoAdNjeY2bUAZvZnSbVePeQZX+9loQj3\n1qLGc+mfPEoSjrPJx3MzuwI4QZ4F8p7RjryWnIU6lRwVSqNf+KC+yYPQse9R5DRfKekFwG/wxR8z\n+01FuxwXv1QBl9o35AnYYXOTPG3294HHA+tK+mAxlgP72nayDYy1cFdeUeO59E8eGW3H2Y7H83dK\neie+e1rEcKJEU2ldqFPpqEJJtcnMZd6SnL5HYRu4Px5I1KPuN5Tj4pcq4FL7hjwBO2y+Wvr/lS1t\nO9kGxlq4k1HUeC79k0dM0nE283j+cmAzM5s1v+DMhTqVLiqUVJvMXOYtae17lLYBM3t61fPFd+qQ\n0lM54f9JAi6jb8gTsEPFzE7IaN7JNjDWwt3yixrPdV3FUZB6nM05nv+e2SlsUGbomSA7qjFSbTJz\nlrckse+5sA3sVH6Q4QEDgxs/d+p/IlPAziWdbAMLxc/9BOBi/APp/atjLv2TR0HqcTbneL4WcImk\nUwrf45OHNtoabCrf+KZ4UNY7S/8GpTW3dq5feFff5GGQ0neHOIhhMM1dtUpl2PDenoB7LJ647dBB\n+l7NmGYbwI35rYz1zr1EclFj5tY/eRQ0Hmc7Hs9bKxGNkBOK/odp5E5RoWTZZLr6Jg+DzL5n0zbQ\nH0Wc4+I3qPFzda6F3Mk2sFCEe2tR47n0Tx4xbcfZLsfz/8V3zJvhGTNnM597zkKdSqsao4NNZi7z\nluT0PZe2gRwXv7k0fs41nWwDC0W4pxQ1nkv/5FHSqK/rqHdeju/wdsJzeBxPhU5zRLQu1Lnk6H4z\nbDKdfJOHRE7fQ7MNSJpocDuEmaqRHBe/QY2fq61apqttYKEI99aixnPpnzxiUo+zOcfzjc1suaTd\nzOwCSbNpu0lZqLPIVGOkGp7nMm9Jct+ZRs02TgKeJOn0wmbVz+59j5Nd/FIFnKRPmdm+pccnmtnu\nFX2PPQtFuF8v6RhcnTAJYGbH1rSdS//kUZB6nM06nkt6ePH3QcBdIxl5Na0LdQeydL+k2WTmMm9J\nct9Dtg38TtK1wH0lXcXUbnnSzDYrguDKDC38X9IbcS+mjSS9pOh7EV5Mhoq+x56FItxTixrD3Oog\nR0HqcTbneP4m/Ie4FV6JqCmt6rDJWahTaVVjdLDJzGXekpy+h2YbMLNXAEg60szemPCWoYX/m9mR\nwJGSDjSzD3a5xrixIIS7pRc1hrn1Tx46qcfZnOO5mV2Ku6PNBTkLdSopaoxcm8xc5i3J6XsUtoED\nJB0K/B2Fwd3MVlS0G1r4v7zOAPjiv3f5tSEs/qslC0K4K72o8ZzWVZxLco7nkg7Dg3fuFayzlX4g\nc6FOpVWN0cEmM5d5S3L6HoVt4Hhcnfk0mg3uw/SAeWDN86uzC+RALAjhTktR4zJz6Z88x+Qcz58H\nbG5mt8/GwMrkLNQZ5KgxUm0yc+m6l9P3KGwDqQb3oYX/l1MLjGDxXy1ZKMK9rahxmbn0T55Lco7n\nP8cjGWdduJOxUGeQo8ZItcnMWd6SzL5HYhtIMbiPIvx/RIv/aslCEe6NRY37mEv/5Lkk53h+KXC1\npGuYygr50FkYI+Qt1DnXTFVjJNlk5jJvSWbfo7ANzKXBfRSL/2rJghDuZnagpKW4Z0Ov0EMdc+mf\nPJfkHM93Bf6BuUnNkLNQp5KsxhhDm8wobAPPMbO5MriPYvFfLVkoggsz6xXKPVNeLu7rNU3n0j95\nLsk5nv8BuGUudO6ZC3UqyWqMMbTJjMI28FxJR5jZXLgRj2LxXy1ZMMK9j7qixjC3/slzyf9v795D\nLSvLOI5/hxjTmQrzkqkzkUg9pVje8NJlQmOQ6KZIahDd6KKJRSpFanSz0qDbH5FaWlEoZGFWFF2o\nHMQJ0SK6+WQzUjnapGOl4iXL0x9rbc45+8w5s9fy7P2us9f3A4dz3BxnP+D4e971rrXfp8nl+Xpg\nS0Rsrf95JjNftMjvLrsGjXrUP6/JNsa03ZMZx72BfajC9Q6q/58m9vdjTM1/RepVuMciQ42HlHw+\nuaQml+enT6imUSzVqMdhqu7JjOnewKvH8GeObLmb/0rVi3CPXQ81nqvk88klNbk8fxx4PfPP/n5C\nh3c1NWKjHoe+3pMZWdZDcgbqgP3LIr8+bpNu/p3Rl7+YSw41HtLXo0WbXJ5fC/wUmPh5HQ0b9Tj0\n9Z7MEzHxgC3Y/DujL+E+6lBjKPt8cjENL88fyMyLxlbM0po06nHo6z2ZxkoEbAeaf2f0JdxHHWq8\nkuYqlvS7iDiD6hnvweFdS53jvZyaNOpx6Os9mZEVDtjSzb8z+hLuow411mgOr78GdjpqbkxGbtRj\n0td7Mk2UDNjSzb8zpjrcI2JdZt7J7FDjwczLfVneGZy9kpkn7Pq3xqZ0o+7rPZkmSgZs6ebfGVMd\n7jQcaqzRRMQ24BnAPVTPND8CbAfeNa7L7w416l7ek2moZMCWbv6dMdXhns2HGms0m4APZ2ZGxMHA\nh6gehfwG47v87kSj9p7MSCYesB1q/p0xydmXxUQ11PhG4EJgc0ScVriklW5dZiZAZm4BnlUPYxjb\nuL25jRo4DXgfcEpmegXWEfUJkDAbsOcBX6YK2HE7t/5+OdVglS/O+bmXpnrlPseoQ401mrsj4hLg\nJuBFwN8jYiPwn3G/cd2o3091BMDzIuJDmel/y24odnXlVfpCvVi5M/pQY43mjcBdVGd3/BV4M/Ag\n1adWx23QqE8BjqZaHaoDunB15VX6rKleuUfzocZaQkQcnZm3UN0ku63+AtiQmT+eUBk26o4rfHXl\nVXptqsOd5kONtbSXA7ewcIU+A4w13G3UK0rJgLX516Y63FsMNdbSPhsRuwHvLPDeNuqVY+IBa/Nf\naKrDfY5RhxpracnCafKDG2ZjHbNno+6+wgFr8x/Sl3AfdaixlpCZB5WuARt1lxULWJv/Qn0J95GG\nGms0EfEa4GxgNdXKfe/MfMGE3t5G3VEdCVibf60X4T6FQ41Lu5hq3/1M4OfAxgm+t426+0oGrM2/\n1ovn3Hc21LhwSSvd3Zm5GSAzvwocOKk3zsxzqJ68WEPVqN89qffWyOYFLDDJgL05Iq6OiHMi4uv0\nuPn3YuXO9A01Lu3RiNgArI6Ik6gOD5uInTXqzPzkpN5fIyl2deVV+qxerNyphxpTPe2xEdircD0r\n3VlU++0XA++ov0/KdcDLqLaF3kh1/IE6pOTVlVfps/qycneo8TLKzG3MHnd76uD1iLiuPhZgnFZl\n5pkRcRXVuDtnmHZM4asrr9JrfVm5Dw81vqNsOVNrzwm8h426+0peXXmVXutLuA+GGp+KQ43HafgD\nTuNgo+6+kgFr86/1JdwHQ42PAI6nmvGolclG3X0lA9bmX+tLuM8bakw1lV0rk426+0oGrM2/1pdL\nFocaT8YkxpnNa9QRYaPunkHA/pPJB+yg+f8jIvYDvoefUJ1qDjVeRhGxnurY390Hr2XmRzPz1MX/\nrWVjo+6+kgFr86/1ItwdarzsrgV+CvytwHvbqLuvZMDa/Gu9CHctuwcy86ISb2yjXhFKBqzNv2a4\nq43fRcQZVB8rnwHIzD+VLUkdUixgbf6zVs3MTOLRZE2TiPj50EszkxyCLGnXDHe1EhF7AwcDWzPz\n3tL1SJqvL8+5axlFxOuAm4ALgF9GxBsKlyRpiOGuNs4FjsrMk4EjgPcUrkfSEMNdbTyemQ8CZOYD\nVKfvSeoQn5ZRG1sj4tPAJmADsKVwPZKGuHJXG28BtlKd+LcFeHvZciQNM9zVxpOB66kmMO0NHFC2\nHEnDDHe18S3gSOBTwGPAFWXLkTTMcFcba6gOg1qXmZcATypcj6Qhhrva2I3q8cdbI+IQqqEMkjrE\ncFcb51Pts38cOBGfc5c6x+MH1EpE7A+sBlYBB2Tm5sIlSZrD59zVWERcSTXibi3V/vsW4LiiRUma\nx20ZtfFC4FCq6TrPx0+oSp1juKuNHZk5A6z1REipm9xzV2P1VJ37gP2A9cBBmXls2aokzWW4q5WI\neArVdswrgJszc3vhkiTN4baMGouIA4HLgB8AzwSeXbQgSQsY7mrjCuAqqkchNwGfL1uOpGGGu9rY\nIzN/RjU7NfFpGalzDHe18UhEnAQ8KSKOw3CXOsdwVxvvoDrTfR+qowjOKluOpGF+QlVtvDczzyhd\nhKTFuXJXG4dExJ6li5C0OJ9zV2MR8RfgQOBeYIbqxqrTmKQOMdy1bCLitZl5fek6JLkto+Xlue5S\nRxjuWk6rShcgqWK4azm5xyd1hOEuSVPIcNdycltG6gg/xKRWIuIw4LnA7zPztvrlzxQsSdIcPgqp\nxiLiIupz3IFjgW9m5ufKViVpLrdl1MYrgZdm5nuBDYBHEUgdY7irje3Amvrn3YB7CtYiaSfcc9fI\nImIz1eOOzwBuj4jfAIcAO4oWJmkBw11NuP0irRDeUFVjEXEMVdDvPngtM99VriJJw1y5q42vAZcC\n/yxdiKSdM9zVxu2Z+dXSRUhanNsyaiwi3kT1nPsfBq9l5kfLVSRpmCt3tXE28G3gX6ULkbRzhrva\n2JGZl5YuQtLi3JZRYxHxdeAh4FfUx/xm5hVFi5I0jyt3tfHn+vsz6++uEKSOceWuViJif2A11TG/\nB2Tm5sIlSZrDlbsai4grgeOBtcAewFbguKJFSZrHg8PUxguBQ4EfUZ0t80jZciQNM9zVxo7MnAHW\nZua9pYuRtJB77mosIj4B3AfsB6wHDsrMY8tWJWkuw12tRMRTgYepJzJl5vbCJUmaw3DXExYRr8rM\n75euQ9Is99y1HJ5TugBJ87lyV2sRsS/VzdXHS9ciaT7DXY1FxAnAlcD9wJ7A2zPzJ2WrkjSX2zJq\n42PASzLzcODFwMWF65E0xHBXG//LzLsAMnMbfohJ6hyPH1Ab90fEOcAmYAPVM++SOsSVu9p4A/As\n4ONUH2J6a9lyJA1z5a6RRcS6zLyT6pOpX6I6EXIG2BeHZUudYririXPrr8uZPcN9EPAnlipK0kI+\nCqlWImIf4NnAnzPTWapSx7jnrsYi4i3AjcCFwOaIOK1wSZKGGO5q4yzg8Mw8BTgaOK9wPZKGGO5q\nYwfwWP3zw4DbMlLHeENVI4uIa5h9OubWiPglcCRVwEvqEMNdTVy2k9eumXgVknbJbRmNLDNvyMwb\ngKcBJ9Y/fwDYvWxlkoa5clcbHwFOqH8+Hfgh1bBsSR3hyl1tPJaZ/waov/+vcD2ShrhyVxs3R8TV\nwGbgGODXheuRNMRPqKqViDgZCOCPmfnd0vVIms9tGTUWEXsBewB3A0+PiA8ULknSELdl1MZ1wB+B\nw6gGdTxUthxJw1y5q41VmXkmkMBGYK/C9UgaYrirjf9GxO7AWqpPrHoFKHWM4a42vgCcD/wWuBO4\no2w5koa54lIbM8DbqKYvPQp8pWw5koa5clcbHwSOycwjgOOBjxWuR9IQw11t7MjMfwBk5nbg/sL1\nSBrih5jUWERcB6wBbgCOAvYHfgGQmReUq0zSgHvuauM7c37eVqwKSYty5S5JU8g9d0maQoa7JE0h\nw12SppDhLklTyHCXpCn0fyPKh4nEXH5PAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x116a27cf8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "feature_importances = pd.Series(model.feature_importances_, index=X_test.columns)\n",
    "feature_importances.sort()\n",
    "feature_importances.plot.bar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#TODO: Try training the model with the most important features and note difference in overal model prediction score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0    143\n",
       "0.0    143\n",
       "2.0     99\n",
       "dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: Training Set - Show number of Hom Ref, Hom Var, Het Var datapoints the model was trained on\n",
    "ytrain = pd.DataFrame()\n",
    "ytrain['ytrain'] = y_train\n",
    "pd.value_counts(ytrain['ytrain'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='traintest_precision'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the training subset: 0.861\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the training subset: {:.3f}'.format(precision_score(pred, y_test, average='micro'))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>pacbio.alt_alnScore_mean</th>\n",
       "      <th>pacbio.alt_alnScore_std</th>\n",
       "      <th>pacbio.alt_count</th>\n",
       "      <th>pacbio.alt_insertSize_mean</th>\n",
       "      <th>pacbio.alt_insertSize_std</th>\n",
       "      <th>pacbio.alt_reason_alignmentScore</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>-318.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>55287869.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8720.00000</td>\n",
       "      <td>4490.347537</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55287551.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.128931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>-23.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>36570283.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15429.08824</td>\n",
       "      <td>6092.111898</td>\n",
       "      <td>34.0</td>\n",
       "      <td>9845.352941</td>\n",
       "      <td>3732.076491</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10351.50000</td>\n",
       "      <td>2457.487657</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36570259.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>-21.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>97850858.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10500.54902</td>\n",
       "      <td>4184.292506</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97850836.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>-22.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>130473394.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>18887.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13565.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10894.00000</td>\n",
       "      <td>3911.380662</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130473371.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>-23.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25982840.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11695.00000</td>\n",
       "      <td>3749.278458</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25982816.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Size  chrom          end  pacbio.GT  pacbio.alt_alnScore_mean  \\\n",
       "85  -318.0    1.0   55287869.0        0.0                   0.00000   \n",
       "441  -23.0   15.0   36570283.0        1.0               15429.08824   \n",
       "96   -21.0    1.0   97850858.0        0.0                   0.00000   \n",
       "172  -22.0   10.0  130473394.0       -1.0               18887.00000   \n",
       "134  -23.0   10.0   25982840.0        0.0                   0.00000   \n",
       "\n",
       "     pacbio.alt_alnScore_std  pacbio.alt_count  pacbio.alt_insertSize_mean  \\\n",
       "85                  0.000000               0.0                    0.000000   \n",
       "441              6092.111898              34.0                 9845.352941   \n",
       "96                  0.000000               0.0                    0.000000   \n",
       "172                 0.000000               1.0                13565.000000   \n",
       "134                 0.000000               0.0                    0.000000   \n",
       "\n",
       "     pacbio.alt_insertSize_std  pacbio.alt_reason_alignmentScore  \\\n",
       "85                    0.000000                               0.0   \n",
       "441                3732.076491                              34.0   \n",
       "96                    0.000000                               0.0   \n",
       "172                   0.000000                               1.0   \n",
       "134                   0.000000                               0.0   \n",
       "\n",
       "         ...        pacbio.ref_insertSize_mean  pacbio.ref_insertSize_std  \\\n",
       "85       ...                        8720.00000                4490.347537   \n",
       "441      ...                       10351.50000                2457.487657   \n",
       "96       ...                       10500.54902                4184.292506   \n",
       "172      ...                       10894.00000                3911.380662   \n",
       "134      ...                       11695.00000                3749.278458   \n",
       "\n",
       "     pacbio.ref_reason_alignmentScore  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "85                               22.0       0.0       0.0         0.0   \n",
       "441                               6.0       0.0       0.0         0.0   \n",
       "96                               51.0       0.0       0.0         0.0   \n",
       "172                              35.0       0.0       0.0         0.0   \n",
       "134                              42.0       0.0       0.0         0.0   \n",
       "\n",
       "     segdup_pct        start  tandemrep_cnt  tandemrep_pct  \n",
       "85          0.0   55287551.0            2.0       0.128931  \n",
       "441         0.0   36570259.0            0.0       0.000000  \n",
       "96          0.0   97850836.0            1.0       1.000000  \n",
       "172         0.0  130473371.0            0.0       0.000000  \n",
       "134         0.0   25982816.0            0.0       0.000000  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add original labels and predicted labels back to the original dataframe\n",
    "df_Xtest = pd.DataFrame(X_test)\n",
    "df_Xtest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "labels = pd.DataFrame(y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['predicted_label'] = pred\n",
    "df_Xtest['GTconswithoutpacbio.GT'] = df_train['GTconswithoutpacbio.GT']\n",
    "df_Xtest['chrom'] = df_train['chrom']\n",
    "df_Xtest['start'] = df_train['start']\n",
    "df_Xtest['end'] = df_train['end']\n",
    "# df_Xtest['Y_test'] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['GTconswithoutpacbio.GT'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['GTconswithoutpacbio.GT'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['GTconswithoutpacbio.GT'].replace(2.0, 'Homozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    57\n",
       "Heterozygous_Variant    57\n",
       "Homozygous_Variant      51\n",
       "dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['GTconswithoutpacbio.GT'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    54\n",
       "Heterozygous_Variant    68\n",
       "Homozygous_Variant      43\n",
       "dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['predicted_label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[52  2  3]\n",
      " [ 7 50  0]\n",
      " [ 9  2 40]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "ytest = df_Xtest['GTconswithoutpacbio.GT']\n",
    "predict = df_Xtest['predicted_label']\n",
    "print(confusion_matrix(ytest, predict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='traintest_confusion_matrix'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>52</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>7</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>68</td>\n",
       "      <td>54</td>\n",
       "      <td>43</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                    52                     2   \n",
       "Homozygous_Reference                     7                    50   \n",
       "Homozygous_Variant                       9                     2   \n",
       "All                                     68                    54   \n",
       "\n",
       "Predicted             Homozygous_Variant  All  \n",
       "True                                           \n",
       "Heterozygous_Variant                   3   57  \n",
       "Homozygous_Reference                   0   57  \n",
       "Homozygous_Variant                    40   51  \n",
       "All                                   43  165  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(ytest, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       0.76      0.91      0.83        57\n",
      "Homozygous_Reference       0.93      0.88      0.90        57\n",
      "  Homozygous_Variant       0.93      0.78      0.85        51\n",
      "\n",
      "         avg / total       0.87      0.86      0.86       165\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(ytest, predict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Predict\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='predict'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the [previous section](#machine_learning), a RF model was trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](#train_test) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data\n",
    " * Reminder: The labels for this training set and the following [prediction step](#prediction_step) are the consensus genotype (GTcons) labels generated from a preliminary R analysis based on reference and alternate read count:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "           \n",
    "   * The trained model is used in the following section to predict labels for 5000 randomly selected Deletions [these datapoints were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "   \n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Load Data **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X2 = df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Impute missing values using KNN **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/3281 with 0 missing, elapsed time: 2.843\n",
      "Imputing row 101/3281 with 0 missing, elapsed time: 2.844\n",
      "Imputing row 201/3281 with 0 missing, elapsed time: 2.845\n",
      "Imputing row 301/3281 with 0 missing, elapsed time: 2.846\n",
      "Imputing row 401/3281 with 0 missing, elapsed time: 2.847\n",
      "Imputing row 501/3281 with 0 missing, elapsed time: 2.848\n",
      "Imputing row 601/3281 with 0 missing, elapsed time: 2.850\n",
      "Imputing row 701/3281 with 0 missing, elapsed time: 2.850\n",
      "Imputing row 801/3281 with 0 missing, elapsed time: 2.852\n",
      "Imputing row 901/3281 with 0 missing, elapsed time: 2.853\n",
      "Imputing row 1001/3281 with 0 missing, elapsed time: 2.854\n",
      "Imputing row 1101/3281 with 0 missing, elapsed time: 2.856\n",
      "Imputing row 1201/3281 with 0 missing, elapsed time: 2.858\n",
      "Imputing row 1301/3281 with 0 missing, elapsed time: 2.859\n",
      "Imputing row 1401/3281 with 0 missing, elapsed time: 2.860\n",
      "Imputing row 1501/3281 with 0 missing, elapsed time: 2.860\n",
      "Imputing row 1601/3281 with 0 missing, elapsed time: 2.861\n",
      "Imputing row 1701/3281 with 0 missing, elapsed time: 2.861\n",
      "Imputing row 1801/3281 with 0 missing, elapsed time: 2.862\n",
      "Imputing row 1901/3281 with 0 missing, elapsed time: 2.863\n",
      "Imputing row 2001/3281 with 0 missing, elapsed time: 2.864\n",
      "Imputing row 2101/3281 with 0 missing, elapsed time: 2.864\n",
      "Imputing row 2201/3281 with 0 missing, elapsed time: 2.865\n",
      "Imputing row 2301/3281 with 0 missing, elapsed time: 2.865\n",
      "Imputing row 2401/3281 with 0 missing, elapsed time: 2.866\n",
      "Imputing row 2501/3281 with 0 missing, elapsed time: 2.867\n",
      "Imputing row 2601/3281 with 0 missing, elapsed time: 2.867\n",
      "Imputing row 2701/3281 with 0 missing, elapsed time: 2.868\n",
      "Imputing row 2801/3281 with 0 missing, elapsed time: 2.868\n",
      "Imputing row 2901/3281 with 0 missing, elapsed time: 2.869\n",
      "Imputing row 3001/3281 with 0 missing, elapsed time: 2.869\n",
      "Imputing row 3101/3281 with 0 missing, elapsed time: 2.870\n",
      "Imputing row 3201/3281 with 0 missing, elapsed time: 2.870\n"
     ]
    }
   ],
   "source": [
    "#Convert dataframe to matrix\n",
    "X2=X2.as_matrix()\n",
    "X2=pd.DataFrame(X2)\n",
    "\n",
    "# Imput missing values from three closest observations\n",
    "X2_imputed=KNN(k=3).complete(X2)\n",
    "X2=pd.DataFrame(X2_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37568322.0</td>\n",
       "      <td>37568587.0</td>\n",
       "      <td>-115.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.818868</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.378230</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>112835104.0</td>\n",
       "      <td>112837661.0</td>\n",
       "      <td>-2534.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.460305</td>\n",
       "      <td>...</td>\n",
       "      <td>159.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11538.50000</td>\n",
       "      <td>5911.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7826.000000</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3557049.0</td>\n",
       "      <td>3557087.0</td>\n",
       "      <td>-37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8573.00000</td>\n",
       "      <td>4836.734918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6021.000000</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end    Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0   37568322.0   37568587.0  -115.0       0.0       0.0         0.0   \n",
       "1    1.0  112835104.0  112837661.0 -2534.0       0.0       0.0         1.0   \n",
       "2    1.0    3557049.0    3557087.0   -37.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000            3.0       0.818868                ...                  \n",
       "1    0.890888            6.0       0.460305                ...                  \n",
       "2    0.000000            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "1                                            159.0   \n",
       "2                                            151.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "1                       270.0                             8.0   \n",
       "2                       207.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "1                            9.0               11538.50000   \n",
       "2                           10.0                8573.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7611.378230              31.0                 8407.903226   \n",
       "1              5911.500000               2.0                 7826.000000   \n",
       "2              4836.734918               3.0                 6021.000000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "1                4015.000000                               2.0  \n",
       "2                3561.905763                               3.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dftest_header = list(df_test.columns.values)\n",
    "X2.columns = dftest_header\n",
    "X2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37568322.0</td>\n",
       "      <td>37568587.0</td>\n",
       "      <td>-115.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.818868</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.378230</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>112835104.0</td>\n",
       "      <td>112837661.0</td>\n",
       "      <td>-2534.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.460305</td>\n",
       "      <td>...</td>\n",
       "      <td>159.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11538.50000</td>\n",
       "      <td>5911.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7826.000000</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3557049.0</td>\n",
       "      <td>3557087.0</td>\n",
       "      <td>-37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8573.00000</td>\n",
       "      <td>4836.734918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6021.000000</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end    Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0   37568322.0   37568587.0  -115.0       0.0       0.0         0.0   \n",
       "1    1.0  112835104.0  112837661.0 -2534.0       0.0       0.0         1.0   \n",
       "2    1.0    3557049.0    3557087.0   -37.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000            3.0       0.818868                ...                  \n",
       "1    0.890888            6.0       0.460305                ...                  \n",
       "2    0.000000            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "1                                            159.0   \n",
       "2                                            151.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "1                       270.0                             8.0   \n",
       "2                       207.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "1                            9.0               11538.50000   \n",
       "2                           10.0                8573.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7611.378230              31.0                 8407.903226   \n",
       "1              5911.500000               2.0                 7826.000000   \n",
       "2              4836.734918               3.0                 6021.000000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "1                4015.000000                               2.0  \n",
       "2                3561.905763                               3.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X3 = pd.DataFrame()\n",
    "X3 = X2\n",
    "X3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X5 = X2.reindex_axis(sorted(X2.columns), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# X5 = df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred_prob = model.predict_proba(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred_prob_log = model.predict_log_proba(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X5['predicted_label'] = pred\n",
    "X5['chrom'] = df_test_2['chrom']\n",
    "X5['GTcons'] = df_test_2['GTcons']\n",
    "X5['start'] = df_test_2['start']\n",
    "X5['end'] = df_test_2['end']\n",
    "X5['Size'] = df_test_2['Size']\n",
    "X5['GTconswithoutpacbio.GT'] = df_test_2['GTconswithoutpacbio.GT']\n",
    "X5['GTsupp'] = df_test_2['GTsupp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.concat([X5, pd.DataFrame(pred_prob, columns=['1','2','3'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X7 = pd.concat([X5, pd.DataFrame(pred_prob_log, columns=['1','2','3'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_DEL_revisedR_LOO.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X7.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_log_DEL_revisedR_LOO.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Note: Reformat X6 csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_DEL_revisedR_LOO.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.rename(columns={'1': 'Homozygous_Reference_LOO_PB'}, inplace=True)\n",
    "X6.rename(columns={'2': 'Heterozygous_Variant_LOO_PB'}, inplace=True)\n",
    "X6.rename(columns={'3': 'Homozygous_Variant_LOO_PB'}, inplace=True)\n",
    "X6.rename(columns={'predicted_label': 'PB_predicted_LOO_label'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homozygous_Reference_LOO_PB</th>\n",
       "      <th>Heterozygous_Variant_LOO_PB</th>\n",
       "      <th>Homozygous_Variant_LOO_PB</th>\n",
       "      <th>GTcons</th>\n",
       "      <th>GTconswithoutpacbio.GT</th>\n",
       "      <th>GTsupp</th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>PB_predicted_LOO_label</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>-115</td>\n",
       "      <td>1</td>\n",
       "      <td>37568587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37568322</td>\n",
       "      <td>3</td>\n",
       "      <td>0.818868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.6</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2534</td>\n",
       "      <td>1</td>\n",
       "      <td>112837661</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>112835104</td>\n",
       "      <td>6</td>\n",
       "      <td>0.460305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-37</td>\n",
       "      <td>1</td>\n",
       "      <td>3557087</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3557049</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Homozygous_Reference_LOO_PB  Heterozygous_Variant_LOO_PB  \\\n",
       "0                          0.9                          0.0   \n",
       "1                          0.6                          0.1   \n",
       "2                          0.1                          0.3   \n",
       "\n",
       "   Homozygous_Variant_LOO_PB  GTcons  GTconswithoutpacbio.GT  GTsupp  Size  \\\n",
       "0                        0.1       0                       0       4  -115   \n",
       "1                        0.3       2                       2       1 -2534   \n",
       "2                        0.6       1                       1       2   -37   \n",
       "\n",
       "  chrom        end  pacbio.GT      ...        pacbio.ref_insertSize_std  \\\n",
       "0     1   37568587        0.0      ...                      4770.945293   \n",
       "1     1  112837661       -1.0      ...                      4015.000000   \n",
       "2     1    3557087       -1.0      ...                      3561.905763   \n",
       "\n",
       "   pacbio.ref_reason_alignmentScore  PB_predicted_LOO_label  refN_cnt  \\\n",
       "0                              31.0                       0         0   \n",
       "1                               2.0                       0         0   \n",
       "2                               3.0                       2         0   \n",
       "\n",
       "   refN_pct  segdup_cnt  segdup_pct      start  tandemrep_cnt  tandemrep_pct  \n",
       "0         0           0    0.000000   37568322              3       0.818868  \n",
       "1         0           1    0.890888  112835104              6       0.460305  \n",
       "2         0           0    0.000000    3557049              0       0.000000  \n",
       "\n",
       "[3 rows x 39 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X6.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/PB_final_GTcons_LOO_DEL_SVanalyzer.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Label Analysis\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "  * The [random forest(RF) model](#train_test) was trained on svanalyzer data. The trained model was used to predict consensus GT labels for the 5000 deletions that were randomly selected from the union_refalt vcf. The following is a comparison of model predicted labels [Conesus GT] to consensus genotype generated by the R script for the 5000 randomly selected datapoints from union_refalt.vcf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 751   18   19]\n",
      " [ 198 1478  136]\n",
      " [  25   64  592]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "consensus_GT = X6['GTconswithoutpacbio.GT']\n",
    "predict = X6['PB_predicted_LOO_label']\n",
    "print(confusion_matrix(consensus_GT, predict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6['GTconswithoutpacbio.GT'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "X6['GTconswithoutpacbio.GT'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "X6['GTconswithoutpacbio.GT'].replace(2, 'Homozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_LOO_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "X6['PB_predicted_LOO_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_LOO_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='predict_precision_score_matrix'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the prediction subset: 0.860\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the prediction subset: {:.3f}'.format(precision_score(consensus_GT, predict, average='micro'))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>1478</td>\n",
       "      <td>198</td>\n",
       "      <td>136</td>\n",
       "      <td>1812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>18</td>\n",
       "      <td>751</td>\n",
       "      <td>19</td>\n",
       "      <td>788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>64</td>\n",
       "      <td>25</td>\n",
       "      <td>592</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>1560</td>\n",
       "      <td>974</td>\n",
       "      <td>747</td>\n",
       "      <td>3281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                  1478                   198   \n",
       "Homozygous_Reference                    18                   751   \n",
       "Homozygous_Variant                      64                    25   \n",
       "All                                   1560                   974   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                 136  1812  \n",
       "Homozygous_Reference                  19   788  \n",
       "Homozygous_Variant                   592   681  \n",
       "All                                  747  3281  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** High Confidence Label Analysis**\n",
    "* **Reminder:** The labels predicted by the model are the following consensus genotype:\n",
    "    * Homozygous Reference: 0 \n",
    "    * Heterozygous Variant: 1 \n",
    "    * Homozygous Variant: 2 \n",
    "* Here **high confidence labels** are the LOO GTcons labels predicted by the model that were also assigned a predict probability of either 0.9 or 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "high_conf_labels = X6[(X6['Homozygous_Reference'] == 1) | (X6['Homozygous_Reference'] == 0.9) | (X6['Heterozygous_Variant'] == 1) | (X6['Heterozygous_Variant'] == 0.9) | (X6['Homozygous_Variant'] == 1) | (X6['Homozygous_Variant'] == 0.9)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='hiconf_precision_score'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>711</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>2</td>\n",
       "      <td>713</td>\n",
       "      <td>2</td>\n",
       "      <td>717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>311</td>\n",
       "      <td>317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>713</td>\n",
       "      <td>742</td>\n",
       "      <td>335</td>\n",
       "      <td>1790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                   711                    23   \n",
       "Homozygous_Reference                     2                   713   \n",
       "Homozygous_Variant                       0                     6   \n",
       "All                                    713                   742   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                  22   756  \n",
       "Homozygous_Reference                   2   717  \n",
       "Homozygous_Variant                   311   317  \n",
       "All                                  335  1790  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_GT = high_conf_labels['GTconswithoutpacbio.GT']\n",
    "predict = high_conf_labels['PB_predicted_LOO_label']\n",
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       1.00      0.94      0.97       756\n",
      "Homozygous_Reference       0.96      0.99      0.98       717\n",
      "  Homozygous_Variant       0.93      0.98      0.95       317\n",
      "\n",
      "         avg / total       0.97      0.97      0.97      1790\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(consensus_GT, predict))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [NIHFAES]",
   "language": "python",
   "name": "Python [NIHFAES]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
