{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PB: GTcons\n",
    "\n",
    "* The following notebook is trained on data generated from revised R script [Oct 12 2017]\n",
    "    * Exact Match [1] and Homozygous [0] Reference data points\n",
    "    * Removed all data points with Gtcons and GTconswithoutXX -1\n",
    "* 5k randomly selected deletions test data was also processed through same R script\n",
    "* Balanced Training Set for GTcons labels\n",
    "    * 200 Homozygous Reference\n",
    "    * 200 Heterozygous Variant\n",
    "    * 200 Homozygous Variant\n",
    "* **Train/Prediction Label:** consensus genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/cross_validation.py:44: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "\n",
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/grid_search.py:43: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Imports\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from fancyimpute import KNN\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy import stats\n",
    "from matplotlib import pyplot\n",
    "from sklearn import preprocessing\n",
    "from scipy.linalg import svd\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import seaborn as sns\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "import plotly.plotly as py\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import metrics\n",
    "from sklearn.grid_search import GridSearchCV\n",
    "from sklearn.metrics import f1_score, precision_score\n",
    "from sklearn import preprocessing\n",
    "from ggplot import *\n",
    "from bokeh.charts import TimeSeries\n",
    "from bokeh.models import HoverTool\n",
    "from bokeh.plotting import show\n",
    "from bokeh.charts import Scatter, Histogram, output_file, show\n",
    "from bokeh.plotting import figure, show, output_file, ColumnDataSource\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.charts import Bar, output_file, show\n",
    "import bokeh.palettes as palettes\n",
    "from bokeh.models import HoverTool, BoxSelectTool, Legend\n",
    "from sklearn import (manifold, datasets, decomposition, ensemble,\n",
    "                     discriminant_analysis, random_projection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>sample</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HG002</td>\n",
       "      <td>21</td>\n",
       "      <td>65326531</td>\n",
       "      <td>65326651</td>\n",
       "      <td>-120</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom sample  id     start       end  Size      type    SVtype  refN_cnt  \\\n",
       "0     1  HG002  21  65326531  65326651  -120  Deletion  Deletion         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            187.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       236.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           11.0               18112.46296   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7134.848382              54.0                 11277.83333   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4197.626206                              54.0  \n",
       "\n",
       "[1 rows x 41 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Training Data\n",
    "# SVanalyzer generated training data\n",
    "df_train = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/training_data/tech_sep/HG002/DEL/PB_EM_HR_HG002_DEL_balanced_label_min1.csv')\n",
    "df_train_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/training_data/tech_sep/HG002/DEL/PB_EM_HR_HG002_DEL_balanced_label_min1.csv')\n",
    "df_train.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set = pd.DataFrame()\n",
    "train_set = df_train_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set['GTcons'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "train_set['GTcons'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "train_set['GTcons'].replace(2, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='imbalance'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Heterozygous_Variant    200\n",
       "Homozygous_Variant      200\n",
       "Homozygous_Reference    200\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(train_set['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train the model only on the rows that have an Exact Match or Homozygous Reference Label\n",
    "# This step removes any row that has in 'Inaccurate Call' label\n",
    "df_train = df_train[(df_train['Label'] == 1) | (df_train['Label'] == 0)]\n",
    "df_train_2 = df_train_2[(df_train_2['Label'] == 1) | (df_train_2['Label'] == 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    163\n",
       "1    437\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# There are only Exact Match [1] and Homozygous Reference Labels [0]\n",
    "pd.value_counts(df_train['Label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='hom_ref'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>sample</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HG002</td>\n",
       "      <td>859</td>\n",
       "      <td>37568322</td>\n",
       "      <td>37568587</td>\n",
       "      <td>-115</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.37823</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom sample   id     start       end  Size       type    SVtype  refN_cnt  \\\n",
       "0     1  HG002  859  37568322  37568587  -115  Insertion  Deletion         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0               7611.37823              31.0                 8407.903226   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "\n",
       "[1 rows x 40 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Test Data\n",
    "# SVanalyzer generated training data\n",
    "df_test = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/test/final_df/tech_sep/DEL/HG002/PB_HG002_DEL_min1.csv')\n",
    "df_test_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/dataframes/test/final_df/tech_sep/DEL/HG002/PB_HG002_DEL_min1.csv')\n",
    "df_test.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_test.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Label'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(c) - set(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_train.drop(['Label'], axis=1, inplace = True)\n",
    "df_train.drop(['GTconswithoutpacbio.GT'], axis=1, inplace = True)\n",
    "df_train.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_train.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_train.drop(['sample'], axis=1, inplace = True)\n",
    "df_train.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_train.drop(['type'], axis=1, inplace = True)\n",
    "df_train.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>65326531</td>\n",
       "      <td>65326651</td>\n",
       "      <td>-120</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end  Size  refN_cnt  refN_pct  segdup_cnt  segdup_pct  \\\n",
       "0     1  65326531  65326651  -120         0         0           0         0.0   \n",
       "\n",
       "   tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0              1            1.0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            187.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       236.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           11.0               18112.46296   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7134.848382              54.0                 11277.83333   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4197.626206                              54.0  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_train['chrom'].replace('X', 23, inplace=True)\n",
    "df_train['chrom'].replace('Y', 24, inplace=True)\n",
    "df_test['chrom'].replace('X', 23, inplace=True)\n",
    "df_test['chrom'].replace('Y', 24, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GTconflict',\n",
       " 'GTconswithoutpacbio.GT',\n",
       " 'GTsupp',\n",
       " 'SVtype',\n",
       " 'id',\n",
       " 'sample',\n",
       " 'type'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(d) - set(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_test.drop(['GTconswithoutpacbio.GT'], axis=1, inplace = True)\n",
    "df_test.drop(['GTcons'], axis=1, inplace = True)\n",
    "df_test.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_test.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_test.drop(['sample'], axis=1, inplace = True)\n",
    "df_test.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_test.drop(['type'], axis=1, inplace = True)\n",
    "df_test.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Impute missing values using KNN\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>65326531</td>\n",
       "      <td>65326651</td>\n",
       "      <td>-120</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>83753489</td>\n",
       "      <td>83753698</td>\n",
       "      <td>-209</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>82.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18304.63415</td>\n",
       "      <td>4939.332831</td>\n",
       "      <td>41.0</td>\n",
       "      <td>11648.90244</td>\n",
       "      <td>2866.042175</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>152326749</td>\n",
       "      <td>152326980</td>\n",
       "      <td>-231</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>177.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16403.08333</td>\n",
       "      <td>7477.361174</td>\n",
       "      <td>36.0</td>\n",
       "      <td>10544.55556</td>\n",
       "      <td>4602.637067</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start        end  Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0     1   65326531   65326651  -120         0         0           0   \n",
       "1     1   83753489   83753698  -209         0         0           1   \n",
       "2     1  152326749  152326980  -231         0         0           1   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0              1            1.0                ...                  \n",
       "1         1.0              1            1.0                ...                  \n",
       "2         1.0              0            0.0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            187.0   \n",
       "1                                             82.0   \n",
       "2                                            177.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       236.0                             1.0   \n",
       "1                       160.0                             1.0   \n",
       "2                       277.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           11.0               18112.46296   \n",
       "1                            7.0               18304.63415   \n",
       "2                            4.0               16403.08333   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7134.848382              54.0                 11277.83333   \n",
       "1              4939.332831              41.0                 11648.90244   \n",
       "2              7477.361174              36.0                 10544.55556   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4197.626206                              54.0  \n",
       "1                2866.042175                              41.0  \n",
       "2                4602.637067                              36.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store training data in a new variable which will be converted to a matrix\n",
    "X = df_train\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/600 with 0 missing, elapsed time: 0.111\n",
      "Imputing row 101/600 with 0 missing, elapsed time: 0.111\n",
      "Imputing row 201/600 with 22 missing, elapsed time: 0.112\n",
      "Imputing row 301/600 with 0 missing, elapsed time: 0.155\n",
      "Imputing row 401/600 with 0 missing, elapsed time: 0.156\n",
      "Imputing row 501/600 with 0 missing, elapsed time: 0.156\n"
     ]
    }
   ],
   "source": [
    "# Convert dataframe to matrix\n",
    "X=X.as_matrix()\n",
    "\n",
    "#Imput missing values from three closest observations\n",
    "X_imputed=KNN(k=3).complete(X)\n",
    "X=pd.DataFrame(X_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>65326531.0</td>\n",
       "      <td>65326651.0</td>\n",
       "      <td>-120.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>187.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18112.46296</td>\n",
       "      <td>7134.848382</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11277.83333</td>\n",
       "      <td>4197.626206</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>83753489.0</td>\n",
       "      <td>83753698.0</td>\n",
       "      <td>-209.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>82.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18304.63415</td>\n",
       "      <td>4939.332831</td>\n",
       "      <td>41.0</td>\n",
       "      <td>11648.90244</td>\n",
       "      <td>2866.042175</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>152326749.0</td>\n",
       "      <td>152326980.0</td>\n",
       "      <td>-231.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>177.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16403.08333</td>\n",
       "      <td>7477.361174</td>\n",
       "      <td>36.0</td>\n",
       "      <td>10544.55556</td>\n",
       "      <td>4602.637067</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end   Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0   65326531.0   65326651.0 -120.0       0.0       0.0         0.0   \n",
       "1    1.0   83753489.0   83753698.0 -209.0       0.0       0.0         1.0   \n",
       "2    1.0  152326749.0  152326980.0 -231.0       0.0       0.0         1.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0            1.0            1.0                ...                  \n",
       "1         1.0            1.0            1.0                ...                  \n",
       "2         1.0            0.0            0.0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            187.0   \n",
       "1                                             82.0   \n",
       "2                                            177.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       236.0                             1.0   \n",
       "1                       160.0                             1.0   \n",
       "2                       277.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           11.0               18112.46296   \n",
       "1                            7.0               18304.63415   \n",
       "2                            4.0               16403.08333   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7134.848382              54.0                 11277.83333   \n",
       "1              4939.332831              41.0                 11648.90244   \n",
       "2              7477.361174              36.0                 10544.55556   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4197.626206                              54.0  \n",
       "1                2866.042175                              41.0  \n",
       "2                4602.637067                              36.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header values in a list, will be used later to re-label the matrix post KNN imputation\n",
    "dftrain_header = list(df_train.columns.values)\n",
    "X.columns = dftrain_header\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Store Labels in a new 'Y' DataFrame\n",
    "Y = pd.DataFrame()\n",
    "Y = X['GTcons']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    200\n",
       "1.0    200\n",
       "2.0    200\n",
       "dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Count the number of labels\n",
    "pd.value_counts(Y.values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Note: originally selected 1000 of each label --> find out why some are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Remove labels from feature set\n",
    "X.drop(['GTcons'],axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X4 = X.reindex_axis(sorted(X.columns), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Machine Learning\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='machine_learning'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the following section a random forest model will be trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](http://scikit-learn.org/0.16/modules/generated/sklearn.cross_validation.train_test_split.html) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data. The trained model will be used in a [later section](#predict) to predict the consensus GT for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "       * In the following section, svanalyzer data was used to train a random forest (RF) model. The features for the svanalyzer dataset include: svviz features, GA4GH features [RefN, Segmental Duplications, Tandem Repeat], preliminary R script analysis [consensus GT, GTsup].\n",
    "       * The RF classifier will predict the consensus GT labels:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "       \n",
    "       * In the [following section](#prediction_step), the trained RF model will be used to predict labels for genotype labels for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Train Random Forest Classifier **"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='train_test'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train Test Split\n",
    "# Train on 70% of the data and test on 30%\n",
    "X_train, X_test, y_train, y_test = train_test_split(X4, Y, test_size=0.3, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n",
       "            max_depth=None, max_features='auto', max_leaf_nodes=None,\n",
       "            min_impurity_split=1e-07, min_samples_leaf=1,\n",
       "            min_samples_split=2, min_weight_fraction_leaf=0.0,\n",
       "            n_estimators=10, n_jobs=1, oob_score=False, random_state=None,\n",
       "            verbose=0, warm_start=False)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%timeit\n",
    "# Train Random Forest Classifier\n",
    "model = RandomForestClassifier() \n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Feature Importance\n",
    "importances = model.feature_importances_\n",
    "indices = np.argsort(importances)[::-1]\n",
    "print(\"Feature ranking:\")\n",
    "\n",
    "for f in range(X_test.shape[1]):\n",
    "    print(\"%d. feature %d (%f)\" % (f + 1, indices[f], importances[indices[f]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Another Resource](https://stackoverflow.com/questions/37877542/how-to-label-the-feature-importance-with-forests-of-trees)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/ipykernel/__main__.py:3: FutureWarning:\n",
      "\n",
      "sort is deprecated, use sort_values(inplace=True) for INPLACE sorting\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x114929a20>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAHeCAYAAABg00vYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXmcZUV1+L8zDIvAoCwjghqJyBwhKi64IIigIRoVwS2o\nISSDgKjgAkYFRUVAjQniBrJHFhFBRcQFQXFhkeAOKBwc4YcgyDIS1rAM078/Tr3p27fvcuq++2Z6\n3pzv59Of7vdevbpVdW+fqjp1llkTExMEQRAE48Xs5d2AIAiCoH9CuAdBEIwhIdyDIAjGkBDuQRAE\nY0gI9yAIgjEkhHsQBMEYMmd5N2DA7bffM80mc9111+TOO+93fd9btu9yUefKWee49SfqXHGfj3nz\n5s6qKj+jV+5z5qzSe9m+y0WdK2ed49afqHO8ng+Y4cI9CIIg6EYI9yAIgjEkhHsQBMEYEsI9CIJg\nDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAZ46EaBEEQTGePT1447b2TPvCS1u/Fyj0IgmAM\nCeEeBEEwhoRwD4IgGENCuAdBEIwhrQeqIjIbOBrYEngQ2FNVFxY+fxPwbmAxcCXwdlVdIiK/Au5O\nxa5X1QV9Nz4IgiCoxmMtswuwhqpuLSIvAI4AdgYQkUcBhwFPV9X7ReQrwKtE5HxglqpuP6J2B0EQ\nBA141DLbAucBqOplwFaFzx4EXqiqgwjyc4AHsFX+miJyvohcmCaFIAiCYBkxa2JiWgKkKYjICcDX\nVfV76fWfgCer6uJSuf2AV6SfpwEvAE4ANgO+B0j5O0UWL35kIjcYfRAEwbiz0wHnTHvv3CN2Lr6s\nzMTkUcvcDcwtvJ5dFNJJJ/8pYD7wOlWdEJFrgYWqOgFcKyKLgI2AG+suUpM+ittvv8fRRH/ZvstF\nnStnnePWn6hz5j8fRYrl582bW1nGo5a5BFuNk9QrV5Y+PxZYA9iloJ7ZA9PNIyIbA+sAt/ibHgRB\nEAyDZ+V+NrCjiFyKLf8XiMibgbWBXwBvAS4CLhQRgM8CJwJfEpGLgQlgjyaVTBAEQdAvrcJdVZcA\n+5Tevqbwd93q/81dGxUEQRAMRzgxBUEQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0I\ngmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJg\nDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ\n4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEe\nBEEwhoRwD4IgGEPmtBUQkdnA0cCWwIPAnqq6sPD5m4B3A4uBK4G3p49qvxMEQRCMFs/KfRdgDVXd\nGvgAcMTgAxF5FHAYsIOqbgM8GnhV03eCIAiC0eMR7tsC5wGo6mXAVoXPHgReqKr3p9dzgAdavhME\nQRCMmFkTExONBUTkBODrqvq99PpPwJNVdXGp3H7AK9LP8Z7vFFm8+JGJOXNWGaYvQRAEY8dOB5wz\n7b1zj9i5+HJW1fdade7A3cDcwuvZRSGddPKfAuYDr1PVCRFp/E4Vd955/7T35s2by+233+Noor9s\n3+WizpWzznHrT9Q585+PIsXy8+bNrSzjUctcgq3GEZEXYIemRY4F1gB2Kahn2r4TBEEQjBDPyv1s\nYEcRuRRb/i8QkTcDawO/AN4CXARcKCIAn636zgjaHgRBENTQKtxVdQmwT+ntawp/163+y98JgiAI\nlhHhxBQEQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAM\nCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnh\nHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGDJneTcgCIJgZWOP\nT15Y+f5JH3hJb9eIlXsQBMEYEsI9CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQ\nBGNICPcgCIIxpNWJSURmA0cDWwIPAnuq6sJSmTWBC4C3qOo16b1fAXenIter6oI+Gx4EQRDU4/FQ\n3QVYQ1W3FpEXAEcAOw8+FJGtgGOAJxTeWwOYparb99vcIAiCwINHLbMtcB6Aql4GbFX6fHXgNcA1\nhfe2BNYUkfNF5MI0KQRBEATLiFkTExONBUTkBODrqvq99PpPwJNVdXGp3I+BfVT1GhF5OvAC4ARg\nM+B7gJS/U2Tx4kcm5sxZZZi+BEEQrBDsdMA5le+fe8TO096rKlsqN6uqLo9a5m5gbuH17CYhnbgW\nWKiqE8C1IrII2Ai4se4Ld955/7T35s2by+233+Noor9s3+WizpWzznHrT9S5/J6PIt7vFMvNmze3\nsoxHLXMJ8AqApF650vGdPTDdPCKyMbAOcIvje0EQBEEPeFbuZwM7isil2PJ/gYi8GVhbVY+r+c6J\nwJdE5GJgAtjDsdoPgiAIeqJVuKvqEmCf0tvXVJTbvvD3Q8Cbh21cEARB0I1wYgqCIBhDQrgHQRCM\nISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh\n3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyD\nIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAI\nxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAyZ01ZARGYDRwNbAg8Ce6rqwlKZNYELgLeo6jWe\n7wRBEASjw7Ny3wVYQ1W3Bj4AHFH8UES2An4KbOr9ThAEQTBaPMJ9W+A8AFW9DNiq9PnqwGuAazK+\nEwRBEIwQj3BfB7ir8PoREVmqzlHVS1T1xpzvBEEQBKPFI3DvBuYWXs9W1cV9f2fddddkzpxVpr0/\nb97citLVeMv2XS7qXDnrHLf+RJ3L7/kYRd0e4X4JsBNwpoi8ALhyFN+58877p703b95cbr/9Hsfl\n/GX7Lhd1rpx1jlt/os7l93wU8X6nWK5O0HuE+9nAjiJyKTALWCAibwbWVtXjvN9xtTgIgmAFZo9P\nXjjtvZM+8JLl0BKHcFfVJcA+pbevqSi3fct3giAIgmVEODEFQRCMISHcgyAIxpAQ7kEQBGNICPcg\nCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiC\nMSSEexAEwRgS2ZGCIAgaqArjC8svlK+XEO5BEKyUzKTY66Mg1DJBEARjSAj3IAiCMSTUMkEQjA0r\nqn58FMTKPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAE\nwRgSwj0IgmAMCeEeBEEwhkT4gSAIZjzjHsFxFMTKPQiCYAwJ4R4EQTCGhHAPgiAYQ1p17iIyGzga\n2BJ4ENhTVRcWPt8J+DCwGDhJVY9P7/8KuDsVu15VF/Tc9iAIgqAGz4HqLsAaqrq1iLwAOALYGUBE\nVgWOBJ4L3AdcIiLfAu4CZqnq9iNpdRAEQdCIRy2zLXAegKpeBmxV+GxzYKGq3qmqDwEXA9thq/w1\nReR8EbkwTQpBEATBMsIj3NfBVuIDHhGROTWf3QM8Grgf+C/gZcA+wJcL3wmCIAhGjEfg3g3MLbye\nraqLaz6bC/wvcC22op8ArhWRRcBGwI11F1l33TWZM2eVae/Pmze3onQ13rJ9l4s6V846x60/K1Kd\nM6ENM71Oj3C/BNgJODOpV64sfHY1sJmIrAfci6lk/gvYA3g68HYR2Rhb4d/SdJE777y/sgO3336P\no4n+sn2XizpXzjrHrT8rUp0Dcsp7y66IddYJeo9wPxvYUUQuBWYBC0TkzcDaqnqciOwPfB9T8Zyk\nqn8WkROBL4nIxcAEsEdhtR8EQRCMmFbhrqpLML15kWsKn58LnFv6zkPAm/toYBAE40uEFRgd4cQU\nBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRB\nMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBJ5TYMg6JWI0T4ziJV7\nEATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSScmIJg\nJSbH4Sick1YsYuUeBEEwhsTKPQjGkFhlB7FyD4IgGENi5R4EKwixGg9yiJV7EATBGBIr9yDIwLt6\nriqXUzZW5MGwhHAPxpZRCOIgWFFoFe4iMhs4GtgSeBDYU1UXFj7fCfgwsBg4SVWPb/tOsGIzCtvo\nEMRB0C+elfsuwBqqurWIvAA4AtgZQERWBY4EngvcB1wiIt8Ctqn7TjBzCfVAEIwPHuG+LXAegKpe\nJiJbFT7bHFioqncCiMjFwHbA1g3fGQtGoVNdniviIAjGC49wXwe4q/D6ERGZo6qLKz67B3h0y3eW\nOXG4FQTBysasiYmJxgIi8mngMlU9M72+SVWfkP5+BvBJVX1Fen0kcAnwwrrvBEEQBKPHY+d+CTAQ\n3i8Arix8djWwmYisJyKrYSqZn7V8JwiCIBgxnpX7wPLlGcAsYAHwbGBtVT2uYC0zG7OWOarqO6p6\nzei6EQRBEBRpFe5BEATBikeEHwiCIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNIBA4LZgQi8jeltx4G\n7lDVh5dHe+pYUdq5IiAiT1DVmwqvRVV1ebapihWlnWVmjLWMiKwCrAKcAeyKmVDOBr6rqpUuoiLy\nBVXdt/D6FFXdPbdczrVz2+ltY+HzdYAlwGuAbw9CO5TKfEhVDyu8/oSqHti1jek7s7AYQWsM3lPV\nn9aUPQh4H3B/qntCVTcufP7huv6p6sdq6rwCeAKgwHwsVtEc4H2qepqI3AJMAKsDawI3pvK3qeom\nNXU+EXhTqU/Trp/G6+9K5S7v0s5CuacBXwTWBU4DrlLVb9fU+XjgP4DHAmcBV6jq/1SUc90jEdkR\n2B8bq0G58nOcNZ5pjP4NeBJwYerPHaUyX1XVXav6WCr3NGDQ5/elt1cBPqGqzyyV/VFq5zRq/t88\nfd+urm3F8cxsp7fOrP6k77SOfRUzaeW+B3AQ8DjsH2cWJuQuKhcUkXcAHwLWE5HXprdnA7/rUi7n\n2t6yGdcufucM4NuYh+9s4LWYkB98/hZgT2BzEXlFensVYFXgwEJVOf0Z8HVMuNyYXk8AlcIdmzA2\nVtX7az6/Nf3eBbgec2p7LlBe9Ra5HniJqt4hIusCJwB7Ad8DTlPVjQBE5DTgQFW9UUQ2xgLX1XEW\n8INCn+r4LiYMBhPpBDb22e0slPss5hNyPHBi+rxSuAPHYcH1DsbG/GTgBRXlvPfoSODdNPS7w3ge\nC9wM7Aj8HDiF5KhYYF7d9UqsC7wR2BCbfMGez6Mryu6Tfn8E+Cb2LD0PeFVN3a19B96Wfm8KrIb1\n51nAvcD2HdvprTO3P+Ab++lMTEzMqJ/58+fvkVH2oJ7L5VzbVdZ77VT2p+n3j9LvH5Q+X33+/Pmb\nzJ8//7j58+c/Kf08cf78+av30J9LM8p+c/78+bMc5c4vvb6goexlpdcXp98/a2pn+XPv9UrlfpLR\nd287f5h+X1i8pzV1Xlj6/eNh7tH8+fO/2/W+141n4ZkctPGSijI3zJ8//+NVPzV1PjujnT+sGrMh\n+/6d+fPnz0l/rzJ//vzzemint05Xf7xjX/Uzk1buAy4QkffRspVOXCUih6jqR0TkPODTqnr+EOVy\nru0t6702wGpplf97EdkAmFv8UFUfBP6fiGyiqjfU1NG1P9eIyMaqerOj3tWAK0VkEFZiQlXfXFFu\nPRHZVFX/KCKCBZWr45ci8hUsfMXWwG9EZFcmdwEDfi8ipwKXYzucXzbUeZWIvBH4NWkrrKrXVpT7\nqYi8DAunQSr3pyHb+VcReSuwVmrD/za084F0/VVSuI4Hasp579FtInIMU/t9XE3ZqwvjuTX14zkn\nPZOIyFxsBVvmfmyn6GV9EfkuU5/P2ih9aec6uO8P1RTL6ftGhb/nYLuiYdvprdPbH/CN/fQveQot\nY7xbaYCPAjukv3fFtr5VgtNbLufa3rLeawN8KpU5AHgncGhNuTtFZGfsH2kJ1AqtnP5sC/xJRG5P\nr6fo0Uv8h6M+gHcBZ4vIhsBNwFvrCqrqO0Tk1VgY6VNV9btpQji3VHRvTFU1HzhDVc9puP4z08+A\nCaDqn3JD4DNMCuAJ7B9umHa+BVON3QFshanK6tgb+C9gA+C9TG7xy3jv0fXp9+MK/aljLybH86sN\n4/lBTIWwEXAZdm/L/EVVT264VhmPCmXAP6c2/BPwe6BqMQF5fT8R+J2IXIWduXyyh3Z66/T2B3xj\nP42ZKNzvUdUPOcs+rKp3AajqXSLyyJDlcq7tLeu9Nqr6jfRQPAPTw/65puhjsYdtQJ3QcvdHVed7\nyiWuBF6G6fpnARsDP6kot66qPmPwQkT+CfhFVYVpRbIGcAuwgYjsrqqnVBRdCztn+DPw6IZyqOoO\nxdcpuF0VT1XVzWs+69rOdwEHD6xoROSTwAdqqn2Pqr6x7dree6Sqh4jIRky9P+V+FA/dB9deRUQu\nrFmVPlFVRUTmYdZBVUKzaRdVxZ9U9QfOsnuq6n6DFyLyCaaeMwG+vhc4B1sAbQr8AVi/h3Z663T1\nJ+EZ+2nMROHu3UoDXC4ip2Nb5Oel7wxTLufa3rLeayMi+2KrqPWALwGbAfuWy2UILXd/kjpgAYV/\nClV9WU29Z2MqjKdjKoQpB6si8iosG9ebRGSwAp6NZeM6s6bOc7BDo+Jh4TDlSGqR/Qt9ehhboZa5\nIvW/OE5122Tv9d8JvFBEXq+q92H3vo4tROQxqtqkunHfIxE5EVOxrAU8CriO6Qe0xUP3QVC/CeoP\n3fcGvqyqt9d8jqq+d9CPpF5cO9V5hlabiraqUDKMCNx9r7GC2QpbZU+xgslop6vO3P4kWse+ipko\n3L1baVR1PxHZBfuHPVNVy1vjrHI51/aWzbg22ApqO+CHqvpZEfl5VaEMoZXTny9iaqHXYyvzugkD\nYJaq7iMiJ2EPalkg/BZbsfwfkzrYJdgqsY7Zqrpbw+e55QDegVkqfAhbTb27ptx2wCsLryeAJw95\n/auAzwPfT4KuabW1BbBIRO7AxqlO3eK9R1tiKoFjMQH+tXIBVT0eOF5E9lDVkxz9WV1Efo3dzwkq\nzlnSc/4hTKh9GFNBPgubQP6zos6yCqWK04Afpn4cnt5bAtxWU7617+RZwXjb6a0ztz/gGPsqZqJw\n/0dgc1X9dXpYvlNXUEQei+mzBXisiFys1XbhrnI51y6vnodtY2I26eal1w/WlPMKLXd/sO3eV0Tk\nH1T1oyJSpWYZsFhE1sBWRxOUniNVvRE4OR3UDfq1NaZbrOMKEXk+8BuaV8/ecgA3q+otIjJXVX8s\nIh+pKjRQHaV7tUhVa1VnOddP+vh7MSFX6w2uqk9quF4R7z1apKoTIrJWMtlsqvMXIrI1JmA+Dnxc\nVX9YUe79jvbth6nrAO5U1QNF5NHABVQL9/9uq7BgRPAeTIA+jK1kTwGqjApa+66qFwEXicizVfVX\nYKHNVbXuoNLTTledHfoDvrGfxkwMP3Aak6vN+ZjNbx1fxdQD78e2X6cOWc59bRG5XkSuK/zUqVu8\n1wY4HbNbfko6nf9mTbmbVfUWYK6q/ph6K5ScsVwiIn8HrJkOCNdrKHsU8B7sYPhGJlc2ZT6NHdh9\nDDsUOr6hzhdjK/trsBVKXfz/FwNfAW4Hrm0oB3BXmtQm0m5ng6pCIrK9iFwHfB/4o5gjzLDt/Aks\ndWDZE9tlVSIiTxeRn4vILSLyaxF5Vk1R7z36pYi8F7hZzHfiUQ39OQZbRHwIu0eVEyCmkngVpnLY\nheoEPLNVdVH6e9D/uyip7Qp8FRvLMzH77dMb2vk14DnYJPEwdiZVRU7fNxeRN4rIvwK3pO8N205v\nnd7+gG/spzETV+6PV9X/BlDVT4l5dNWiqsekP38rdmA3TLmcaz81/Z6F3aQ3DNtGbIXzQ+Bp9jW9\noqacS2iR15/9se3s57CHt3arrqpfBxCR9YCzVPXumqLPVdV3i8iPVHUHEalaEQ7q3LKhbUXeiU0u\nN2O7lv/XUHZP7GDrQMwCab+acocB26rqzWLeot/A7kV2O2XSVf10ERmoyu6j3ikKbMz3VNXfisgz\nsf5tU1HOdY9U9SARWRs7D/lHzNyujgcwx7rV1JLZ1+1aTsIE9pexCe5LwKtLZZYKUlX9aOH9VWra\nufXgbxF5DM0Cbk3gW8C7VHV3Efn7mjpz+v6uVOYMzMHufMxqaZh2uur09ifhGftpzEThPiEi81X1\nWhHZlJoHI3GNiPwz8CNMwC4a/EOVDg695dzXTturAZeInXYP00aAE1V1Wwr21jXsCTyFSaE17dC1\nQ39+JyIPYYe4u2Cmi5WIuVofneo7S0RuUNUTK4quIiLPwbahq1Gy2091fUFV9xWRn1HSS6tqlTni\noZiO/GuYzvISzPysivsx/e/fYKaKV9WUe0ST7biq/llEptmZZ7Rz//RzbCq3HvAIljC+7rxjlqr+\nNtX1GxGpTCTvvUcyPZzBJky3wx8wgakEvpsWHnUxctZX1c+nv38jIq+vKPMzEdlXVb9QaMs+mDFB\nG3dRf84Bdr7wLmxlvgWmEpxGZt8H9/keVX1QRDzysK2d3jpd/Ul4xn4aM1G4vxv4qpht9M002EZj\nq+enYsJuwOCf6iUdyrmvnYT54J98Y+odC7zXBrhPLMl40X69apXwCZ2MV3OAiJxCtSliTn9cljqJ\nwzAB+3VMT1snYE/BJoE9sIPAYyvKDGz5W00BE0tU9a8igqo+ICL3NJT1um3fLSL7YSqx7YC/ZrRz\n9dLr05KK7vnYVvoYzH7+kIZ2PiJmYXRRun7lWUvGPfKGMwDzq3gedi6wPal/IvIkneoo9ygReZyq\n/iU9T1ULhQ8CJ4nIHpgK8m/T78pYSoWJchYWuqByt5Q4AJvQDgd2o97WO6fvf8Tsxt8jdh5TuVPO\nbKerzoz+gG/spzHjhLtawKZpOkcR+YiqHlIqW3moKSIf7VjOfW2m6lp/C5xXdQ3vtROXpt8b1nyn\nKl7NLGoOKjP747LUSQwE7ESTgFXVo5m0Flh66Fu8vqoOVlWrY/rH+Zia4ICaay9ME+v6IvIB6g+h\nADZV1T1FZFtVPTeVr2I3bFwPx8ZymsNRoZ27quqnUj+ehk0Yzy4U/U/gX1X1IRE5DHg5sBATnt+q\nuf4e2Nb9k+n6e9WU896jR6nqhWIB5rRqJ1Lo1x1YbB2w3eWA/2bq4uNg4FIRuQtYBzsELNd1P/DG\nJIA2AW5S1TpfjUF/BjxQGOOqdl7K5P9HcWdwtqq+plA0p+8LRGRtVb1XRH6hqn9Jde6sU525ctrp\nqjOjP+AY+ypm4oFqHS/OKFsboa1juaprfxmz430eZiL1f866aq+dBN4vUl2/qZjMjlIL+HQ4thrd\nARMadYKwjqr+eC11IE/Aeq9/CrbKfQEmWL5U89190vUuxnTZdYIQ/G7bX1TVf1fVV6nq+7Temgng\naSKyTzooO53pevxVVPUKsSBca6nqr9KZRJMp5L3Acar6NGy1eVdNOe898oYzaGJW8YWqXoCt7HcE\n3qTVFjWDsreq6v8MBHvalVTxCDapfRc4RkQ26dDOx5ReZ/VdVe9Nv/9SeLu8is5qp7POOsr9yRr7\nIiuScJ/VXiS77DDljsV0bxdgq5QTnHXV1pkE5gIszsS/ikjVQQzAS7Ht4eHYgU1TZETXtfFb6sBU\nAXsvU1VOXa9/n6p+T1XvUtXvUCOIVXWxqh6jqu9Q1c9rvRkk2Gr8EkzvfhlmtVPF6iLyDBFZQ0RW\nk3qnMLDQqy/GVuTPVdVLSp8PdNYvx0I/ICKrYguBOs5gUr3zV6ZGlyzivUd7Y89RWziDJqZMRmJO\nPP+UVvr/LCKfzahrs5r3j8esx7bB1Cd1ZyfudtJP38vPZx/t9MqaaYuArmM/49QyDeQEnveWHabc\nZqo6WH1/U0QurSiTW+d2qroNQLqBl9V8dxC+90OqeoaINK1eXddW1S+IWbM8DbhGVZvMrdbCdNkD\n3fRrqPc8dV0fuFFEPoTFq34O8KCI/ENqW10snkZU9SdYboV52uzdNx/zPC22b8qhWekgdVXMWeZH\nSfdfPFD9gYhcAjwReHU6yP4CZk5Xx1qaYr2r6ukiUjlZeu+Rqt4kIntTCHTVA89W1X1S/e8Skbpw\n0EsRc5dfpKp1i481VHWgqvqmmO33UPTU9/Lz2Uc7h0mckT32sGIJ95nGGiKypqreLyKPwnnI0cKq\nMun4MIv6B2JV7IDypyKyA83epC5E5HmYbnENYIcktN5eU/x8TDdcDLSVI9yrmMDMFp+M9f1WzNNv\ngvpAa42ImYm+FbtXAKjqFuVyqvp0R3WuA19V/Q8R+RZwl5pp5aaYyuXshq89JGZbfxmm5qvctXjv\nUTpg3wZT7wyeo2eXy7VQtbNcX1UXiZkD1sqO9EyeCNwNPEZE9kqqhTJzROTpqnqliHjuQSs99b1M\n7+3MxTv2RVYk4T7T1DKfxezWr8Lcx+ucP3Lq/CpmVnkZZm1Rt9pbgOnfTsTitfxrD9c+GTMha9I3\nD7hLVRdkXrPx+ukg6lmYJ+/vWnYOXt6FWcc09imtlN+N2R4P2jNl5T6wHJGK7E6U1D2qWgwd/EfM\ngqKJPTGd7uewSbPOqsl7j0RVN20pM/ULIuupatFK6MJSkY9h3qx3Yk5zdRM/mGWRx2/gnZh1zUbY\nTtB1UFhEl2SdAAAgAElEQVSiPBbZfa+g/Hz20U6vrKm6tzljv5QZI9xlem7KpajF1t69S9lU/lHY\nP4xglhjHqgUy6lQuXefLInIRZk97m9bE/86s8wgR+T5mOnlSw7b7D1jEOWhYMedcG/iDqn6prq4S\n3xezX15qpaPV6d7q0oNVpUI8FLPOuBx4Z7IaqHJZz+EK4EZtDicAppd9JfCXlnKQF0bZhaouBHZJ\n1jcPpgmhCu89ulzEl+dTRF6MOU2tIiJnATeo6omqOiXctKp+W0S+B8wrHRRW0eo3kPg9sLdOhsdo\nylJWmYpQVV9XKurue6p3M+xM4Argz2oRFz/dtZ3eOjP6kzv2S5kxwh1bpQ5sSUl/PwWbqVZXi1fS\npSyYu7pi5orbYNYYuw1RDjEb1tXVPOLOSmZPVXHOc+rcC5ivqv8uIueLyKmq2hSuoA33tYGvi7lr\nFwV23QHki7ADwIHVS126t0o785rr/yPwPFVdkiaFn1EdjySHC4HrROSPTOZ6rXIkukN9yU8gLyx0\nI0kVcyKmjtoDcy+/XUROUNWqA3rvPboL+LlYXJtpOW5LDJzCKn0WRORvMcH0Biw+0Fmp3t1Ute5M\nyOM3AGZx9h3MvX4+Ftu8LiCW137d3XeZ6jdwMiZD9tXpwf3c7cyos7U/Hcd+KTNGuOtUF9/VsK3I\n2tg/feeyifVVdRB855y04h6mHMCrVfU5qT1vSIdoVcI9p863MRka9pXYTR9GuOdc+x3YP3hj2NnE\n2qra5C49wGtnDuZtORf751yVeq/CHN6K/SNW9klEPp7+XC3tmH7FZDCwg2rqzAkL3caHsQnt4TQ2\nO2I7gh9TbX3lvUcvAdZT1UpP1xJtPgtfwHZ8i0Xk08C/YJPLl5maG7RIq99AIic8htd+PafvRb+B\nz0i930BOO711evrTZeyXMmOE+wAR2RKzcf4BZmpWa2+dUfZ3IrKNql6SDkRuEDNPm6VTTem85cCC\nOK2m5qyyKvVmpTl1PjJ4KNM//DAn7LnXXlSz86jCK+By0oNtDFwrIr/FzjAekmSBpNVhCDzcBPxc\n66P9Dbbud2O+Bf+LrV6PaKgzJ4xyGw+reR0+Of29EEDq47t479G1mCNckwPRgDafhbVV9Vsisj6W\nNOKC1MZpz7tMxtXZEDMfHBxozqNal5wTasRrv57Td6/fQE47+/RFcI99FTNGuKcGH4Rtd/ZW1Yv7\nKJt4EfAyEXmYyeh81zLd5M1bDszh5iqxPKJPpT71XE6dg9X15Zg5YJ1Ho5eca98hIscydfVaFyBp\ny/QzoE7A5aQHqw28NgSrM3noPejT0u20ppRwaXX1RrVcrxdhC4ay3nXwHW+iFA8TYrFHXolFpEQs\n6NWaNeW992gbLJ7PHYPrNKhl9sEOdAdOYWUzzIHQeSnpkFVEZlEdibQcVwcmBXzV85ETasSbijCn\n71/BdsdPkma/gZx2euv09Cdn7Kcxa2Ji2MVhP4jI/2AHb/8JTNkalh/gnLKjRMyO98nAH9NBYZXr\n8qDs+sBftSVFllhUQAGu1vqokLntbL22VMQ61+khCtquMyWsgYj8s9rBc2t6MBF5Cibgi1mGmv6J\nPO2Z5gmrqj+RUtwUEbm0uDsQkR+r6vY1dU5LlKJ5KQqLde2O6VxXxYTfmpgD0+e1IhDbsPeo6tkU\ny1y0v6ZUkCLyJVX9t8LnR2CT81aYN/C1mBr0XlWttfdOO7ZNgIXakmGq4rvTwmOIyBGqmuuJXfx+\n3f/l5qQorFg6PXdbq9rprdPTn65jP2DGrNyZjG+xFs3efLllSTqyciS/aSsJb7nCZ7djccWLvIuC\nQ4z4IygiIs/AwqbeABwpInWJE1zkXLtOSEh1rIs6ysI0Jz3Y6Vj6vm2x1VHrfW1DzYmpinLclBuS\n/n2QCrFpS+9NlOJp3yki8k3MQuZBMVO7BapamRugh3s05dlM/APwHbF0gH/BFk1F3ot5235aVX+R\n1Hu/x0yBKxGRBVj+gquBpyYhmOMHURWewpWKsIFpfRc7aH+tqp6VXl9InoptWjsz6vT0J3vsi8yk\n8AMvTA/vLFU9pPgzZFmwrefbMPvQE6lP5Ost10TZnnUQQfEvmD63yUZ1kDjhgzQnTvCSc+06psW6\naKDc99XFkk+cISJfEcslW8e9qvoJLNjUv1ETPK0nyu1cgKU5ewU2WdcdAII/UUorYlZBDwBfSWch\ni4Crk0DIwXuPqmytF2KC71wRESyOylJUdUJVv4cJGVT1SjWP08On1TTJ24BnpglnK/JjH1W1c5CK\n8FaxpCY391Dn/cBRYk5XdWVGVWdrfzqO/VJm0sp9AzE72xelh2wpOj1fYE5ZVKfYvF4jlqR2Gt5y\nLZRVD64Iiglv4gQvOdeuY5iwDznpwSZE5HHAXBFx7ciGoLw7ewD4jPO73kQpHopJqhUTBIPQEjkM\nFW5DVX+ZVETTMhdJt4TOi5iMsTM4qM6hKjxGeUeRS1Xfbwdeh4UU2ADwWNj0UqenPx3HfikzSbi/\nFHgGZhdaFfe7a1nEYk0M2IgaweEtl0lOBEVv4oRRXHsU/BrTKW+B6QsPbSh7CGYffCrt6QiXJ+VE\nKXXZnVrR/CTVo+CrqS1Xp0nrqNLn7oTOIvIVJq1jfinmaf1s8iOmTiOpJE4CnoDtRPeoU19lMEtV\n7xSRV2LJX54xZH3uOp396ZJMeykzRrgn3dNPxeJnbIc5dlyGCYXOZRMbFf5+ALN9HqZcE+Vt2Nux\nFZonRK03cYKXnGv3Qbnv7vRgah6uA0eoYa2E2sjdfiMpiFmBeZiFy9BxfYALROR9FEIaaL0D2TAs\n7beIbKWqv8CcvIp9+1zxCzqZ0HkfTMUyaOPfMt1x7Rim85Vh2llqlycVYU6dB4KF6BWRV2NhBpYi\nyfyxzzoLtPYnc+ynMWOEe4FDsNlsc0z/fCAWyyO7rEyGKfjvpgt6y1V87+mk5BKqOkjeUTah+7aq\nlgVDJepPnODFfe0Gptkniz+sgDs9mIgcDryFgi18gwlbFtIeN8XD4LkqekYPXncKbFZg2JAGVfeo\nzQX+pVjugPL/Vl1/voa5yt9YKDdFwAwOsEVkJ2ArVf2IiJxHTUhqSakLC69PUdXdqQ6P4UpFmOpp\n7LuI7KnmAbxzEsB1nAo8v+HAuhgSxVtndn9wjH0VM1G4b6uq24klVT5ZRJriMbeVHQTeWh/zfrwS\nSzB8K1MjxXnLLUUsPO0gAe8BInKmqn5Gp7sZ35lu9rVMps7L9WjMXmnmXlsyYl3gDyuQkx7slcCT\ntMFpLRdxxk1xsrGqvkxqzN+GxBXSwHuPxOECr8kZSlMAOBH5O+AhtbhFVTxO/c5kh2CJZMB2o98j\n2fGna5Uzis1KP79Lbaqa5LqkIqxz/x/Ufw3NXCcitwGPTgeeS8OdqOrGpXZ668zqTyJn7Jcyk6xl\nBswRkTWwA6tVKJ3e55RV1a3VQhX8DovZ8g/YSvumLuVKvBJ4kZq96XbUh4R9LPAe4IuYUKzaurbR\n1Rkh59rHYWqUVbFVQZO51aaq+mHg/9I/TZ3FyCA92K+xlGIHN9T5a/qNPw6TcVOGsRYaMDjEf5uI\nnF786aGdV4nIG8WYLymBegXee/RGbOL9X1X9DBZhdAoisqOI/ElEVk0Hw98CTm0wIrhGLLuUh4c1\n2c2n3+X/y0FGsUOSkNxIVR+nqi9tqHMPLPrpJZgbflMqwsa+q+r3RWRLNSe2r2Bna6tSOudR1Tep\n6mOBEwrt3KhqR+mts0N/IG/slzITV+6fwUwQ5wH/Q3OWIW/ZJ6jqPQCqep+YPfEw5cBW9WtimYhW\nY7q9O6meHareXxZkXtudexJnWAFVvUBEtsfMw56sltO1jquAW0TkL0wGfGrKMu+hD2uhAVmH+Jl4\nQxp475HHBb4prk2VL8S2wJ/EPD8naPb8vDxNegO/gSkHhQXDhTtKRgxNToiDVIQXpNV551SEIrI/\nsKuIbIM5Qj6J5FtCtRf1gWJRSx8PfBvbMS0csk5vfyBv7JcyE4X7vtjBwmbA9UmXO2zZ80XkJ5iO\n8XnUuwS3lpPJjDyPBf4gk7FQFlVVKJYk+S0UVt8ddMmd1DKZ187JPekKKyCWHmyhqv6XiBws5rFa\nF4JgV+ygqKuTShW9WQsVD/HVwiZPoUEv66l7BxF5NObR+UdNOTgr8N4jjwt8VlwbzfDCVdX9xCxv\nBDhLJ7MYDahbNDXtUM9gcqcySEVYlZvV0/c3AC9M13szllXtf6U+m9qJTBo5/CW9Ljsw5dbp7U/W\n2BeZicJ9AjtAVCw4V1OEPldZVf2giDwHU7WcrDVu/c5yrow8BV4FbJKjS+7pADD32t7YHWBBjETa\nwwrkpAe7Acuj2pvOnfa4KdlUCfZEjrPXFETkdZgOeg5wZtppHFZR1HWPVPXzIvIDCi7wFcWy4tok\nnfwxwLqYILpKU2rAirLrYfbytwDrisiBag5qg/YVQ1RsRCHkRFV9CW8qQk/f71HVR0Tk2cB1Oukl\nWreIWl9VTxKR3VT1UqkO3JVbp6s/kDf2RWaicM+x93WVFZFiuR3FAmndCBylhUz3nnI6mZGnmPJs\nQJVOd6BLbhVaPR8AZl0beI+qeicud1gB8acHeyLwRxG5Lr2e6HKIVOJYLG7KMaktX8KsfEbBMEGa\n9sdieZ+HeRX/Iv0u47pH4nOBPxkLD7Aq8BKxRCGnAZ+nms9hnrzHM7mSrRMwZ6e6n47tLu6vaeeJ\nWJzytbDJ4DqqY7SDPxWhp+8T6VxjAcnsVszCpskC56np9xNqyuXW6epPImfslzLjhHs6kOi77KOw\nVGcXYQ/PczFHgJOZanftLQf+lGc5uuTGxAkdyLl2TuyO1cUOSZVJHWBV8oKc9GC7Oq6bS1vclJnC\nI2qxZSZUdUJE7qsp571HAxf4g1X1R1SnNcyKa5O+szC18faW84tZqrpPWiztSb3H7ZaYVdqxmKPO\n1xrq9KYibO07tks6FVOxHJgWVadRH5n0nZiGYPPUxqrnOLdOb3+ArLFfyowT7iNinqoO7Hm/LyLn\nq+rBFWoCbznwpzzL0SX3eQCYe+1BrIs7sFVE06GNK6yAOtKDyaR98D5MX/3WqeO8LMTUF+eKyG40\nW14tTy4W8+58QjqnqEvw4L1HrS7wMhnX5gwR2RU7M7pTRC7U6mB5fxWzqllLLJZ/0zO1WMyKbS3s\nntbJmUVpMltLVe+QqZFEpqD+VIStfVfVn2P2609Qcza6DIvuWrfwebkWEgTVtC+rzoz+QN7YL2Vl\nEe7riMhTVfWatL2aKxYGtxxewFsO/CnPcnTJfYcLcF9b82J3NIYVkLz0YHX2wb3EotaGuCk940ks\nXolaqsaXY3Har67Tp2bcI48LfFVcm0ew84kq3pLK34F5SzbFXToKm1SvxMyJ61buvxSR9wI3p/+l\nafdH8lMRtvY9CdSNgU+JeQaDWdl8kqlWSwNeISJHakMuXm+dHfoDeWO/lJVFuO8LfDltPW/Ewrbu\nyvToat5y4E95lqNLLh4A3svwB4Dua0te7I62sALu9GCqOnBuea6WvBUx56hhaIubko2IPBHz6pwS\nKkCrnb28df4tdog/C1O9bKGqn6oo571HrS7wOhnX5u2qerSjmfdhYZkH/X4K5sBXxQT27N6JnfdU\nen6nSW0uFntm4BBYJjcVocf9f13sHm6InZsNArbVjcMG2AR0PZNqyPL/kbfO3P5A3tgvZaUQ7qp6\nuYi8hKmmZr/oWi7hTXmWo0teC/P8HFjKvAbIiYM9zLVzYne0hRXISc1W9lYE+8f4fbmsF3HGTenI\nsKECqjgH+Abtq//GeyT5LvBgZnse4f5dzJ/jTiazK722puzBmAC7Tcwz+VwKHqpFNPmVYKqzVzH9\noNBlspnTd1W9CLhIRD5cs9sus1NbgYw6c1MrQt7YL2WlEO7iNDXzlkt4U54tpuQyTr265XxMqA12\nAxMMJ9xzrp0T66ItrIA7PZiqHoUdgB2kqh8vf96R3LgpObhCBWRyo6p+1FGu7R7lusAD3CciR5LM\niVPdVc/xGqpalUSjikWqeluq61YRudv5vc0q3vOabHbp+0uxQ/9GtBS0L01Cdf9HbXXmplaEvLFf\nykoh3PGbmnnLgR3WgeksoV5HfByWcPlgzLniZOrNve7SFOujJ3KunRPrYhBW4C5gHcw0sshVYh6K\nWwF7JTXXx6iw15du3oqNaH7clBy8ycFzOFdEPsnU85sqlVTjPdKCC7xYbte9Upkmk+GBk01bcpSf\nijlQXV24XpUNOcA9IvJ9THX3HGBNsUxXlX4oYv4Si9QSUZSpM9ksR6/s0vfVZdLqazCxVVl9lama\nhLx1uvpTImfsl7KyCHevqZm3HKp6iPgcMHLc+r8vFt6z+E/eGv2tgZxr74GZZg2ETG2sC20PK1CX\nHqzqAW4K8dCJhkOr47UmzWAG3lABObwR+8fdvFBnFY33SPJd4AfP8d9jVh1NYbM3xMJ9FHeVdWdH\nRa/Q2pSFYtmKTgTuBh4jInsNVHiF9rlMNrv0nbxkMm2TkKtOb39K5Iz9UlYW4X5xWkm2mZp5y+U4\nYOS49b8IWJ1J12ZXaM8Gcq7tjnUhLWEF1DxWv5fKPjbV9XUsNseUFYcmb0WZDLvcB7lxU9yohQpY\nH5s4rtPm8BheHlTVJo/gAW33KNcFnrSi9oTYfqqqbl7x/jTU739yKBbZ9WaxiJffAC4oF1LVu0Vk\nrojsTDpUFJGnl3Y32X3HVKrvxxZm38bUltPwTEI5dTr7U8Q99kVWFuF+NLALtjpagNnBDlMO/A4Y\nOW79a6vq3zd8nkvOtd2xLnCGFRCRo7G8pINwqU0rjq+mz2djtvl/wAImdaHLoZULEXkDpqq7Gnia\niHxUVU8bstobRORApp7fVJ0NtN2jXBd48IfYviItEIrqqIc8nWvgEVW9OdX155ad5TnYc1SMaV6k\nS99PwhYhL6Y+Zgw4J6HMOtv6U6TT2K8swv3LwEcx88WDMBvsqoiJ3nLgdMBQ1Zvwx6PpVZ+beW13\nrAtwhxV4Hqa2aXKtHrR1qZNIqrOTvj3R5dDKy/7Ac9RM7eZi5wjDCvdVMVPIQYCouoPftnuU7VaP\nP8T2dth4Lr0W9U4/Xu4Wkf2w3el2TFqJVTFbVXdr+LxL3z0xYyBvEvLW2dafIp3GfmUR7kuwB+iD\nqnqGiNTpk73loMUBQ0RuwW7C6phQuRFTS9yuqpvU1Lll+hnQSZ/b8do5sS68YQUWYtvOytgiDdzF\ncIIjN25KDks0RW1U1Xta/tFdZByit92jXBd4MJ10a9hsVe0jv2iZ3bA2H46dIezRUPYKEXk+8Buq\nV69d+u6JGQN5k5C3zrb+LKXr2K8swn1V4FPYqfMO1Oe99JZrdcBQS0aAiJwGHKiqN4oF3K89jNFS\n/PV06p9Nl2uTEetCHWEFEn+DqRwGlkVNTlSDUMqzMEHzg4Z6G+l4aOXlOhE5gsl/9Ca3cRdJJfN+\nbBIcxACqOqBvvEea71aPqp4lFkXxKZg6oy509Y5Y4pei81ang+TUvpuwg8LjmVTZzaPe1v/FTLU3\nn7J67dJ3fDFjIG8S8tbZ2J8iXcd+ZRHuC7BDtROBnbEMKMOUA1wOGGBqiRtT+ZubDg7F4kfsz6QF\nzsNMbtW74L62OmJdSF5YAajPfVtFUX30gKremvHdcjtz46bksAATqjti/+gfGLI+sL5vrKqNO5y2\neyT5bvWIWcrMwXwVzhALuFWVXepI4N3047y1f/o5lkld80DAV94fVd2y6v0BXfquqldhzzEi8kQt\npffrMgm11entT4lOY79SCHc1++aBjXOtU5C3XA11tq+/F5FTsZX9C7EtcB3vwNzzP4Q5Hb07sw3Z\n15a8WBeusAJi8XHqDojqgoEN3LYH1gOoate0eLlxU3JYHTsM+zpmirgxw8cAuh7bAVaScY9y3erB\nVqNvZtLb9UzM1b3Mn1S1826qiKrun37vIBbcaxPM+mpaKI/Cjq6qnuIuMLvvIvLvmHnhY4AFInLe\noG2J7Emorc6M/hTpNPYrhXAfJdJu+7o3FkZgPnCGqp7TUN3NqnqLiMxV1R+LyEeGbJ7n2jmxLrxh\nBXK8BAd4Qyi3ovlxU3L4GpaT9vXYxHYc8LIh61wNuFJErkyvyyGUXfdI893qwVRBtwKLk4VR3aR8\nm5gJbPGwf5hDb0RkAaaOuhp4qljy8fKiymUQ0LHvr8NUa+ep6hZisd+LdbonIW+d5Cf7gY5jH8K9\nI+K3fV0L2/L+GcuivrvW27PeJRbgaiKpaDYYspmea+eYDbrCCmiyc04WK8/Fl2nHG0I5B2/clBzW\nxGKlvFtVd09qjWFpi1GUa9rpcqtP3I15ZB8nFufntppy16ffbR7ZObwNeKZaeOu1sGdqinDXyeQ4\nT8HUgcVnqepcKKfvj2D9GagA67JQeSYhV50d+gMdxz6Ee3e8tq859qx7YgdbBwIHAPsN2UbPtXPM\nBt1hBRJnYw/v47FJ5mYsx2UV3hDKOXjjpuSwGubx+EsR2QKbQIelbAf9sFj0ya+qpfXLNe1cXfxu\n9f8EbKqqv0966+Nryv2ooo0DnXRXFmHnSmBqqaYV8enY87Qt9hxVheGGvL7/OP3slp6T79SUa52E\nOtTp7Q90HPsQ7t3x2r7m2LMuwZINPIKZA17fXLwVz7VzYl3khBUA2EBVtxaRE7CJqs7xA/whlHPw\nxk3J4b3YYfvhmBVFnWt7Dltiwm2QAeyJWP7Rl2HnGrnxSHLc6h8LvEqmRvasmlQPw1aOvwSeBTwE\nrCEW0uE/M66HWGKSwcHkL5Nly7NpOHcA7lXVT4jIZqq6h4jUxYh3911VP4gle0dEfq71+XHdk1BG\nnd7+QMexD+HeHa/tq9ueFdPnHoPp7frQ57ZeO8dsUDPCCiQG1h9rqer/NehzwR9C2Y3646bk1HmJ\nWHz8dTD1TJOqyctjdDIe/LFiGcD+RUQuTtfMNe10udUnvCGM7weekVavq2P3/bXY858l3LFnvEzd\njm7AhIg8Dkugsxb1K11336VkYpgO8V9S+Dx7Emqrs0N/oOPYh3Dvjtf21W3Pim2zvwW8qyd9ruva\nmhnrQvxhBb4hIgcDv03/GPc2tNUbQtmN+OOm5NRZjCm0JmbnXhdp08tjRGQDNU/n9bHzkVUpqF0y\n75HXBR78IYznqeoDqS0PpvY+VHGY3oqq/iS1fydgK1X9iIicR7MfxiFYaJBTsThOp9aUy+l7m4lh\nl0nIa7bo7Q90HPsQ7plIpu1rpj1rr/rczGvnnA24wgqo6lEiMkstTMN3SGGSRWTnCsudcgjlPvDG\nTckhJ6mzl48A/yMW93xtTIV1ANODnHnvkdcFHvwhL76ZdhKXY4fk30rjeVV792o5hMnwHrtiQrku\nqcdPmQyi963B++lw85BC0Zy+N5oYdpyEXGaLGf2BjmMfwj2fLNtXsaww72DyVHx9rXcnPgCbzXvR\n52ZeO+dswB1WIKlyUNUrC2+/CxNUxXLlBxoAETlbVV/jbFcZb9yUHNxJnb2oefx+F1sg3JbG7LyK\nou57JD4XeHCGMFbVQ0XkHGwXdJKqXiVmBly1uvXysKreleq/q8H6p4lpq/KMvntNDN2TUEaddUzr\nT9exD+GeSQfb18MwE6d9sFPvHRvqvpTJQ8AvDN4fQsC5r03e2YA7rEANTZH6yjwmo2wZV9yUTFqT\nOnsRkS+o6r5ScmxJetqq8fTeI68L/LSQFxVt3FNVT5CpjmlbpjbWOaR5uVzM+upn2G6wS3iI8rPk\n7jt+E8OcSWhYk9Gl/Rl27EO4d0T8tq+3qOrPRGQfVf2SiPxbh8t1FXA51845GxhKb03eA9/Znlqd\ncVMy6zxIzAzxAeqTOns5NP32OrZ4z1BaXeBF5Guq+nqZDDJX/H7xkLhL+joXqrqfmF+HAGep6rfa\nvlNBue2evg9Uq2368wGtk1CHOuso9meosQ/h3h2v7euDIrIdsKpY4oz1O1yrq4BzX9ujn5duYQWW\nG+KPm5JT5+OZmpd2EyYdVnJ5V4NaZ9p4es9QpN2tHlV9ffpdmQmrcC4yIZZk/BbPtXMQkfWwnc8t\nwLoicqCqfmLIOlv7TqZq1TkJZYcqcDDU2Idw747X9vVybJwPw1Zqy3LMW69dVgkUqVAN9LV6y1HL\nDIM3bkoOOXlp21BPocx7BO0u8B4G5yJ1u7S6mPM5nI3tfJ+O7YRyQ0PD9Gepte9F1WpVheVDTc8k\n5K1TRObXHFhX9WeosQ/hnok4bV9F5C2Yx+nmTHpdrlIuN6I25lzbHetCO4QVEHN0mg/8TlUHk8On\nvddkuFgz3rgpOeTkpW0kYzxz45G43OpbmJXaWBlrXszWflhmqeo+InIS9rzWOvIMzicKr09R1d2B\n3UtF++h7+VCzj0loUOepWGjiunO0pf0ZduxDuOfjtX09Dfghtr0+PL23hPrYHU3kCjj3tbVbrAtX\nWAER+RCTeukDRORMVf2Mqp5bUfaJ2EqlGLP6Yzrp3NMFb9yUHHLy0nppHM8O9+jH+Fzgm5gyEYrI\nxzBV5GqYwLwWMwkdhsXJmmmtdL2qneU7MH+S9UTktVi/ZwG/A6jQqf+Y4fte3g24JyFHndeJyG2Y\nL8PNhfcnVHXjmjOCTmMfwj0Tr+2rqj4I/D8sMqOLvgRcl2uTF+vCG1bglcA2qrokrU4vxrK4V+H1\nlMzBGzclh5y8tF684+m6R+p3gc/h1ZhD2JHYzquPgGxHYWN4JXATFUJTVY8CjhKRg1T1420V9tT3\n8g6vdRLy1qmqb0ptO0pV3+H8bqexD+HenRzbVy+jEHBecmJdeMMK3IqtNO7FVh23N9Tp9ZTMwRs3\nJYf3qGqXsK1NeMfTdY/E7wLfRHn1eouad+RcVV0oHbOElZjAVsJ3Yh7E/10uICKDBcodhb+Bavvx\nnvpepnUS6sCBInIotlv7NnCFpmifFXQa+xDu3enDAaPMKAScl5xYF41hBQoHgI8F/iAivwW2wA6h\n65Bnab0AAB41SURBVOg1OXhiFJPlFiLymAa/hi54wzR471FW5h7nuchNIrIHFmnzEwznfzDgYCxO\n/W0isiEWq6e8QKrTL9dNgK19zzzUHFyrcRJyUK7zRGxBuD3tYRI6jX0I9+704YBRZhQCzos71oW2\nhxXosrJ1eUpmMorJcgtsFXkH1sa6fKduHOM5wHuP3Jl7Ms5F3opFqzwL+DfMCgkRedLgTKADi1T1\nNgBVvVUs/MIUSlYrG9F+iO/pu/tQM+GZhAZt9B785oRJ6DT2syYm+jAgWDkp2L5e3dEBo1xfOW7z\nRA9byqEom4W1lJ2So1REnkchdR7QmDpPLGjWppjD0R3dW720viMxz9SRT5YVgriPOl05XytM976E\nHfS2usCnXdaUcxFVdZt2ettY892zMbXdT4DnYKv0H6f2HlQqWwzY9ijsGZnWTk/fk8XbS7EkM4so\nHWpW1PkDVf37utfpvaUHv4U6Z2G7oZdW1Hkh5j17NCb4T60zo6yjbexj5d4RGYEDhlpIg14FXA/U\nbRWrKG893anzROQNmD3+1cDTROSjqnpaxrWrGMVuoI5p8XJ6wOsPUL5HOS7wOeciVQzjs/DNwt9/\nbinrDdjW2vcOh5r3iMj3mZyE1hSLOLp0Eso9+CUvTEIdjWMfwr07fdi+TmFEAm5Ycv55y/9IOanz\n9geeo6r3ishczON3qL7nroSGZBSOWd5t9SzIc4HveC4yTBunMbDzd9IYsC2n7wW8h5qtk1DuwS/w\nclXdOqOtVTSOfQj37vRh+1qmdwHXA8Po7XJS5y1R1XtTmXtkCOcg8cdN6ZPlqd8cXDvHBb5vi59R\n0xawrYv7v+tQ0zkJ5R78vkJEjlTVPgwxKgnh3p0+bF/L9CbglhPl1WtO6rzrROQIJjNb/bFrI9Qf\nN2Wmk7Ub0Ay3ep10jJp2LkKeimCZhJJQC9g2F/OynhawLafvBXIONdval3vwuwE2UV3P5MF8TmRV\nCLXMyBiF7WtvAq5HKh8gp/lcTuq8BZhVwI7YSv8DHdrqZXnqxysRkc2AzbC0cH9Wi+n+6fRZrule\nHVXnJ+5zkRq6xK3phKrek/48V0RehalSvFSeHYk/9ruLqoNfqmMP7VTxXi6NYx/CvTt92L6WWZYC\nbgoZJlw55nM5qfNWxwTu14G9sBVPVxO7NoYVxMPGyynXty/wGszS4mQsRPG+hfHMNd2ro6rfrnOR\n5By0P3afAFDVl6jqofXfGimbZZav6nsfh5plXAe/ZRPGNFlVPu9dxz6Ee3fctq8ZLEsBB3SK3QH+\nsALl1HlNeumvAV8EXk8/ycGb6Kwfz5jYcngjtlP7oap+RkR+Xvo8Ox5JDVX99p6LZDlGjQqxDESL\nVDU38UpV3/s41CzTNVNX02TVaexDuHen1QGjA8tSwAGdTLjAaT6nFuLUo38k1Xcu8G7tJzn4qMiJ\nl+NlNknvml4/WPywg+leDt5zEbdj1CgQkR2wA8+7sYTie6lqXQweL6M41MzK1OWcrDqNfQj37rTa\nvnZgmQu4HBOuXPO5DP0j9JwcvIVh1DLD2oVXcTp2zvIksVyq36wp12i611E37z0XGTY36LAciiU8\nv1ksYco3qA+wVkVV3/s41JxC28HvgMzJqtPYh3DvTo4DhpdlKeAG5Jhw5ZrPeR1PwA6nd6an5OAD\n+tKP92gXXsUFWIjmpwGqqlfUlGsz3euim/eei5Sdg5Y1j6jqzQCq+uc6S7KcsyP6OdSchvPgN2ey\n6jT2Idw7kumA4WUkAq6JHBOuDuZzbv2jql4iItcB62C7l6Ht0XvWj4/SLvxEVd0Wc15ros10r4tu\n3nUuklRsr8Qma10OpqR3i8h+TFqS/bX4YZezo5xDzSGo06W7Jqv0eaexD+E+gxiFgPOSqULxms+5\n9Y+l66+JmYF2TV83oDf9eI924VXcJxYHR7GkKk2xYGpN97ro5r3nImLRCDfDxvBfReRFqvpezzV6\nYjdMeB+OnUftUfyw49lRmVwLnFocuvTGyapUV6exD+E+gxiRgPOSo0Jxmc959Y8dru9lFPrxYe3C\nq7g0/d6wpZzXdM8dKzxjUt9OVbdJ3/kscFlLW3tBJsMKbIglWxl4nc6jcA9yzo4qrtHVAqeqLq8u\nvXGyKtFp7EO4zyxGIeC85JhwucMKZDiedDUhm8aI9eM58XJcZGy7vaZ7ObHCvc/cqiIyW1WXMClg\nlwXesAK57v+jssBp1KV7J6sSncY+hPvMojcB14EcE66csAJFmra9WSZkLYxSP54TL8dFxrbba7qX\n41bvfea+Clwilkzk+cAZLW3oBS2EFRCRDYBNgIVaSpaSc3ZUYFgLnCradOldYuB0GvsQ7jOLPgVc\nFpkqlJywAq5tb7r+2liEzbbrNzJi/XjXia0J77bbbbrXpJsv4XrmVPWIZPr7VOAEVf2do1+9ISIL\ngPdjh85PFYsVc2ZFuZyzI/ehZgaNunTvZFX6TqexD+E+g+hTwHW8vleF4jKfy9n2ppXTf2CqlLOw\nh/7W7r0BRqMfz5rYnHi33V7TPbdbfYZd9jMwgXkT8DkR+biq/tDZnj54G/BMVX1ALMXghcA04U6e\natN9qJmBS5funaxS2U5jH8J9BjEiAdeVJhWKN6xAzrb3OOAILKzDTzHBPOxhcu/6cfLi5Xgpb7u/\nWlUow3Qvy63eOakfA+yLpfr7IPApzDZ/WbEIeDj9/X/U75xyVJs5h5qNdNCleycr6Dj2IdxnFqMQ\ncFk4VSjesAI5295HqeqFIvIhVdWetsi968fJi5fjorTtPklVr3R+tW4CHsatvq7OBzCb8dVU9TLp\nJyF8K2Ip8QZC8pdpAnw2JuCraFUzdTzUbCNXl+6drKDj2Idwn1mMQsC5yFShePWaOdveB0TkZcAq\nIvIC7IEelt714xkTmxsR2QuYr6r/LiLni8ipqlqboNwxAWe71TvqnABOAb4rIv/EpGAaNcdUvFeb\nbcmpZupyqNmIV5feYbKCjmMfwn1mMQoB5yVHheLVa+Zse/cG/gsTTO/Ftq3D0rt+PPPAzsvbgOel\nv1+JTYbThHvGBOx2q8+oc9fUxoGJ5TLJ5KSqP0nt3AnYSlU/IiLnYZES677TqGbqcqjpxaFLz5qs\nEp3GvnPmkWAk7I3FdO9TwHmZokKheWJZpJZMYi2tSOKdLDRgctt7AHACtlqp4z2q+kZV/TtVfb2q\nXt9Q1ssdInKsiLxVRPaWknNLRwYT2/cx2/k+JuBHVHUxgKo+TPsZxjOBbbB8u9NQ1RuKP1ie3zpc\ndWKes48G/gX4GyzvwLLkECZjAu0KfMT5vdqzoySIL8b02D9Lq+JhGejSXwNshT37S1HVn6QJax3g\nJenvA5lq0VWm09jHyn1m8R5VXV65LXNUKKPIZ7mFiDymj9VTgd7144zGF+EcEbkIUyE8m/osUV1N\n95oOx711DhLCPwNTIQydED6Th1X1LgBVvatN7+w5OyLvUNOLV5d+CDBICbgrtiqvywfRaexDuM8s\nRiHgvLhVKG16zY7b3i2wlfYdTOqJh9Jnj0I/zgh8EVT1MBH5NiDAyVofFTLLdM8p4Lx1jiIhfA6X\ni8jpwM8wFcWvqwrlnB2Rd6jZSAddes5k1WnsQ7jPLHoXcG10tRzwmM/l2PKq6pNq2tc5mfUo9ONe\nu3Bn+9bAttifA+7AJtcHROS9qvqXiq94bahzBJx3Uh9FQng3qrqfiOyCTYBnqeq3aoq2nh11PNRs\nI1eX7pqsEp3GPoT7DGIUAs5BH5YDddv+Pra9wySzHkmsngxnrzY+hwU1m40lXP85ZvL2RSynKtBp\nAvYIuNw6jwLeA5yPpXu7uFuXuyEi62ET9C3AuiJyoKp+oqKoR83U5VCzkdyD34zJCjqOfQj3FYNh\nBFwjw1gOOLb9fWx7h8matCxi9QwTJvbvVHWbtCp7EfB6VX1YRA4olcudgD0CLqtOVf06LBWyZ6lq\nH2klcxjonZ+OHWLX6Z1b1UxdLHAycOnSMyarzmMfwn3FYBgB5yJHhdK27e952zvMIejIYvU49dlt\nDHYA2wCXJ0sZKLWzwwTsEXBZdYrIdsDRwCrAWSJyg6qe6O1oD3j1zjnmtzmHml68unTvZNV57MMU\ncsVgWYRXbTThKtFmPncMtiI8ANt1fAX4d+DDfTe6CbVctsdipmZfBl49bJ0isoNYQpULsMxHOw5R\n3b3JPPP9wBkiMltE/gX4U821vaZ7u2Emc4cDT6RBwGXUeRg2UfwF+DjDB2DLpVHv3NH8doogBvrw\nur1cRE4Xkf1E5FTqdemzVHUfLEHLjsB6DXV2GvsQ7sGAHBVKo018R1veOobatajqPaq6WC213nOH\nqSvhtQv3sA+wKbZiPBlbRb6eev+Gxgm4o4DzTupLVPWv2CH/A0zuOpYVR2G+H1diAbTKfhD7p9/H\nYouLLxb+rsMriN2o6n7YudKamArlnTVFcw5JO419qGVWDEamlumoQvGaz2Vte6WnZNYN9JFGrbcw\nsWoOYO8XkVepOYX9EPhhWj1XqcTaJuAuh+PeSX2hWNz59UXkA/Sfa7SNCUwdcyfwIBb1cildzo4y\nDzVdZOjSy5NVk3ljp7EP4T7DWAYCrkwXywGvXtNtyyv9JrMu191bGjV6DBObLG1eCLxZRAaxX2Zj\nSdLPLJRzTcA5Aq7DpP527D5fDNwH7NWhy8NwMPA8Vb1NRDbEcgxXHVTmnB25DzUz8OrSGyerEp3G\nPoT7DGKUAq6OHMuBDuZzOba8vSWzLrR3FGnUegsTC/wWWB8TqJreW8L0TDtZE7BTwOVO6t9W1X9o\n+HzULFLV2wBU9VYRqbMYyTG/dR9qZuA9+HVNVolOYx/CfWbRu4DLwKNCyTWfy9n2jiKZdW9p1DpM\nbK2o6o3AyWJRIJc0lMs13WsVcB3qvFNEXg1ci01AqOq13r72wD1iYZF/AjwHWFNEPp7acVChXM7Z\n0Si8br26dO9kBR3HPoT7zGIUAs5Lqwqlg/lc67ZXRpvMus80ar2HiS3wfhF5P7ZynEW9Z7L3DCNH\nwHnrfCzmSDOgj37n8M3C338uf9jx7GgUXrdeXbp3soKOYx/CfQYwYgHnxa1CydBrera9owyU1pt+\nfBhnLwdvBDZW1Ta1QOME3FHAuc5FVHWHqvfTvT+k6rM+UdWTW4p0OTvKOdT04tWlN05WRbqOfQj3\nmcHyigS5lEwVilev2brt1dEms+5TPw7kHdhlcD0+B6+2CbiLgMs5F6nixZnlR0IHNRPkHWp6cenS\nHZOVh8axDzv3GYBOxt3eEHP6eX/hZ5lQpUJpKO7d9udse08GrsD+EQY/neho7+0lx9nLy2rAlSLy\nlWR3fXpVoTYb6i7+BRl22XWM3Hs6k5y47wNB/CwswNyhPVx/ii4dO8wfFY1jHyv3mcXJWILsrnkc\nh6FVhdJh25+z7e0zmfUo9eO9hYkt4MoWlWG65/Yv6MEccFl4T+eQE0o351DTS44ufVgaxz6E+8yi\nTwGXi8dyIHfbn7Pt7S2Z9Sj04x312V5+he3SNsYiTNbFc/ea7uUIuFGYAy5PctRMoxDEbl36qAnh\nPrPoTcB1oFWF0kGvmWPL23sy6571472HiS1wEra6fjEWP+REqvWpXtO9HAHXWKeIzG8xu5tRapnM\ns6PeBXFPunQvoZZZgehdwGWQo0Jxm+RlbHt7T2ZNj2nUOh7YeVlfVU8Skd1U9VIRqTsLc51hZAq4\ntjpPBZ4vImenc4YyuzfUvczJUTMtY0HcGRH5gqruW3h9iqruTsvYh3CfWYxCwHnJUaF4t/052947\nRORYTEUxkcoc17UziVHox0cRJhYReWr6/QRgcU0x1wScqUdvq/M6EbkNeLSI3MzkanFCVTdOjlgz\nibFRM4nIOzBrr/VE5LXY2M/CErrQNvYh3GcWoxBwXnJUKN5tf862t7dk1iPWj2clanbyTmwy3RzL\nFlVnAuqdgHMEXFtArjcBiMhRqvoOR1+WN8s712tvqOpRwFEicpCqfjz3+yHcZxa9CbgOuFUo3m1/\nzrZX+01mPUr9+LB24dNQ1aswU7w2vBNwjoDz1nmgiBwKPJ506KuqCyvKLW+Wa67XPhGL9Q+26Nu7\n+Jln0bfCdnwc6VnA5eJWofRgPjcN6TGZ9Sj145n6bBcicjjmYLV0Mq8JP+CdgHMEnLfOEzEV1PY0\nH/oub0bhdbq82KjmfdeiL4T7DKJPAdeBHBXKKPSao0hm3bt+fBQTGxYwbhNVfbClnHcCzhFw3jq9\nh77Lm1F4nS4XiqEFuiz6QrjPLEYh4FxkWg6MQq85imTWo9CPj2Ji+zXmRdom3L0TcI6Ac0/qzkPf\n5U3O2dEKQddFXwj3mcUoBNwoGIVecxTJrHvXjzOaie0q4BYR+QuTUSGfXC6UMQG7BVxGnd5D3+XN\nKLxOlzedFn0h3GcWoxBwo6B3vaaqHiQiczGLlkHCkmHr7F0/zmgmtl2Bv6U//4ZRCLiXq6rn0Hd5\nsyzd/5cVnRZ9IdxnEKMQcCNiJHpNVR0k/j1XLAXdt4epb0T68VEc2N0A3OfQuXsZhYB7hYgcqap9\nqLZGyYxx/++RTou+EO4zjL4F3IhYFnrNPpJZj0I/PoqJ7YnAH0XkusE1VPWFTV9oYRQCbgNMuFyP\njcGwbRwJK4rXaQ5dF30h3Gc2fQi4UTAyvab0m8x6FPrxUUxsuw7frElGJOB2GkGdgZMui74Q7jOQ\nngXcKOh92y+jSWY9Cv34KCa2JcCbmBp3fVkFjHOhKanKgCRgbqgpHowW16IvhPsMYkQCbhSMYtvf\nWzLrAqPQj49Cn/3/27v3UEurMo7j30HGxpkK834ZRZF6SjGveOkyoTFIFKVI5h9iGWmjYuGFwktY\nXkqDbn9EammGoZDFZEVRSeYgHhEtIkt/6YxUjjbqaKl4yfT0x9qbOTqT2rv3e571vu/vA4e92cwc\nnr9+51lrr3c91wM3ArXd0/JKal1V9tb/2/Q53OvSRsBNXUvL/mkOsx5rY3+8jT9sT0o6b0q/q1Ud\nWFX2TtOmz+FelzYCriumNsx6jqnvj7f0h+2uiDiWcg5/fGHcK92hPu86tKrso0ZNn8O9Lm0EXFdM\nfZg13XmgZd/Rz9g0xgFOWydWlT3VqOlzuNeljYCrWkQslfQAG4ZZj2edbsvks2Q78UCLpMNe/V+l\nG/KqMlujps/hXoGWA652bQ6z7sQDLRGxFtgOeIRynvxZYB1wSkVbH0NeVWZr1PQ53OvQZsBVTS0M\ns57zu7vyQMsq4POSFBF7AOdTjkJ+n3q2Pga3qsw2adNX67WdgzI34IBjgM8AR0nqdbDPFWWY9S3A\nucBMRByTXNJ8WipJAJJWA7uOBmGk37w4ugESNgTMmcB3KAFj7Tpj9Ho5ZQDNt+a8f1Xu3CsyCrjP\nUh6Zf2tEnC+p0UDnDpraMOsOeigiLgFuBd4B/CMilgP/zi0LGPCqMtukq1p37nUZB9xRwIGULmko\n2hhm3RXHAw9S7g35G/Ax4CnKU6upvKrM13RV6869LoMLuGh3mHXVIuJASXdQvqC8Z/QDsEzSr/Iq\n29jAV5XZGq1qHe4VGHLA0e4w69q9F7iDjTv0WaCqcGfY22bZGjV9Dvc6DDbg2hxm3QFfi4jNgU9m\nF/IaDG5VmW3Sps/hXoGBB9zY1IdZd4DYeJL9+MvKjcbsZRj4qjLbRE2fw70uQwy4sTaGWVdN0u7Z\nNbwGg11VZpu06XO412VwATdHG8OsOyEiPgicCiykdO5bS3p7blWFV5VVaNT0OdzrMtiAa2mYdVdc\nRNl3XwHcBCzPLWeThryqzNao6fM594pIOo1yAmExJeA+lVzSvNnUMOvkkubTQ5JmACRdDeycW84m\nvSRggCGtKrPdHhHXRsRpEXENr7Hpc+dekU0FnKQvJZc1X9oYZt0Vz0XEMmBhRBxBuTysNoNdVWZr\nuqp1516XlcB7KEv04ymPog/FAkkrKCdIlgNbJdczn06m7LdfBJw0eq3KkFeV2Zquat2512WBpBUR\ncRVlPNw0Zn52RRvDrDtB0lo2XEl89PjziFg5uooi3cBXldkarWrduddlsAHHxsOs788tpwpbZhcw\nx5BXldkarWod7nUZcsCNh1kfzfSGWXfdyx9wyjTkbbNsjZo+h3tdhhxw42HW+wGHUmZ2Wj2GvKrM\n1qjpc7jXZcgB95Jh1kCtw6yHasirymyNmj7/9a3LSwIuIoYUcJ0YZj3PapqfOw6YxxneqjLbuOl7\nOCK2B36Kn1DtnCEHXCeGWbchInahXPu7aPyZpAskHf2//9e8axQwNhWNmj6He10GG3AdGmbdhuuB\nG4G/ZxfyCoa8qszWqOlzuFdk4AE3ZE9KOi+7iFcx5FVltkZNn8PdLN9dEXEs5ZH+WQBJf8ktaSOD\nXVVma9r0LZidrekordnwRMRNL/to1gOobVIOd7MKRMTWwB7AGkmPZtdj3edz7mbJIuLDwK3AOcBt\nEXFccknWAw53s3xnAAdIOhLYD/h0cj3WAw53s3wvSnoKQNKTlJv/zCbi0zJm+dZExFeAVcAyYHVy\nPdYD7tzN8p0ArKHctrgaODG3HOsDh7tZvtcBN1AmMG0N7JRbjvWBw90s3w+B/YEvA88DV+SWY33g\ncDfLt5hyEddSSZcAmyXXYz3gcDfLtznl+OOdEbEnZSCG2UQc7mb5zqLss18MHI7PudsU+PoBswpE\nxI7AQmABsJOkmeSSrON8zt0sWURcSRmruISy/74aOCS1KOs8b8uY5dsH2Isy2eht+AlVmwKHu1m+\n9ZJmgSW+EdKmxXvuZslGE40eA7YHdgF2l3RwblXWdQ53swpExOsp2zHvA26XtC65JOs4b8uYJYuI\nnYHLgJ8DOwC7pRZkveBwN8t3BXAV5SjkKuAbueVYHzjczfJtIek3lNmpwqdlbAoc7mb5no2II4DN\nIuIQHO42BQ53s3wnUe5034ZyFcHJueVYH/gJVbN8p0s6NrsI6xd37mb59oyILbOLsH7xOXezZBHx\nV2Bn4FFglvLFqqcx2UQc7maViogPSbohuw7rJm/LmNXL97pbYw53s3otyC7AusvhblYv75laYw53\nM7Mecrib1cvbMtaYH2Iyq0BE7A28BfiTpHtGH381sSTrOB+FNEsWEecxuscdOBj4gaSv51ZlXedt\nGbN87wfeLel0YBngqwhsYg53s3zrgMWj95sDjyTWYj3hPXezJBExQznuuB1wb0T8AdgTWJ9amPWC\nw90sj7dfrDX+QtUsWUQcRAn6RePPJJ2SV5H1gTt3s3zfAy4FHs8uxPrD4W6W715JV2cXYf3ibRmz\nZBHxUco59z+PP5N0QV5F1gfu3M3ynQr8CPhndiHWHw53s3zrJV2aXYT1i7dlzJJFxDXA08DvGF3z\nK+mK1KKs89y5m+W7b/S6w+jVHZdNzJ27WQUiYkdgIeWa350kzSSXZB3nzt0sWURcCRwKLAG2ANYA\nh6QWZZ3ni8PM8u0D7AX8knK3zLO55VgfONzN8q2XNAsskfRodjHWD95zN0sWEV8EHgO2B3YBdpd0\ncG5V1nUOd7MKRMQbgGcYTWSStC65JOs4h7tZZSLiA5J+ll2HdZv33M3q8+bsAqz73LmbVSIitqV8\nufpidi3WfQ53s2QRcRhwJfAEsCVwoqRf51ZlXedtGbN8FwLvkrQv8E7gouR6rAcc7mb5XpD0IICk\ntfghJpsCXz9glu+JiDgNWAUso5x5N5uIO3ezfMcBuwIXUx5i+nhuOdYH7tzNkkTEUkkPUJ5M/Tbl\nRshZYFs8LNsm5HA3y3PG6OdyNtzhPg74w7OKsn7wUUizCkTENsBuwH2SPEvVJuY9d7NkEXECcAtw\nLjATEcckl2Q94HA3y3cysK+ko4ADgTOT67EecLib5VsPPD96/wzgbRmbmL9QNUsSEdex4XTMnRFx\nG7A/JeDNJuJwN8tz2SY+u27eq7Be8raMWRJJN0u6GXgjcPjo/dnAotzKrA/cuZvl+wJw2Oj9R4Bf\nUIZlmzXmzt0s3/OS/gUwen0huR7rAXfuZvluj4hrgRngIOD3yfVYD/gJVbMKRMSRQAB3S/pJdj3W\nfd6WMUsWEVsBWwAPAW+KiLOTS7Ie8LaMWb6VwN3A3pRBHU/nlmN94M7dLN8CSSsAAcuBrZLrsR5w\nuJvl+09ELAKWUJ5Y9YraJuZwN8v3TeAs4I/AA8D9ueVYH7hDMMs3C3yCMn3pOeC7ueVYH7hzN8v3\nOeAgSfsBhwIXJtdjPeBwN8u3XtLDAJLWAU8k12M94IeYzJJFxEpgMXAzcACwI/BbAEnn5FVmXeY9\nd7N8P57zfm1aFdYr7tzNzHrIe+5mZj3kcDcz6yGHu5lZDznczcx6yOFuZtZD/wUJaJPUSaIydgAA\nAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11491fe48>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "feature_importances = pd.Series(model.feature_importances_, index=X_test.columns)\n",
    "feature_importances.sort()\n",
    "feature_importances.plot.bar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#TODO: Try training the model with the most important features and note difference in overal model prediction score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0    139\n",
       "2.0    136\n",
       "0.0    145\n",
       "dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: Training Set - Show number of Hom Ref, Hom Var, Het Var datapoints the model was trained on\n",
    "ytrain = pd.DataFrame()\n",
    "ytrain['ytrain'] = y_train\n",
    "pd.value_counts(ytrain['ytrain'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='low_precision'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the training subset: 0.939\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the training subset: {:.3f}'.format(precision_score(pred, y_test, average='micro'))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>pacbio.alt_alnScore_mean</th>\n",
       "      <th>pacbio.alt_alnScore_std</th>\n",
       "      <th>pacbio.alt_count</th>\n",
       "      <th>pacbio.alt_insertSize_mean</th>\n",
       "      <th>pacbio.alt_insertSize_std</th>\n",
       "      <th>pacbio.alt_reason_alignmentScore</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>-36.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>67518800.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15534.50000</td>\n",
       "      <td>6819.368757</td>\n",
       "      <td>32.0</td>\n",
       "      <td>9820.281250</td>\n",
       "      <td>4126.553808</td>\n",
       "      <td>32.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67518763.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>-1926.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>87382085.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4070.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2582.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10618.191180</td>\n",
       "      <td>3941.521434</td>\n",
       "      <td>68.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>87380158.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.033731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>-26.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>165220279.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17449.30435</td>\n",
       "      <td>6003.157745</td>\n",
       "      <td>23.0</td>\n",
       "      <td>10925.130430</td>\n",
       "      <td>3864.813717</td>\n",
       "      <td>23.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9352.789474</td>\n",
       "      <td>3920.835445</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>165220252.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>-27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>51460422.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12129.85714</td>\n",
       "      <td>6914.355397</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7604.357143</td>\n",
       "      <td>3970.730476</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51460394.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>-27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5928819.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16790.16667</td>\n",
       "      <td>4330.578782</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10764.222220</td>\n",
       "      <td>2746.611965</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11108.692310</td>\n",
       "      <td>2980.377554</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5928792.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.518519</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Size  chrom          end  pacbio.GT  pacbio.alt_alnScore_mean  \\\n",
       "434   -36.0   11.0   67518800.0        2.0               15534.50000   \n",
       "122 -1926.0   10.0   87382085.0       -1.0                4070.00000   \n",
       "224   -26.0    1.0  165220279.0        1.0               17449.30435   \n",
       "479   -27.0   19.0   51460422.0        2.0               12129.85714   \n",
       "205   -27.0    1.0    5928819.0        1.0               16790.16667   \n",
       "\n",
       "     pacbio.alt_alnScore_std  pacbio.alt_count  pacbio.alt_insertSize_mean  \\\n",
       "434              6819.368757              32.0                 9820.281250   \n",
       "122                 0.000000               1.0                 2582.000000   \n",
       "224              6003.157745              23.0                10925.130430   \n",
       "479              6914.355397              14.0                 7604.357143   \n",
       "205              4330.578782              18.0                10764.222220   \n",
       "\n",
       "     pacbio.alt_insertSize_std  pacbio.alt_reason_alignmentScore  \\\n",
       "434                4126.553808                              32.0   \n",
       "122                   0.000000                               1.0   \n",
       "224                3864.813717                              23.0   \n",
       "479                3970.730476                              14.0   \n",
       "205                2746.611965                              18.0   \n",
       "\n",
       "         ...        pacbio.ref_insertSize_mean  pacbio.ref_insertSize_std  \\\n",
       "434      ...                          0.000000                   0.000000   \n",
       "122      ...                      10618.191180                3941.521434   \n",
       "224      ...                       9352.789474                3920.835445   \n",
       "479      ...                          0.000000                   0.000000   \n",
       "205      ...                      11108.692310                2980.377554   \n",
       "\n",
       "     pacbio.ref_reason_alignmentScore  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "434                               0.0       0.0       0.0         1.0   \n",
       "122                              68.0       0.0       0.0         0.0   \n",
       "224                              19.0       0.0       0.0         0.0   \n",
       "479                               0.0       0.0       0.0         0.0   \n",
       "205                              26.0       0.0       0.0         0.0   \n",
       "\n",
       "     segdup_pct        start  tandemrep_cnt  tandemrep_pct  \n",
       "434         1.0   67518763.0            1.0       1.000000  \n",
       "122         0.0   87380158.0            2.0       0.033731  \n",
       "224         0.0  165220252.0            0.0       0.000000  \n",
       "479         0.0   51460394.0            1.0       1.000000  \n",
       "205         0.0    5928792.0            1.0       0.518519  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add original labels and predicted labels back to the original dataframe\n",
    "df_Xtest = pd.DataFrame(X_test)\n",
    "df_Xtest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "labels = pd.DataFrame(y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['predicted_label'] = pred\n",
    "df_Xtest['GTcons'] = df_train['GTcons']\n",
    "df_Xtest['chrom'] = df_train['chrom']\n",
    "df_Xtest['start'] = df_train['start']\n",
    "df_Xtest['end'] = df_train['end']\n",
    "# df_Xtest['Y_test'] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['GTcons'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['GTcons'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['GTcons'].replace(2.0, 'Homozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Heterozygous_Variant    61\n",
       "Homozygous_Variant      64\n",
       "Homozygous_Reference    55\n",
       "dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Heterozygous_Variant    62\n",
       "Homozygous_Variant      65\n",
       "Homozygous_Reference    53\n",
       "dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['predicted_label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[57  0  4]\n",
      " [ 0 53  2]\n",
      " [ 5  0 59]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "ytest = df_Xtest['GTcons']\n",
    "predict = df_Xtest['predicted_label']\n",
    "print(confusion_matrix(ytest, predict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>62</td>\n",
       "      <td>53</td>\n",
       "      <td>65</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                    57                     0   \n",
       "Homozygous_Reference                     0                    53   \n",
       "Homozygous_Variant                       5                     0   \n",
       "All                                     62                    53   \n",
       "\n",
       "Predicted             Homozygous_Variant  All  \n",
       "True                                           \n",
       "Heterozygous_Variant                   4   61  \n",
       "Homozygous_Reference                   2   55  \n",
       "Homozygous_Variant                    59   64  \n",
       "All                                   65  180  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(ytest, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       0.92      0.93      0.93        61\n",
      "Homozygous_Reference       1.00      0.96      0.98        55\n",
      "  Homozygous_Variant       0.91      0.92      0.91        64\n",
      "\n",
      "         avg / total       0.94      0.94      0.94       180\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(ytest, predict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Predict\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='predict'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the [previous section](#machine_learning), a RF model was trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](#train_test) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data\n",
    " * Reminder: The labels for this training set and the following [prediction step](#prediction_step) are the consensus genotype (GTcons) labels generated from a preliminary R analysis based on reference and alternate read count:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "           \n",
    "   * The trained model is used in the following section to predict labels for 5000 randomly selected Deletions [these datapoints were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "   \n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Load Data **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X2 = df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Impute missing values using KNN **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/3281 with 0 missing, elapsed time: 2.688\n",
      "Imputing row 101/3281 with 0 missing, elapsed time: 2.688\n",
      "Imputing row 201/3281 with 0 missing, elapsed time: 2.689\n",
      "Imputing row 301/3281 with 0 missing, elapsed time: 2.690\n",
      "Imputing row 401/3281 with 0 missing, elapsed time: 2.692\n",
      "Imputing row 501/3281 with 0 missing, elapsed time: 2.693\n",
      "Imputing row 601/3281 with 0 missing, elapsed time: 2.694\n",
      "Imputing row 701/3281 with 0 missing, elapsed time: 2.695\n",
      "Imputing row 801/3281 with 0 missing, elapsed time: 2.697\n",
      "Imputing row 901/3281 with 0 missing, elapsed time: 2.699\n",
      "Imputing row 1001/3281 with 0 missing, elapsed time: 2.699\n",
      "Imputing row 1101/3281 with 0 missing, elapsed time: 2.702\n",
      "Imputing row 1201/3281 with 0 missing, elapsed time: 2.704\n",
      "Imputing row 1301/3281 with 0 missing, elapsed time: 2.705\n",
      "Imputing row 1401/3281 with 0 missing, elapsed time: 2.705\n",
      "Imputing row 1501/3281 with 0 missing, elapsed time: 2.706\n",
      "Imputing row 1601/3281 with 0 missing, elapsed time: 2.706\n",
      "Imputing row 1701/3281 with 0 missing, elapsed time: 2.707\n",
      "Imputing row 1801/3281 with 0 missing, elapsed time: 2.708\n",
      "Imputing row 1901/3281 with 0 missing, elapsed time: 2.708\n",
      "Imputing row 2001/3281 with 0 missing, elapsed time: 2.709\n",
      "Imputing row 2101/3281 with 0 missing, elapsed time: 2.710\n",
      "Imputing row 2201/3281 with 0 missing, elapsed time: 2.710\n",
      "Imputing row 2301/3281 with 0 missing, elapsed time: 2.711\n",
      "Imputing row 2401/3281 with 0 missing, elapsed time: 2.711\n",
      "Imputing row 2501/3281 with 0 missing, elapsed time: 2.712\n",
      "Imputing row 2601/3281 with 0 missing, elapsed time: 2.713\n",
      "Imputing row 2701/3281 with 0 missing, elapsed time: 2.713\n",
      "Imputing row 2801/3281 with 0 missing, elapsed time: 2.714\n",
      "Imputing row 2901/3281 with 0 missing, elapsed time: 2.715\n",
      "Imputing row 3001/3281 with 0 missing, elapsed time: 2.715\n",
      "Imputing row 3101/3281 with 0 missing, elapsed time: 2.716\n",
      "Imputing row 3201/3281 with 0 missing, elapsed time: 2.716\n"
     ]
    }
   ],
   "source": [
    "#Convert dataframe to matrix\n",
    "X2=X2.as_matrix()\n",
    "X2=pd.DataFrame(X2)\n",
    "\n",
    "# Imput missing values from three closest observations\n",
    "X2_imputed=KNN(k=3).complete(X2)\n",
    "X2=pd.DataFrame(X2_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37568322.0</td>\n",
       "      <td>37568587.0</td>\n",
       "      <td>-115.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.818868</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.378230</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>112835104.0</td>\n",
       "      <td>112837661.0</td>\n",
       "      <td>-2534.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.460305</td>\n",
       "      <td>...</td>\n",
       "      <td>159.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11538.50000</td>\n",
       "      <td>5911.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7826.000000</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3557049.0</td>\n",
       "      <td>3557087.0</td>\n",
       "      <td>-37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8573.00000</td>\n",
       "      <td>4836.734918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6021.000000</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end    Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0   37568322.0   37568587.0  -115.0       0.0       0.0         0.0   \n",
       "1    1.0  112835104.0  112837661.0 -2534.0       0.0       0.0         1.0   \n",
       "2    1.0    3557049.0    3557087.0   -37.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000            3.0       0.818868                ...                  \n",
       "1    0.890888            6.0       0.460305                ...                  \n",
       "2    0.000000            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "1                                            159.0   \n",
       "2                                            151.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "1                       270.0                             8.0   \n",
       "2                       207.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "1                            9.0               11538.50000   \n",
       "2                           10.0                8573.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7611.378230              31.0                 8407.903226   \n",
       "1              5911.500000               2.0                 7826.000000   \n",
       "2              4836.734918               3.0                 6021.000000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "1                4015.000000                               2.0  \n",
       "2                3561.905763                               3.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dftest_header = list(df_test.columns.values)\n",
    "X2.columns = dftest_header\n",
    "X2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37568322.0</td>\n",
       "      <td>37568587.0</td>\n",
       "      <td>-115.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.818868</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13276.22581</td>\n",
       "      <td>7611.378230</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8407.903226</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>112835104.0</td>\n",
       "      <td>112837661.0</td>\n",
       "      <td>-2534.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.460305</td>\n",
       "      <td>...</td>\n",
       "      <td>159.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11538.50000</td>\n",
       "      <td>5911.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7826.000000</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3557049.0</td>\n",
       "      <td>3557087.0</td>\n",
       "      <td>-37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8573.00000</td>\n",
       "      <td>4836.734918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6021.000000</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end    Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0   37568322.0   37568587.0  -115.0       0.0       0.0         0.0   \n",
       "1    1.0  112835104.0  112837661.0 -2534.0       0.0       0.0         1.0   \n",
       "2    1.0    3557049.0    3557087.0   -37.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000            3.0       0.818868                ...                  \n",
       "1    0.890888            6.0       0.460305                ...                  \n",
       "2    0.000000            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            149.0   \n",
       "1                                            159.0   \n",
       "2                                            151.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       209.0                             4.0   \n",
       "1                       270.0                             8.0   \n",
       "2                       207.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           16.0               13276.22581   \n",
       "1                            9.0               11538.50000   \n",
       "2                           10.0                8573.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              7611.378230              31.0                 8407.903226   \n",
       "1              5911.500000               2.0                 7826.000000   \n",
       "2              4836.734918               3.0                 6021.000000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                4770.945293                              31.0  \n",
       "1                4015.000000                               2.0  \n",
       "2                3561.905763                               3.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X3 = pd.DataFrame()\n",
    "X3 = X2\n",
    "X3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X5 = X2.reindex_axis(sorted(X2.columns), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# X5 = df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred_prob = model.predict_proba(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred_prob_log = model.predict_log_proba(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X5['predicted_label'] = pred\n",
    "X5['chrom'] = df_test_2['chrom']\n",
    "X5['GTcons'] = df_test_2['GTcons']\n",
    "X5['start'] = df_test_2['start']\n",
    "X5['end'] = df_test_2['end']\n",
    "X5['Size'] = df_test_2['Size']\n",
    "X5['GTconswithoutpacbio.GT'] = df_test_2['GTconswithoutpacbio.GT']\n",
    "X5['GTsupp'] = df_test_2['GTsupp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.concat([X5, pd.DataFrame(pred_prob, columns=['1','2','3'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X7 = pd.concat([X5, pd.DataFrame(pred_prob_log, columns=['1','2','3'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_DEL_revisedR.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X7.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_log_DEL_revisedR.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Note: Reformat X6 csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/preliminary_df/PB_pred_prob_DEL_revisedR.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.rename(columns={'1': 'Homozygous_Reference_GTcons_PB'}, inplace=True)\n",
    "X6.rename(columns={'2': 'Heterozygous_Variant_GTcons_PB'}, inplace=True)\n",
    "X6.rename(columns={'3': 'Homozygous_Variant_GTcons_PB'}, inplace=True)\n",
    "X6.rename(columns={'predicted_label': 'PB_predicted_GTcons_label'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homozygous_Reference_GTcons_PB</th>\n",
       "      <th>Heterozygous_Variant_GTcons_PB</th>\n",
       "      <th>Homozygous_Variant_GTcons_PB</th>\n",
       "      <th>GTcons</th>\n",
       "      <th>GTconswithoutpacbio.GT</th>\n",
       "      <th>GTsupp</th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>PB_predicted_GTcons_label</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>-115</td>\n",
       "      <td>1</td>\n",
       "      <td>37568587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4770.945293</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37568322</td>\n",
       "      <td>3</td>\n",
       "      <td>0.818868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2534</td>\n",
       "      <td>1</td>\n",
       "      <td>112837661</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4015.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.890888</td>\n",
       "      <td>112835104</td>\n",
       "      <td>6</td>\n",
       "      <td>0.460305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-37</td>\n",
       "      <td>1</td>\n",
       "      <td>3557087</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3561.905763</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3557049</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Homozygous_Reference_GTcons_PB  Heterozygous_Variant_GTcons_PB  \\\n",
       "0                             1.0                             0.0   \n",
       "1                             0.1                             0.5   \n",
       "2                             0.1                             0.6   \n",
       "\n",
       "   Homozygous_Variant_GTcons_PB  GTcons  GTconswithoutpacbio.GT  GTsupp  Size  \\\n",
       "0                           0.0       0                       0       4  -115   \n",
       "1                           0.4       2                       2       1 -2534   \n",
       "2                           0.3       1                       1       2   -37   \n",
       "\n",
       "  chrom        end  pacbio.GT      ...        pacbio.ref_insertSize_std  \\\n",
       "0     1   37568587        0.0      ...                      4770.945293   \n",
       "1     1  112837661       -1.0      ...                      4015.000000   \n",
       "2     1    3557087       -1.0      ...                      3561.905763   \n",
       "\n",
       "   pacbio.ref_reason_alignmentScore  PB_predicted_GTcons_label  refN_cnt  \\\n",
       "0                              31.0                          0         0   \n",
       "1                               2.0                          1         0   \n",
       "2                               3.0                          1         0   \n",
       "\n",
       "   refN_pct  segdup_cnt  segdup_pct      start  tandemrep_cnt  tandemrep_pct  \n",
       "0         0           0    0.000000   37568322              3       0.818868  \n",
       "1         0           1    0.890888  112835104              6       0.460305  \n",
       "2         0           0    0.000000    3557049              0       0.000000  \n",
       "\n",
       "[3 rows x 39 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X6.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Oct122017_Revised_R/summary_plots/data/PB_final_GTcons_df_DEL_SVanalyzer.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Label Analysis\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "  * The [random forest(RF) model](#train_test) was trained on svanalyzer data. The trained model was used to predict consensus GT labels for the 5000 deletions that were randomly selected from the union_refalt vcf. The following is a comparison of model predicted labels [Conesus GT] to consensus genotype generated by the R script for the 5000 randomly selected datapoints from union_refalt.vcf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 741   23   24]\n",
      " [  40 1659  113]\n",
      " [  12   54  615]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "consensus_GT = X6['GTcons']\n",
    "predict = X6['PB_predicted_GTcons_label']\n",
    "print(confusion_matrix(consensus_GT, predict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "X6['GTcons'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "X6['GTcons'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "X6['GTcons'].replace(2, 'Homozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_GTcons_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "X6['PB_predicted_GTcons_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_GTcons_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>1659</td>\n",
       "      <td>40</td>\n",
       "      <td>113</td>\n",
       "      <td>1812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>23</td>\n",
       "      <td>741</td>\n",
       "      <td>24</td>\n",
       "      <td>788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>54</td>\n",
       "      <td>12</td>\n",
       "      <td>615</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>1736</td>\n",
       "      <td>793</td>\n",
       "      <td>752</td>\n",
       "      <td>3281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                  1659                    40   \n",
       "Homozygous_Reference                    23                   741   \n",
       "Homozygous_Variant                      54                    12   \n",
       "All                                   1736                   793   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                 113  1812  \n",
       "Homozygous_Reference                  24   788  \n",
       "Homozygous_Variant                   615   681  \n",
       "All                                  752  3281  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the prediction subset: 0.919\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the prediction subset: {:.3f}'.format(precision_score(consensus_GT, predict, average='micro'))) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** High Confidence Label Analysis**\n",
    "* **Reminder:** The labels predicted by the model are the following consensus genotype:\n",
    "    * Homozygous Reference: 0 \n",
    "    * Heterozygous Variant: 1 \n",
    "    * Homozygous Variant: 2 \n",
    "* Here **high confidence labels** are the labels predicted by the model that were also assigned a predict probability of either 0.9 or 1\n",
    "* The following is an analysis of predicted svanalyzer labels with predict probability >0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "high_conf_labels = X6[(X6['Homozygous_Reference'] == 1) | (X6['Homozygous_Reference'] == 0.9) | (X6['Heterozygous_Variant'] == 1) | (X6['Heterozygous_Variant'] == 0.9) | (X6['Homozygous_Variant'] == 1) | (X6['Homozygous_Variant'] == 0.9)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>1504</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>1550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>4</td>\n",
       "      <td>712</td>\n",
       "      <td>9</td>\n",
       "      <td>725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>455</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>1525</td>\n",
       "      <td>737</td>\n",
       "      <td>492</td>\n",
       "      <td>2754</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                  1504                    18   \n",
       "Homozygous_Reference                     4                   712   \n",
       "Homozygous_Variant                      17                     7   \n",
       "All                                   1525                   737   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                  28  1550  \n",
       "Homozygous_Reference                   9   725  \n",
       "Homozygous_Variant                   455   479  \n",
       "All                                  492  2754  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_GT = high_conf_labels['GTcons']\n",
    "predict = high_conf_labels['PB_predicted_GTcons_label']\n",
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       0.99      0.97      0.98      1550\n",
      "Homozygous_Reference       0.97      0.98      0.97       725\n",
      "  Homozygous_Variant       0.92      0.95      0.94       479\n",
      "\n",
      "         avg / total       0.97      0.97      0.97      2754\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(consensus_GT, predict))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [NIHFAES]",
   "language": "python",
   "name": "Python [NIHFAES]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
